---
title: "Benchmarking approaches"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Goals

Determine the biases introduced by using salmon to quantify pre-mRNA.


To establish the best approach the following analyses will be performed:

1) Examine global pre-mRNA dynamics during development.

```{r load_libs}
source("../../R/globals.r")
```

## get metadata

```{r get_metadata}

mdata1 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "EICHHORN",
                             "GSE83616_run_info_geo.txt"))
mdata2 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("EICHHORN", "RISSLAND")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```

## Get transcript annotations
```{r check_monoexonic}
gtf <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, 
                exon_number, 
                transcript_id, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  mutate(max_exons = max(exon_number))

monoexonic_genes <- dplyr::filter(gtf_summary, 
                                  max_exons == 1) %>%
  pull(gene_id) %>% 
  unique()

monoexonic_transcripts <- dplyr::filter(gtf_summary, 
                                  max_exons == 1) %>%
  pull(transcript_id) %>% 
  unique()

```


## Load in naive method featurecounts data

```{r load_dat, message = F}
count_data_dir <- file.path(data_dir, 
                            "featurecounts",
                            "drosophila")

count_files <- dir(count_data_dir,
                   pattern = "_counts.tsv$",
                   recursive = T,
                   full.names = T)

sample_names <- str_split(count_files, "\\/") %>% 
  map_chr(., ~.x[10]) %>% 
  str_replace(., "_counts.tsv", "")  

sample_names_simple <- str_replace(sample_names, "_exon|_intron", "")
 
dat <- map(count_files, 
           ~read_tsv(.x,
                     skip = 2,  
                     col_names = F) %>% 
             select(gene_id = X1, 
                    length = X6, 
                    counts = X7))
             
# name with libraryID
names(dat) <- sample_names

dat <- bind_rows(dat, .id = "sample_name")

dat <- mutate(dat, 
              Run = str_replace(sample_name, 
                                             "_exon|_intron", 
                                             ""),
              count_type = ifelse(str_detect(sample_name, "exon"),
                                  "exon",
                                  "intron"))
#join with main data
dat <- inner_join(dat, 
                  mdata,
                  by = c("Run"))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = "gene_id")

fcount_dat <- group_by(dat, Run) %>% 
  mutate(lib_size = sum(counts),
         length = as.numeric(length)) %>% 
  mutate(rate = log(counts) - log(length),
         total_rates = log(sum(exp(rate))),
         TPM = exp(rate - total_rates + log(1e6)))

# convert gene symbols to gene id
gsymbol2gid <- gtf %>% 
  tbl_df() %>% 
  filter(type == "gene") %>%
  dplyr::select(gene_id, gene_name) %>% 
  unique()

fcount_dat <- left_join(fcount_dat, gsymbol2gid, 
                        by = c("gene_id" = "gene_name")) %>% 
  dplyr::select(-gene_id) %>% 
  dplyr::rename(gene_id = gene_id.y)

fcount_dat$expt <- "featureCounts"
```

## salmon data 
```{r read_in_salmon}

fa_types <- c(
  "original_gtf",
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon", 
                        "drosophila",
                        mdata$study,
                        .x,
                        mdata$Run,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$Run
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "Run"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("Run"))


# convert gene symbols to gene id
#tx2gid <- gtf %>% 
 # tbl_df() %>% 
#  filter(type == "exon") %>%
#  dplyr::select(gene_id, transcript_id) %>% 
#  unique()

#dat <- left_join(dat, tx2gid,
#                 by = c("Name" = "transcript_id")) %>% 
#  mutate(gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
#                          Name,
#                          gene_id))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_annotations
salmon_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))

```

```{r}
pdata <- mutate(mdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"),
                strain_id = str_replace(strain, "-", "_"))

pdata_split <- split(pdata, pdata$study)

eichorn_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "drosophila",
                   pdata_split$EICHHORN$study,
                   .x,
                   pdata_split$EICHHORN$Run,
                   "quant.sf"))

rissland_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "drosophila",
                   pdata_split$RISSLAND$study,
                   .x,
                   pdata_split$RISSLAND$Run,
                   "quant.sf"))

tids <- read_tsv(eichorn_files[[4]][1]) %>% 
  dplyr::select(Name)

tx2gene <-gtf %>% 
  filter(type == "transcript") %>% 
  select(transcript_id, gene_id) %>% 
  tbl_df() %>% 
  unique()

tx2gene_mapping <- full_join(tids, tx2gene,
                             by = c("Name" = "transcript_id"))

# add in "gene id" for pre mRNA transcript
tx2gene_mapping <- mutate(tx2gene_mapping,
                          gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
                                           Name,
                                           gene_id)) %>% 
  dplyr::rename(transcript_id = Name)
          
                                 
rs_tx_objs <- map(rissland_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))

                                 
ei_tx_objs <- map(eichorn_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))

# remove pre-mRNA annotations
mats <- map(rs_tx_objs, ~.x$abundance[!str_detect(rownames(.x$abundance), "pre_"), ])

plt_dat <- map_dfr(1:ncol(mats[[1]]), 
    ~data_frame(ids = rownames(mats[[1]]), 
                original = mats[[1]][, .x], 
                primary = mats[[2]][, .x],
                primary_masked = mats[[3]][, .x],
                primary_masked_exons = mats[[4]][, .x]),
    
    .id = "col") 
plt_dat <- plt_dat %>%
  gather("primary_type", "expr", primary:primary_masked_exons)


plot_expr_scatter <- function(plt_dat, col_to_plot) {
  plt_summary <- plt_dat %>%
    group_by(col, primary_type) %>%
    summarize(
      n_zero = paste0("there are \n",
                      sum(original > 0 &
                            expr == 0), " genes\non the x axis"),
      x = 1e-2,
      y = 1e4
    ) %>%
    filter(col == col_to_plot)
  
  plt <- ggplot(filter(plt_dat, col == col_to_plot),
                aes(original, expr)) +
    geom_density_2d() +
    geom_point(size = 0.25) +
    geom_abline(aes(slope = 1, intercept = 0)) +
    geom_text(data = plt_summary,
              aes(x = x, y = y, label = n_zero)) +
    facet_wrap( ~ primary_type) +
    scale_x_log10() +
    scale_y_log10() +
    coord_fixed()
  save_plot(paste0(col_to_plot, "_normal_vs_premrna.pdf"),
            plt,
            base_aspect_ratio = 3)
}

walk(unique(plt_dat$col), ~plot_expr_scatter(plt_dat, .x))
```


Figure out what the genes are that have no primary expression 

```{r gtf_attr}

gtf_info <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(exon_number, 
                gene_name, 
                gene_id,
                gene_biotype,
                exon_id,
                transcript_id) %>% 
  group_by(transcript_id) %>% # per transcript exon count
  mutate(max_tx_exons = max(exon_number)) %>% 
  group_by(gene_id, gene_name, gene_biotype) %>% 
  summarize(max_tx_exons = max(max_tx_exons)) %>% # max per transcript exon count
  ungroup()

plt_dat <- left_join(plt_dat, gtf_info, by = c("ids" = "gene_id"))

# pull out primary == 0 expression genes and summarize
weird_genes <- plt_dat %>% 
  filter(primary_type == "primary", 
         expr == 0,
         original > 0) 
  
# these genes look low abundance, perhaps this is just
# due to noise due to low counts

# get counts
weird_gene_counts <- rs_tx_objs[[1]]$counts %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("ids") %>% 
  gather("col", "counts", -ids) %>%
  mutate(col = str_replace(col, "^V", "")) %>% 
  left_join(weird_genes, ., by = c("ids", "col"))

plt <- ggplot(weird_gene_counts, aes(counts)) + 
  geom_density() +
  facet_wrap(~col, ncol = 1, scales = "free_x") +
  labs(y = "count density",
       x = "number of counts per gene")

save_plot("count_distribution_for_weird_genes.pdf", plt,
          ncol = 1, nrow = 10)

```