---
title: "Benchmarking approaches"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Background

Goal is to compare different methods for quantifing intronic abundance. 

1) Naive unique mapping reads in introns versus exons
2) Salmon spiked in with pre-mRNA annotations
3) Salmon spiked in with pre-mRNA annotations, with regions with coverage in the t = 0 
stages masked in the pre-mRNA annotations. Masking converts all sequences to Ns
4) Salmon spiked in with pre-mRNA annotations, with regions with coverage in the t = 0 
stages masked in the pre-mRNA annotations. Additionally mask internal exonic sequences to avoid extensive sequence overlap with the spliced mRNA transcript. Masking converts all sequences to Ns
5) Use bowtie2 for alignment then quantify with salmon, using the annotaitons from 3.
6) Use bowtie2 for alignment then quantify with salmon, using the annotations from 4.

The masking process converts many nucleotides to N which could be problematic for salmon. When building the transcript index, `salmon` will by default insert pseudo-random nucleotides for the N's. The impact of this is unknown and could be problematic. 

To potentially mitigate mismapping artifacts, bowtie2 will be used to map the transcriptome. It will not replace the N's. Salmon can then be run on the alignments to perform quantification.

To establish the best approach the following analyses will be performed:

1) Examine global pre-mRNA dynamics during development.

```{r load_libs}
source("../../R/globals.r")
```

## get metadata

```{r get_metadata}

mdata1 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "EICHHORN",
                             "GSE83616_run_info_geo.txt"))
mdata2 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("EICHHORN", "RISSLAND")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```

## Get transcript annotations
```{r check_monoexonic}
gtf <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, 
                exon_number, 
                transcript_id, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  mutate(max_exons = max(exon_number))

monoexonic_genes <- dplyr::filter(gtf_summary, 
                                  max_exons == 1) %>%
  pull(gene_id) %>% 
  unique()

monoexonic_transcripts <- dplyr::filter(gtf_summary, 
                                  max_exons == 1) %>%
  pull(transcript_id) %>% 
  unique()

```


## Load in naive method featurecounts data

```{r load_dat, message = F}
count_data_dir <- file.path(data_dir, 
                            "featurecounts",
                            "drosophila")

count_files <- dir(count_data_dir,
                   pattern = "_counts.tsv$",
                   recursive = T,
                   full.names = T)

sample_names <- str_split(count_files, "\\/") %>% 
  map_chr(., ~.x[10]) %>% 
  str_replace(., "_counts.tsv", "")  

sample_names_simple <- str_replace(sample_names, "_exon|_intron", "")
 
dat <- map(count_files, 
           ~read_tsv(.x,
                     skip = 2,  
                     col_names = F) %>% 
             select(gene_id = X1, 
                    length = X6, 
                    counts = X7))
             
# name with libraryID
names(dat) <- sample_names

dat <- bind_rows(dat, .id = "sample_name")

dat <- mutate(dat, 
              Run = str_replace(sample_name, 
                                             "_exon|_intron", 
                                             ""),
              count_type = ifelse(str_detect(sample_name, "exon"),
                                  "exon",
                                  "intron"))
#join with main data
dat <- inner_join(dat, 
                  mdata,
                  by = c("Run"))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = "gene_id")

fcount_dat <- group_by(dat, Run) %>% 
  mutate(lib_size = sum(counts),
         length = as.numeric(length)) %>% 
  mutate(rate = log(counts) - log(length),
         total_rates = log(sum(exp(rate))),
         TPM = exp(rate - total_rates + log(1e6)))

# convert gene symbols to gene id
gsymbol2gid <- gtf %>% 
  tbl_df() %>% 
  filter(type == "gene") %>%
  dplyr::select(gene_id, gene_name) %>% 
  unique()

fcount_dat <- left_join(fcount_dat, gsymbol2gid, 
                        by = c("gene_id" = "gene_name")) %>% 
  dplyr::select(-gene_id) %>% 
  dplyr::rename(gene_id = gene_id.y)

fcount_dat$expt <- "featureCounts"
```

## salmon data 
```{r read_in_salmon}

fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon", 
                        "drosophila",
                        mdata$study,
                        .x,
                        mdata$Run,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$Run
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "Run"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("Run"))


# convert gene symbols to gene id
#tx2gid <- gtf %>% 
 # tbl_df() %>% 
#  filter(type == "exon") %>%
#  dplyr::select(gene_id, transcript_id) %>% 
#  unique()

#dat <- left_join(dat, tx2gid,
#                 by = c("Name" = "transcript_id")) %>% 
#  mutate(gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
#                          Name,
#                          gene_id))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_annotations
salmon_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))
```
## bowtie2 data
```{r read_in_bt2_salmon}

fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon_bt2_masked", 
                   "drosophila",
                   mdata$study,
                   .x,
                   mdata$Run,
                   "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$Run
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "Run"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("Run"))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_ annotations
bt2_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))

# append bt2 to expt to distinguish
bt2_dat$expt <- str_c("bt2_", bt2_dat$expt)
```
## combine all expts
```{r}
keep_cols <- intersect(colnames(salmon_dat), colnames(fcount_dat))

dat <- bind_rows(list(salmon_dat[, keep_cols], 
                      fcount_dat[, keep_cols], 
                      bt2_dat[, keep_cols]))

keep_tx_cols <- append(keep_cols, "transcript_id")

tpm_dat <- bind_rows(list(salmon_dat[, keep_tx_cols], 
                      fcount_dat[, keep_cols], 
                      bt2_dat[, keep_tx_cols]))

```

## alignment info

```{r}
library(jsonlite)
fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon", 
                        "drosophila",
                        mdata$study,
                        .x,
                        mdata$Run,
                        "aux_info",
                        "meta_info.json"))

align_dat <- map(files, ~map(.x, jsonlite::fromJSON)) %>% 
  map(., ~map(.x, ~.x$percent_mapped))

names(dat) <- fa_types


map(
  align_dat,
  ~ data_frame(
    percent_alignment = unlist(.x),
    sra_number = mdata$Run,
    stage = mdata$developmental_stage,
    strain = mdata$strain
  )
) %>% bind_rows(., .id = "fasta_type")


```

## DE testing

### Salmon with bt2 alignment
First read in salmon data into tximport
```{r}
fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)


pdata <- mutate(mdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"),
                strain_id = str_replace(strain, "-", "_"))

pdata_split <- split(pdata, pdata$study)

eichorn_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon_bt2_masked", 
                   "drosophila",
                   pdata_split$EICHHORN$study,
                   .x,
                   pdata_split$EICHHORN$Run,
                   "quant.sf"))

rissland_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon_bt2_masked", 
                   "drosophila",
                   pdata_split$RISSLAND$study,
                   .x,
                   pdata_split$RISSLAND$Run,
                   "quant.sf"))

tids <- read_tsv(eichorn_files[[1]][1]) %>% 
  dplyr::select(Name)
tx2gene <-gtf %>% 
  filter(type == "transcript") %>% 
  select(transcript_id, gene_id) %>% 
  tbl_df() %>% 
  unique()

tx2gene_mapping <- full_join(tids, tx2gene,
                             by = c("Name" = "transcript_id"))

# add in "gene id" for pre mRNA transcript
tx2gene_mapping <- mutate(tx2gene_mapping,
                          gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
                                           Name,
                                           gene_id)) %>% 
  dplyr::rename(transcript_id = Name)
          
                                 
rs_tx_objs <- map(rissland_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))

                                 
ei_tx_objs <- map(eichorn_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))
```

Run deseq with post-mzt v. pre-mzt
```{r}
ei_ddobjs <- map(ei_tx_objs, 
              function(x) {
                dds <- DESeqDataSetFromTximport(x, pdata_split$EICHHORN, ~mzt )
                dds <- DESeq(dds)
                dds
              })

rs_ddobjs <- map(rs_tx_objs, 
              function(x) {
                dds <- DESeqDataSetFromTximport(x, pdata_split$RISSLAND, ~mzt + strain_id)
                dds <- DESeq(dds)
                dds
              })

names(ei_ddobjs) <- fa_types
names(rs_ddobjs) <- fa_types
```

Extract results and make MA plots
```{r}
ei_res <- map_dfr(ei_ddobjs, ~results(.x, 
                                       name = "mzt_zygotic_vs_maternal", 
                                       tidy = T), .id = "fasta_type") %>% 
  tbl_df() %>% 
  mutate(count_type = ifelse(startsWith(row, "pre_"), 
                        "intron",
                        "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))

ris_res <- map_dfr(rs_ddobjs, ~results(.x, 
                                       name = "mzt_zygotic_vs_maternal", 
                                       tidy = T), .id = "fasta_type") %>% 
  tbl_df() %>% 
  mutate(count_type = ifelse(startsWith(row, "pre_"), 
                        "intron",
                        "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))



plot_ma <- function(df, plt_title = NULL) {
  n_sig_up <- nrow(dplyr::filter(df, sig == "Up")) %>%
    formatC(big.mark = ",")
  n_sig_down <-
    nrow(dplyr::filter(df, sig == "Down"))  %>%
    formatC(big.mark = ",")
  n_notsig <-
    nrow(dplyr::filter(df, sig == "Not Sig"))  %>%
    formatC(big.mark = ",")
  df <- mutate(df, 
               sig = factor(sig, 
                               levels = c("Down",
                                          "Not Sig",
                                          "Up")))
  pre_plot <- ggplot(df,
                     aes(baseMean,
                         log2FoldChange)) +
    geom_point(aes(colour = sig), size = 0.1) +
    scale_colour_viridis(
      discrete = TRUE,
      name = "",
      drop = FALSE,
      labels = c(
        paste0("Significantly decreased (n = ", n_sig_down, ")"),
        paste0("Not Significant (n= ", n_notsig, ")"),
        paste0("Significantly increased (n = ", n_sig_up, ")")
      )
    ) +
    facet_wrap(~count_type) +
    scale_x_log10() +
    labs(x = expression(paste("log"[10], " Average Expression")),
         y = expression(paste("log"[2], " ", frac(
           "Zygotic", "Maternal"
         ))),
         title = plt_title) +
    guides(colour = guide_legend(nrow = 3,
                                 override.aes = list(size = 3))) +
    theme(legend.position = "top")
  
  pre_plot
}

plts <- imap(split(ei_res, 
           list(ei_res$fasta_type,
                         ei_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("eichorn_ma_plot_salmon_bt2.pdf", plt, nrow = 2, ncol = 3)


plts <- imap(split(ris_res, 
           list(ris_res$fasta_type,
                         ris_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("rissland_ma_plot_salmon_bt2.pdf", plt, nrow = 2, ncol = 3)

all_res <- bind_rows(list(EICHHORN = ei_res,
               RISSLAND = ris_res),
          .id = "experiment")
all_res$method <- "salmon_bt2"
```

### Salmon direct alignment

First read in salmon data into tximport
```{r}
fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

pdata <- mutate(mdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"),
                strain_id = str_replace(strain, "-", "_"))

pdata_split <- split(pdata, pdata$study)

eichorn_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "drosophila",
                   pdata_split$EICHHORN$study,
                   .x,
                   pdata_split$EICHHORN$Run,
                   "quant.sf"))

rissland_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "drosophila",
                   pdata_split$RISSLAND$study,
                   .x,
                   pdata_split$RISSLAND$Run,
                   "quant.sf"))

tids <- read_tsv(eichorn_files[[1]][1]) %>% 
  dplyr::select(Name)
tx2gene <-gtf %>% 
  filter(type == "transcript") %>% 
  select(transcript_id, gene_id) %>% 
  tbl_df() %>% 
  unique()

tx2gene_mapping <- full_join(tids, tx2gene,
                             by = c("Name" = "transcript_id"))

# add in "gene id" for pre mRNA transcript
tx2gene_mapping <- mutate(tx2gene_mapping,
                          gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
                                           Name,
                                           gene_id)) %>% 
  dplyr::rename(transcript_id = Name)
          
                                 
rs_tx_objs <- map(rissland_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))

                                 
ei_tx_objs <- map(eichorn_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))
```

Run deseq with post-mzt v. pre-mzt
```{r}
ei_ddobjs <- map(ei_tx_objs, 
              function(x) {
                dds <- DESeqDataSetFromTximport(x, 
                                                pdata_split$EICHHORN, 
                                                ~mzt )
                dds <- DESeq(dds)
                dds
              })

rs_ddobjs <- map(rs_tx_objs, 
              function(x) {
                dds <- DESeqDataSetFromTximport(x, 
                                                pdata_split$RISSLAND, 
                                                ~mzt + strain_id)
                dds <- DESeq(dds)
                dds
              })

names(ei_ddobjs) <- fa_types
names(rs_ddobjs) <- fa_types
```

Extract results and make MA plots
```{r}
ei_res <- map_dfr(ei_ddobjs, ~results(.x, 
                                       name = "mzt_zygotic_vs_maternal", 
                                       tidy = T), 
                  .id = "fasta_type") %>% 
  tbl_df() %>% 
  mutate(count_type = ifelse(startsWith(row, "pre_"), 
                        "intron",
                        "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))

ris_res <- map_dfr(rs_ddobjs, ~results(.x, 
                                       name = "mzt_zygotic_vs_maternal", 
                                       tidy = T), 
                   .id = "fasta_type") %>% 
  tbl_df() %>% 
  mutate(count_type = ifelse(startsWith(row, "pre_"), 
                        "intron",
                        "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))



# plot_ma <- function(df, plt_title = NULL) {
#   n_sig_up <- nrow(dplyr::filter(df, sig == "Up")) %>%
#     formatC(big.mark = ",")
#   n_sig_down <-
#     nrow(dplyr::filter(df, sig == "Down"))  %>%
#     formatC(big.mark = ",")
#   n_notsig <-
#     nrow(dplyr::filter(df, sig == "Not Sig"))  %>%
#     formatC(big.mark = ",")
#   
#   pre_plot <- ggplot(df,
#                      aes(baseMean,
#                          log2FoldChange)) +
#     geom_point(aes(colour = sig), size = 0.1) +
#     scale_colour_viridis(
#       discrete = TRUE,
#       name = "",
#       labels = c(
#         paste0("Significantly decreased (n = ", n_sig_down, ")"),
#         paste0("Not Significant (n= ", n_notsig, ")"),
#         paste0("Significantly increased (n = ", n_sig_up, ")")
#       )
#     ) +
#     facet_wrap(~count_type) +
#     scale_x_log10() +
#     labs(x = "Average Expression",
#          y = expression(paste("log"[2], " ", frac(
#            "Post-MZT", "Pre-MZT"
#          ))),
#          title = plt_title) +
#     guides(colour = guide_legend(nrow = 3,
#                                  override.aes = list(size = 3))) +
#     theme(legend.position = "top")
#   
#   pre_plot
# }

plts <- imap(split(ei_res, 
           list(ei_res$fasta_type,
                       ei_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("eichorn_ma_plot_salmon.pdf", plt, nrow = 2, ncol = 3)

plts <- imap(split(ris_res, 
           list(ris_res$fasta_type,
                         ris_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("rissland_ma_plot_salmon.pdf", plt, nrow = 2, ncol = 3)


res <- bind_rows(list(EICHHORN = ei_res,
               RISSLAND = ris_res),
          .id = "experiment")
res$method <- "salmon"

all_res <- bind_rows(all_res, res)
```
### Featurecounts

```{r}
eichorn_files <- map(c("exon", "intron"),
                     ~file.path(data_dir,
                   "featurecounts", 
                   "drosophila",
                   pdata_split$EICHHORN$study,
                   paste0(pdata_split$EICHHORN$Run, 
                          "_",
                          .x,
                          "_counts.tsv")))

rissland_files <- map(c("exon", "intron"),
                     ~file.path(data_dir,
                   "featurecounts", 
                   "drosophila",
                   pdata_split$RISSLAND$study,
                   paste0(pdata_split$RISSLAND$Run, 
                          "_",
                          .x,
                          "_counts.tsv")))
ei_count_dat <- pmap(eichorn_files, 
                  function(x, y){
                    exon <- read_tsv(x, 
                                     skip = 2, 
                                     col_names = F) %>% 
                      dplyr::select(X1, X7)
                    intron <- read_tsv(y,
                                       skip = 2, 
                                       col_names = F) %>% 
                      dplyr::select(X1, X7) %>% 
                      mutate(X1 = str_c("pre_", X1))
                    bind_rows(exon, intron)
                  })

names(ei_count_dat) <- pdata_split$EICHHORN$Run


ris_count_dat <- pmap(rissland_files, 
                  function(x, y){
                    exon <- read_tsv(x, 
                                     skip = 2, 
                                     col_names = F) %>% 
                      dplyr::select(X1, X7)
                    intron <- read_tsv(y,
                                       skip = 2, 
                                       col_names = F) %>% 
                      dplyr::select(X1, X7) %>% 
                      mutate(X1 = str_c("pre_", X1))
                    bind_rows(exon, intron)
                  })
names(ris_count_dat) <- pdata_split$RISSLAND$Run

ei_count_dat <- imap(ei_count_dat, function(x, y){
  colnames(x) <- c("gene_id", y)
  x
})

ris_count_dat <- imap(ris_count_dat, function(x, y){
  colnames(x) <- c("gene_id", y)
  x
})

ei_count_mat <- Reduce(function(x, y) {
  inner_join(x, y, by = "gene_id")}, 
  ei_count_dat)

ris_count_mat <- Reduce(function(x, y) {
  inner_join(x, y, by = "gene_id")}, 
  ris_count_dat)

ei_count_mat <- as.data.frame(ei_count_mat) %>% 
  tibble::column_to_rownames("gene_id")
ris_count_mat <- as.data.frame(ris_count_mat) %>% 
  tibble::column_to_rownames("gene_id")
```


Run deseq with post-mzt v. pre-mzt
```{r}
ei_ddobj <- DESeqDataSetFromMatrix(ei_count_mat, 
                                   pdata_split$EICHHORN,
                                   ~ mzt)
ei_ddobj <- DESeq(ei_ddobj)

ris_ddobj <- DESeqDataSetFromMatrix(ris_count_mat, 
                                   pdata_split$RISSLAND,
                                   ~ mzt + strain_id)
ris_ddobj <- DESeq(ris_ddobj)
```

plot ma

```{r}
ddojbs <- list(EICHHORN = ei_ddobj,
               RISSLAND = ris_ddobj)

fc_res <- map_dfr(ddojbs,
                  ~ results(.x,
                            name = "mzt_zygotic_vs_maternal",
                            tidy = T),
                  .id = "Experiment") %>%
  tbl_df() %>%
  mutate(
    count_type = ifelse(startsWith(row, "pre_"),
                        "intron",
                        "exon"),
    sig = ifelse(
      padj < 0.05 & log2FoldChange > 0,
      "Up",
      ifelse(padj < 0.05 & log2FoldChange < 0,
             "Down", "Not Sig")
    )
  ) %>%
  filter(!is.na(sig))

plts <- imap(split(fc_res, 
           list(fc_res$Experiment,
           fc_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("both_ma_plot_featurecounts.pdf", plt, nrow = 2, ncol = 2)

res <- fc_res %>% dplyr::rename(experiment = Experiment)
res$method <- "featurecounts"
res$fasta_type <- "genome"
res <- res[, colnames(all_res)]

# remove pre annotation from gene symbols for intron counts
res <- mutate(res, 
              row = ifelse(count_type == "intron",
                           str_replace(row, "^pre_", ""),
                           row))

# convert gene symbols to gene id
gsymbol2gid <- gtf %>% 
  tbl_df() %>% 
  filter(type == "gene") %>%
  dplyr::select(gene_id, gene_name) %>% 
  unique()

res <- left_join(res, gsymbol2gid,
                 by = c("row" = "gene_name")) %>% 
  mutate(row = gene_id) %>% 
  dplyr::select(-gene_id)

all_res <- bind_rows(all_res, res)
```

### summarize results

```{r}
plt_dat <- filter(all_res, count_type == "intron")

plts <- imap(split(plt_dat, plt_dat$experiment),
    ~ggplot(.x, aes(method)) +
  geom_bar(aes(fill = fasta_type), 
           position = "dodge") +
  facet_wrap(~sig, scales = "free_y") +
  scale_fill_viridis_d(name = "") +
  scale_y_continuous(labels = scales::comma) +
  labs(x = "",
       y = "Number of genes",
       title = .y) +
  theme(legend.pos = "right",
        axis.text.x = element_text(angle = 90)))

plt <- plot_grid(plotlist = plts, nrow = 2, ncol = 1)

save_plot("sig_intron_summary.pdf", plt, nrow = 2, base_aspect_ratio = 2.5)
plt
```


```{r}
gene_lists <- split(plt_dat, 
                    list(plt_dat$experiment)) %>%
  map(., ~split(.x, .x$sig)) %>% 
  at_depth(., 2, ~split(.x, list(.x$method, .x$fasta_type))) %>% 
  at_depth(., 3, ~dplyr::pull(.x, row) %>% 
             ifelse(str_detect(., "^pre_"),
                    .,
                    str_c("pre_", .)))  %>% 
  modify_depth(., 3, ~.x[length(.x) > 0])
  
library(UpSetR)

to_plot <- c("featurecounts.genome",
             "salmon.primary_transcripts",
             "salmon_bt2.primary_transcripts",
             "salmon.primary_transcripts_masked",
             "salmon_bt2.primary_transcripts_masked",
             "salmon.primary_transcripts_masked_exons",
             "salmon_bt2.primary_transcripts_masked_exons")

iwalk(gene_lists,
    function(expt, name){
      iwalk(expt, 
           function(x, id){
              pdf(paste0("de_overlap_", name, "_", id, ".pdf"))
             sets_to_plot <- to_plot[to_plot %in% names(x)]
             upset(fromList(x), sets = sets_to_plot, 
                   nsets = length(x), 
                   nintersects = 20, order.by = "freq")
             dev.off()
           })
    })


```

## Compare to eisen data


```{r classify_txs_mzt}

mzt_status <- dplyr::select(pdata, study, Run, developmental_stage, mzt)

## organize tpm data
tpm_dat <- tpm_dat %>% 
  left_join(mzt_status, by = c("study", 
                               "Run", 
                               "developmental_stage"))

## sum up transcript TPMs to gene-level
tpm_dat_summary <- group_by(tpm_dat,
    expt,
    Run,
    study,
    developmental_stage,
    strain,
    source_name,
    count_type,
    gene_id,
    mzt) %>%
  summarize(TPM = sum(TPM)) %>%
  ungroup()

tpm_dat_simple <- tpm_dat_summary %>% 
  group_by(expt,
    study,
    count_type,
    gene_id,
    mzt) %>% 
  summarize(mean_tpm = mean(TPM)) %>% 
  ungroup()
  

tpm_dat_simple <- mutate(tpm_dat_simple,
                         method = ifelse(str_detect(expt, "bt2"),
                                      "salmon_bt2",
                                      ifelse(expt == "featureCounts",
                                             "featurecounts",
                                             "salmon")),
                         fasta_type = str_replace(expt, "^bt2_", ""),
                         fasta_type = ifelse(expt == "featureCounts",
                                             "genome",
                                             fasta_type)) %>% 
  dplyr::select(-expt) %>% 
  dplyr::rename(experiment = study)

tpm_dat_simple <- spread(tpm_dat_simple, mzt, mean_tpm)

# fix gene_ids to remove pre_ prefix if it exists
all_res_simple <- all_res %>% 
  dplyr::select(-c(lfcSE, stat, pvalue, baseMean)) %>% 
  mutate(row = ifelse(count_type == "intron",
                     str_replace(row, "^pre_", ""),
                     row))
  

all_res_w_tpm <- left_join(all_res_simple, tpm_dat_simple, 
               by = c("experiment",
                      "count_type",
                      "row" = "gene_id",
                      "method",
                      "fasta_type")) %>% 
  dplyr::rename(pre_mzt = maternal,
                post_mzt = zygotic,
                gene_id = row)


```

```{r}

mzt_class_w_tpm <- mutate(all_res_w_tpm, 
                    mzt_class = ifelse(pre_mzt > 1 & 
                                         log2FoldChange > 0 &
                                         padj < 0.05,
                           "matzyg", 
                           ifelse(pre_mzt <= 1 & 
                                    log2FoldChange > 0 &
                                    padj < 0.05,
                                  "zyg",
                                  ifelse(pre_mzt > 1,
                                         "mat", 
                                         NA))))

mzt_class <- dplyr::select(mzt_class_w_tpm, 
                           experiment, 
                           fasta_type,
                           gene_id,
                           count_type, 
                           method,
                           mzt_class)    

mzt_class <- spread(mzt_class, count_type, mzt_class)


mzt_class <- filter(mzt_class, !(is.na(exon) & is.na(intron))) %>%
  mutate(mzt_class = ifelse(is.na(intron),
                            exon,
                            ifelse(
                              is.na(exon),
                              intron,
                              ifelse(
                                str_detect(intron,
                                           "zyg") &
                                  exon %in% c("mat",
                                              "matzyg"),
                                "matzyg",
                                ifelse(intron == "zyg" &
                                        exon == "zyg",
                                       "zyg",
                                       NA)
                              )
                            )))

                       
gene2tx <- gtf[, c("gene_id", "gene_name")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)
gene2tx <- as_data_frame(gene2tx)

mzt_class <- left_join(mzt_class, gene2tx, by = "gene_id") %>% 
  unique()

mzt_class %>% 
  group_by(experiment, fasta_type, method, mzt_class) %>% 
  summarize(n())
```

```{r}
download.file("https://ndownloader.figshare.com/files/400050",
              "eisen_data.tsv")

edata <- read_tsv("eisen_data.tsv")

```

```{r}

cas_cols <- str_subset(colnames(edata), "cas")

edata_confident <- edata[rowSums(edata[, cas_cols], na.rm = T) > 0 ,]

res_eisen <- edata_confident %>% 
  dplyr::select(NAME, CLASS)

res_eisen <- inner_join(mzt_class, res_eisen, 
            by = c("gene_name" = "NAME")) %>% 
  dplyr::rename(kr_class = mzt_class,
                eisen_class = CLASS)

summary_dat <- res_eisen %>% 
 group_by(kr_class, eisen_class,
          experiment, fasta_type, method) %>% 
 summarize(n())


classification_summary <- res_eisen %>% 
  mutate(eisen_class = ifelse(is.na(eisen_class),
                              "not classified",
                              eisen_class),
         kr_class = ifelse(is.na(kr_class),
                                 "not classified",
                                 kr_class),
         kr_class = factor(kr_class,
                           levels = c(
                             "mat",
                             "matzyg",
                             "zyg",
                             "not classified"
                           )))

classification_summary <- classification_summary %>% 
  filter(!gene_id %in% monoexonic_genes)

eichhorn_res <- filter(classification_summary, 
                       experiment == "EICHHORN")
rissland_res <- filter(classification_summary, 
                       experiment == "RISSLAND")

plt <- ggplot(eichhorn_res, aes(kr_class)) +
  geom_bar(aes(fill = eisen_class)) +
  facet_grid(method~fasta_type) +
  labs(y = "# of genes", 
       x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

plt

save_plot("eisen_vs_eichhorn_classification.pdf", plt,
          base_height = 7, base_aspect_ratio = 1.75)

plt <- ggplot(rissland_res, aes(kr_class)) +
  geom_bar(aes(fill = eisen_class)) +
  facet_grid(method~fasta_type) +
  labs(y = "# of genes", 
       x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

save_plot("eisen_vs_rissland_classification.pdf", plt,
          base_height = 7, base_aspect_ratio = 1.75)

plt <- ggplot(rissland_res, aes(exon)) +
  geom_bar(aes(fill = eisen_class)) +
  facet_grid(method~fasta_type) +
  labs(y = "# of genes", 
       x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

save_plot("eisen_vs_rissland_classification_by_exon.pdf", plt,
          base_height = 7, base_aspect_ratio = 1.75)


plt <- ggplot(rissland_res, aes(intron)) +
  geom_bar(aes(fill = eisen_class)) +
  facet_grid(method~fasta_type) +
  labs(y = "# of genes", 
       x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

save_plot("eisen_vs_rissland_classification_by_intron.pdf", plt,
          base_height = 7, base_aspect_ratio = 1.75)

plt <- ggplot(edata_confident, aes(CLASS)) +
  geom_bar(aes(fill = CLASS)) +
  labs(y = "# of genes", 
       x = "") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

save_plot("eisen_classification.pdf", plt,
          base_height = 7)
plt
```


## Plot out TPMs for Eisen classes

```{r tidy_eisen}
tidy_eisen <- gather(edata_confident, genotype, expr, 
                     -c(NAME:`F:M SLOPE`)) %>% 
  filter(str_detect(genotype, "^w1_|^cas_" )) %>% 
  separate(genotype, c("strain", "timepoint"), sep = "_") %>% 
  separate(timepoint, c("sex", "timepoint"), sep = 1) %>% 
  group_by(CLASS, NAME, strain, timepoint) %>% 
  summarize(expr = mean(expr)) 
```

Next plot out a few genes to verify that patterns are the same as the paper.
```{r}

pub_genes <- c(
  "Adh", # maternal
  "aur",
  "Nap1",
  "Neu3", # maternal and zygo
  "foi",
  "casp",
  "Bro", # zygotic
  "SCNF",
  "corn"
)

tmp <- filter(tidy_eisen, NAME %in% pub_genes)
plt <- ggplot(tmp, aes(timepoint, expr)) +
  geom_point(aes(color = strain)) +
  scale_color_brewer(palette = "Set1") +
  facet_wrap(~NAME + CLASS, scales = "free_y", nrow = 3) +
  labs(y = "Expression")
plt
save_plot("eisen_example_genes.pdf", plt, base_height = 7, base_aspect_ratio = 1.)

tmp <- filter(tidy_eisen, !is.na(CLASS)) %>% 
  group_by(., timepoint, CLASS, strain) %>% 
  summarize(expr = mean(expr))
ggplot(tmp, aes(timepoint, expr)) +
  geom_point(aes(color = strain)) +
  facet_wrap(~CLASS, scales = "free_y") +
  labs(y = "Avg. Expression")


group_by(tidy_eisen, timepoint, CLASS, strain) %>% 
  summarize(min(expr), max(expr))

eisen_rissland <- left_join(tidy_eisen, rissland_res,
                            by = c("NAME" = "gene_name")) %>% 
  filter(!(is.na(CLASS) | kr_class == "not classified"))

tmp <- filter(eisen_rissland, !is.na(kr_class)) %>% 
  group_by(timepoint, kr_class, strain) %>% 
  summarize(expr = mean(expr))
plt1 <- ggplot(tmp, aes(timepoint, expr)) +
  geom_point(aes(color = strain)) +
    scale_color_brewer(palette = "Set1") +
  facet_wrap(~kr_class, scales = "free_y") +
  labs(title = "intron classification",
       y = "Avg. Expression")

tmp <- group_by(eisen_rissland, timepoint, CLASS, strain) %>% 
  summarize(expr = mean(expr))
plt2 <- ggplot(tmp, aes(timepoint, expr)) +
  geom_point(aes(color = strain)) +
    scale_color_brewer(palette = "Set1") +
  facet_wrap(~CLASS, scales = "free_y") +
  labs(title = "eisen classification",
       y = "Avg. Expression")

plt <- plot_grid(plt1, plt2, nrow = 2)
save_plot("compare_classes.pdf", plt, nrow = 2, base_aspect_ratio = 2.33)
plt
```


```{r}
eisen_gene_annotation <- edata_confident %>% 
  dplyr::select(NAME, CLASS) %>% 
  unique()

kr_class <- dplyr::select(classification_summary, 
                          experiment, 
                          fasta_type, 
                          gene_id, gene_name, 
                          method) %>% 
  unique()

tpms <- left_join(tpm_dat_summary, gsymbol2gid)
tpms_simple <- mutate(tpms,
                         method = ifelse(str_detect(expt, "bt2"),
                                      "salmon_bt2",
                                      ifelse(expt == "featureCounts",
                                             "featurecounts",
                                             "salmon")),
                         fasta_type = str_replace(expt, "^bt2_", ""),
                         fasta_type = ifelse(expt == "featureCounts",
                                             "genome",
                                             fasta_type)) %>% 
  dplyr::select(-expt) %>% 
  dplyr::rename(experiment = study)

tpm_class <- left_join(tpms_simple, 
                       kr_class, 
                       c("gene_id",
                         "gene_name",
                         "experiment",
                         "method",
                         "fasta_type"))

tpm_class <- left_join(tpm_class, mzt_class)
tpm_class <- left_join(tpm_class, 
                       ungroup(tidy_eisen) %>% select(CLASS, NAME) %>% unique(), by = c("gene_name" = "NAME"))
#tpm_class <- filter(tpm_class, 
#                    !is.na(CLASS))

tpm_class <- mutate(tpm_class, 
                       developmental_stage = factor(developmental_stage, 
                                         levels = c(
                                           "Stage 11 oocyte",
                                           "Stage 12 oocyte",
                                           "Stage 13 oocyte",
                                           "Stage 14 oocyte",
                                           "Activated egg",
                                           "0-1 hr embryo",
                                           "1-2 hr embryo",
                                           "2-3 hr embryo",
                                           "3-4 hr embryo",
                                           "4-5 hr embryo",
                                           "5-6 hr embryo")
                                         )
                       )


tpm_filtered <- tpm_class %>% 
    filter(experiment == "RISSLAND",
           method == "salmon_bt2", 
           fasta_type == "primary_transcripts_masked") 

tmp <- tpm_filtered %>% 
  filter(gene_name %in% pub_genes) %>% 
  group_by(developmental_stage, count_type, CLASS, mzt_class, gene_name) %>% 
  summarize(TPM = mean(TPM)) 

ggplot(tmp,
       aes(developmental_stage, TPM)) +
  geom_point(aes(color = count_type)) +
  facet_wrap(~gene_name + CLASS + mzt_class, scales = "free_y") +
  theme(axis.text.x = element_text(angle = 90))



tmp <- filter(tpm_filtered) %>% 
  group_by(developmental_stage, count_type, CLASS) %>% 
  summarize(expr = mean(TPM))
plt1 <- ggplot(tmp, aes(developmental_stage, expr)) +
  geom_point(aes(color = count_type)) +
    scale_color_brewer(palette = "Set1") +
  facet_wrap(~CLASS, scales = "free_y") +
  labs(title = "eisen classification",
       y = "Avg. Expression")

tmp <- group_by(tpm_filtered, 
                developmental_stage, count_type, mzt_class) %>% 
  summarize(expr = mean(TPM))
plt2 <- ggplot(tmp, aes(developmental_stage, expr)) +
  geom_point(aes(color = count_type)) +
  scale_color_brewer(palette = "Set1") +
  facet_wrap(~mzt_class, scales = "free_y") +
  labs(title = "intron classification",
       y = "Avg. Expression")

plt <- plot_grid(plt1, plt2, nrow = 2)
save_plot("compare_classes_rissland.pdf", plt, nrow = 2, base_aspect_ratio = 2.33)
plt
```


## Try new classification

```{r}
url_name <- "https://datadryad.org/bitstream/handle/10255/dryad.52325/supplemental_file_1.xlsx?sequence=1"

fn_name <- "supplemental_file_1.xlsx"
download.file(url_name, fn_name)
pol2chip_dat <- readxl::read_excel(fn_name, sheet = 2)

matzyg_classification <- filter(pol2chip_dat, !is.na(MBT_group)) %>% 
  dplyr::select(gene_id = fb_gene_id, 
                gene_name = fb_symbol, 
                MBT_group)

res_eisen_reclass <- edata_confident %>% 
  dplyr::select(NAME, CLASS)

res_eisen_reclass <- left_join(res_eisen_reclass,
                               matzyg_classification, 
                               by = c("NAME" = "gene_name"))

res_eisen_reclass <- mutate(res_eisen_reclass,
                            new_class = ifelse(!is.na(MBT_group) & 
                                                 CLASS == "mat",
                                               "matzyg",
                                               CLASS)) %>% 
  dplyr::select(NAME, new_class)

eisen_tpm_dat <- left_join(eisen_rissland, res_eisen_reclass, by = "NAME")
rissland_tpm_dat <- left_join(tpm_filtered, res_eisen_reclass, 
                              by = c("gene_name" = "NAME"))

## combine cas and w1 values and generate a matrix for vis.
eisen_mat <- eisen_tpm_dat %>% 
  select(NAME, timepoint, strain, expr) %>% 
  unique() %>% 
  group_by(timepoint, NAME) %>% 
  summarize(expr = sum(expr)) %>% 
  spread(timepoint, expr) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("NAME") %>% 
  as.matrix()  
eisen_mat <- log10(eisen_mat + 1)

eisen_mat_classes <- eisen_tpm_dat %>% 
  ungroup() %>% 
  select(new_class, CLASS, NAME) %>% 
  unique() %>% 
  left_join(data_frame(NAME = rownames(eisen_mat)), # reorder to match matrix above
            ., by = "NAME")
  
Heatmap(eisen_mat, 
        show_row_dend = FALSE, 
        show_column_dend = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        split = eisen_mat_classes$CLASS)

## Note that the RPKM values in the supplement are similar but do not match the RPKM values in the GEO repository.

```


```{r}
rs_exon_intron_comp <- rissland_tpm_dat %>% 
  group_by(gene_name, developmental_stage, count_type) %>% #average replicate timepoints
  summarize(TPM = mean(TPM)) %>% 
  ungroup() %>% 
  spread(count_type, TPM)

rs_exon_intron_comps <- list()
rs_exon_intron_comps$exon <- rs_exon_intron_comp %>% 
  select(-intron) %>% 
  spread(developmental_stage, exon) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("gene_name") %>% 
  as.matrix()

rs_exon_intron_comps$intron  <- rs_exon_intron_comp %>% 
    select(-exon) %>% 
  spread(developmental_stage, intron) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("gene_name") %>% 
  as.matrix()

rs_exon_intron_comps_log <- map(rs_exon_intron_comps, 
                            ~log1p(.x))

rissland_mat_annot <- rissland_tpm_dat %>% 
  ungroup() %>% 
  select(gene_name, exon, intron, mzt_class, CLASS, new_class) %>% 
  unique() %>% 
  left_join(data_frame(gene_name = rownames(rs_exon_intron_comps_log$exon)), # reorder to match matrices above
            ., by = "gene_name") %>% 
  filter(!is.na(mzt_class))

# only keep genes annotated by intron class
rissland_mat_annot <- rissland_mat_annot %>% 
  filter(!is.na(mzt_class))
  
rs_exon_intron_comps_log <- map(rs_exon_intron_comps_log, 
                          ~.x[rownames(.x) %in% rissland_mat_annot$gene_name, ])


## for monoexonic genes, set intron count to NA
## plot heatmaps side by side
## first all annotations (i.e. maternal, mat.zyg, zg)
## then classify zyg into each timepoint and replot just zyg


## All classes
h1 <- Heatmap(rs_exon_intron_comps_log$exon, 
                            na_col = "black",
              name = "Exon Abundance",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        show_row_names = FALSE,
                column_title = "Exon Expression",
        split = rissland_mat_annot$mzt_class)

h2 <- Heatmap(rs_exon_intron_comps_log$intron, name = "Intron Abundance",
              na_col = "black",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        column_title = "Intron Expression",
        split = rissland_mat_annot$mzt_class)

outdir <- "exon_intron_heatmaps"
dir.create(outdir)
pdf(file.path(outdir, "all_classes.pdf"))
draw(h1 + h2)
dev.off()

## mat-zyg

mz_genes <- rissland_mat_annot %>% 
  filter(mzt_class == "matzyg") %>% 
  pull(gene_name)

rs_exon_intron_comps_matzyg <- map(rs_exon_intron_comps, 
                                ~.x[mz_genes, ])

### make new classifcation based on first intronic apperance

intron_apperance <- rs_exon_intron_comps_matzyg$intron > 0.1 
intron_apperance <- as.data.frame(intron_apperance) %>% 
  tibble::rownames_to_column("gene_name") %>% 
  tbl_df() %>% 
  gather(dev_stage, intron_expr, -gene_name)

matzyg_intron_class <- intron_apperance %>% 
  filter(intron_expr) %>% 
  group_by(gene_name) %>% 
  arrange(dev_stage, .by_group = T) %>% 
  summarize(first_expr = dplyr::first(dev_stage)) %>% 
  mutate(first_expr = str_replace(first_expr, " hr embryo", ""))

#drop mRNAs without introns 
rs_exon_intron_comps_matzyg <- map(rs_exon_intron_comps_matzyg, 
                                   function(x){
                                     log1p(x[matzyg_intron_class$gene_name, ])})

h2 <- Heatmap(rs_exon_intron_comps_matzyg$intron, name = "Intron Abundance",
              na_col = "black",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        split = matzyg_intron_class$first_expr,
        column_title = "Intron Expression")

h1 <- Heatmap(rs_exon_intron_comps_matzyg$exon, 
                            na_col = "black",
              name = "Exon Abundance",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        show_row_names = FALSE,
        split = matzyg_intron_class$first_expr, 
        column_title = "Exon Expression")

 h2 + h1
pdf(file.path(outdir, "matzyg_classes.pdf"))
draw(h2 + h1)
dev.off()
 
 ## zyg only
z_genes <- rissland_mat_annot %>% 
  filter(mzt_class == "zyg") %>% 
  pull(gene_name)

rs_exon_intron_comps_zyg <- map(rs_exon_intron_comps, 
                                ~.x[z_genes, ])

### make new classifcation based on first intronic apperance
intron_apperance <- rs_exon_intron_comps_zyg$intron > 0.1 
intron_apperance <- as.data.frame(intron_apperance) %>% 
  tibble::rownames_to_column("gene_name") %>% 
  tbl_df() %>% 
  gather(dev_stage, intron_expr, -gene_name)

zyg_intron_class <- intron_apperance %>% 
  filter(intron_expr) %>% 
  group_by(gene_name) %>% 
  arrange(dev_stage, .by_group = T) %>% 
  summarize(first_expr = dplyr::first(dev_stage)) %>% 
  mutate(first_expr = str_replace(first_expr, " hr embryo", ""))

#drop mRNAs without introns 
rs_exon_intron_comps_zyg <- map(rs_exon_intron_comps_zyg, 
                                   function(x){
                                     log1p(x[zyg_intron_class$gene_name, ])})

h2 <- Heatmap(rs_exon_intron_comps_zyg$intron, name = "Intron Abundance",
              na_col = "black",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        split = zyg_intron_class$first_expr,
        column_title = "Intron Expression")

h1 <- Heatmap(rs_exon_intron_comps_zyg$exon, 
                            na_col = "black",
              name = "Exon Abundance",
        show_row_dend = F, 
        show_column_dend = F,
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        show_row_names = FALSE,
        split = zyg_intron_class$first_expr, 
        column_title = "Exon Expression")

 h2 + h1
 
 pdf(file.path(outdir, "zyg_classes.pdf"))
draw(h2 + h1)
dev.off()
```

### compare classes

```{r}
mzt_classes <- rissland_mat_annot %>% 
  filter(!is.na(CLASS)) %>% 
  dplyr::select(gene_name, 
                intron_classification = mzt_class, 
                eisen_classification = CLASS,
                eisen_with_pol2 = new_class) %>% 
  gather(classification_method, classification, -gene_name, -intron_classification)

plt <- ggplot(mzt_classes, aes(intron_classification)) +
  geom_bar(aes(fill = classification)) +
  facet_wrap(~classification_method) +
  labs(y = "# of genes", 
       x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5))

save_plot("compare_eisen_classes_w_wo_pol2.pdf", plt, base_aspect_ratio = 2)
```

Check out early group of zyg genes
```{r}
prembt <- filter(pol2chip_dat, !is.na(preMBT_group)) %>% 
  pull(fb_symbol)
zyg_intron_class <- zyg_intron_class %>% 
  mutate(pol2 = ifelse(gene_name %in% prembt,
                       "early_zyg",
                       "not_early_zyg"))

ggplot(zyg_intron_class, aes(first_expr)) +
  geom_bar(aes(fill = pol2))
```

```{r}
early_zyg <- read_delim("derenzis_sup_tbl_8.txt", col_names = F, skip = 1, delim = " ")
early_zyg <- dplyr::select(early_zyg, gene_id = X2)
early_zyg <- left_join(early_zyg, gtf) %>% 
  select(gene_id, gene_name) %>% 
  unique() %>% 
  mutate(microarray = "early_zyg")

zyg_intron_class <- zyg_intron_class %>% 
  mutate(microarray = ifelse(gene_name %in% early_zyg$gene_name,
                       "early_zyg",
                       "not_early_zyg"))

ggplot(zyg_intron_class, aes(first_expr)) +
  geom_bar(aes(fill = microarray))


left_join(rissland_mat_annot, 
          early_zyg) %>% 
  filter(!is.na(microarray)) %>%
  group_by(mzt_class) %>% 
  summarize(n = n())


filter(rissland_mat_annot, 
          gene_name %in% prembt) %>% 
  group_by(mzt_class) %>% 
  summarize(n = n())
```