---
title: "Benchmarking approaches"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Background

Goal is to compare different methods for quantifing intronic abundance. 

1) Naive unique mapping reads in introns versus exons
2) Salmon spiked in with pre-mRNA annotaitons
3) Salmon spiked in with pre-mRNA annotations, with regions with coverage in the t = 0 
stages masked in the pre-mRNA annotations. Masking converts all sequences to Ns
4) Salmon spiked in with pre-mRNA annotations, with regions with coverage in the t = 0 
stages masked in the pre-mRNA annotations. Additionally mask internal exonic sequences to avoid extensive sequence overlap with the spliced mRNA transcript. Masking converts all sequences to Ns
5) Use bowtie2 for alignment then quantify with salmon, using the annotaitons from 3.
6) Use bowtie2 for alignment then quantify with salmon, using the annotaitons from 4.

The masking process converts many nucleotides to N which could be problematic for salmon. When building the transcript index, `salmon` will by default insert pseudo-random nucleotides for the N's. The impact of this is unknown and could be problematic. 

To potentially mitigate mismapping artifacts, bowtie2 will be used to map the transcriptome. It will not replace the N's. Salmon can then be run on the alignments to perform quantification.

To establish the best approach the following analyses will be performed:

1) Examine global pre-mRNA dynamics during development.

```{r load_libs}
source("../../R/globals.r")
```

## get metadata

```{r get_metadata}

mdata1 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "EICHHORN",
                             "GSE83616_run_info_geo.txt"))
mdata2 <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("EICHHORN", "RISSLAND")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```

## Get transcript annotations
```{r check_monoexonic}
gtf <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  bed_merge() %>% 
  dplyr::summarize(total_exons = n())

monoexonic_genes <- dplyr::filter(gtf_summary, total_exons == 1) %>%
  pull(gene_id)
```


## Load in naive method featurecounts data

```{r load_dat, message = F}
count_data_dir <- file.path(data_dir, 
                            "featurecounts",
                            "drosophila")

count_files <- dir(count_data_dir,
                   pattern = "_counts.tsv$",
                   recursive = T,
                   full.names = T)

sample_names <- str_split(count_files, "\\/") %>% 
  map_chr(., ~.x[10]) %>% 
  str_replace(., "_counts.tsv", "") 

sample_names_simple <- str_replace(sample_names, "_exon|_intron", "")
 
dat <- map(count_files, 
           ~read_tsv(.x,
                     skip = 2,  
                     col_names = F) %>% 
             select(gene_id = X1, 
                    length = X6, 
                    counts = X7))
             
# name with libraryID
names(dat) <- sample_names

dat <- bind_rows(dat, .id = "sample_name")

dat <- mutate(dat, 
              Run = str_replace(sample_name, 
                                             "_exon|_intron", 
                                             ""),
              count_type = ifelse(str_detect(sample_name, "exon"),
                                  "exon",
                                  "intron"))
#join with main data
dat <- inner_join(dat, 
                  mdata,
                  by = c("Run"))

dat <- anti_join(dat, 
                 data_frame(gene_id = monoexonic_genes),
                 by = "gene_id")

fcount_dat <- group_by(dat, Run) %>% 
  mutate(lib_size = sum(counts),
         eff_length = as.numeric(length) - 150 + 1) %>% 
  filter(eff_length > 0 ) %>% 
  mutate(rate = log(counts) - log(eff_length),
         total_rates = log(sum(exp(rate))),
         TPM = exp(rate - total_rates + log(1e6)))

fcount_dat$expt <- "featureCounts"
```

## salmon data 
```{r read_in_salmon}

fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "drosophila",
                   mdata$study,
                   .x,
                   mdata$Run,
                   "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$Run
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "Run"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("Run"))

dat <- anti_join(dat, 
                 data_frame(gene_id = monoexonic_genes),
                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_ annotations
salmon_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))
```
## bowtie2 data
```{r read_in_bt2_salmon}

fa_types <- c(
  "primary_transcripts_masked",
  "primary_transcripts_masked_exons"
)

files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon_bt2_masked", 
                   "drosophila",
                   mdata$study,
                   .x,
                   mdata$Run,
                   "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))
names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$Run
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "Run"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("Run"))

dat <- anti_join(dat, 
                 data_frame(gene_id = monoexonic_genes),
                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_ annotations
bt2_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))

# append bt2 to expt to distinguish
bt2_dat$expt <- str_c("bt2_", bt2_dat$expt)
```
## combine all expts
```{r}
keep_cols <- intersect(colnames(salmon_dat), colnames(fcount_dat))

fcount_dat <- fcount_dat[, keep_cols]
salmon_dat <- salmon_dat[, keep_cols]
bt2_dat <- bt2_dat[, keep_cols]
dat <- bind_rows(list(salmon_dat, fcount_dat, bt2_dat))
```

## plot pre_mRNA ratios

```{r, calc_premRNA}
# sum primary and mature TPMs for each gene 
summary_dat <- dplyr::select(dat, 
                             expt,
                             TPM, 
                             gene_id, 
              Run, 
              count_type, 
              developmental_stage,
              study, 
              strain) %>% 
  group_by(expt,
           gene_id, 
           Run, 
           developmental_stage, 
           count_type,
           strain) %>% 
  dplyr::summarize(total_tpm = sum(TPM, na.rm = F)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- summary_dat %>% 
  spread(count_type, total_tpm, fill = 0) %>% 
  mutate(primary_ratio  = intron / (intron + exon))


## calc mean primary ratio
tpm_means <- tpm_summary %>%  
  filter(!is.na(primary_ratio)) %>% 
  group_by(expt,
           gene_id, 
           developmental_stage,
           strain) %>% 
  summarize(mean_primary = mean(primary_ratio,
                                na.rm = T)) 
  
tpm_means <- ungroup(tpm_means)

tpm_means <- mutate(tpm_means, 
                       developmental_stage = factor(developmental_stage, 
                                         levels = c(
                                           "Stage 11 oocyte",
                                           "Stage 12 oocyte",
                                           "Stage 13 oocyte",
                                           "Stage 14 oocyte",
                                           "Activated egg",
                                           "0-1 hr embryo",
                                           "1-2 hr embryo",
                                           "2-3 hr embryo",
                                           "3-4 hr embryo",
                                           "4-5 hr embryo",
                                           "5-6 hr embryo")
                                         )
                       )

tpm_means <- mutate(tpm_means,
                       strain = ifelse(strain == "wt",
                                       paste(strain, 
                                             "(Eichhorn)"),
                                       paste(strain,
                                             "(Rissland)")))
                                        
```

```{r plt_it}

plt_dat <- split(tpm_means, tpm_means$expt)

plts <- imap(plt_dat, 
  ~ggplot(.x, 
       aes(developmental_stage, mean_primary)) +
  geom_boxplot(aes(fill = developmental_stage), 
               outlier.shape = NA,
               coef = 1e11) + 
  facet_grid(~strain, 
             scale = "free") +
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM")),
       title = .y) +
  scale_fill_brewer(palette = "Paired",
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5),
    strip.text = element_text(size = 8)
  )
)

plt <- plot_grid(plotlist = plts, nrow = 4, ncol = 2)
plt

save_plot("exon_intron_boxplt-pre-mRNA-ratios.pdf", 
          plt, 
          base_width = 6,
          base_height = 4.5,
          nrow = 4,
          ncol = 2)

```

## Classify transcripts into maternal or zygotic

Next I'll try to classify transcripts into catagories based upon their abundance changes throughout embryogenesis. 

```{r}
pre_mzt <- c(
  "Stage 11 oocyte",
  "Stage 12 oocyte",
  "Stage 13 oocyte",
  "Stage 14 oocyte",
  "Activated egg",
  "0-1 hr embryo",
  "1-2 hr embryo"
  )
post_mzt <- c("2-3 hr embryo",
  "3-4 hr embryo",
  "4-5 hr embryo",
  "5-6 hr embryo")

# average pre and post MZT TPMs
avg_abundance <- summary_dat %>% 
  mutate(mzt = ifelse(developmental_stage %in% pre_mzt,
                      "pre_mzt",
                      "post_mzt")) %>% 
  group_by(expt, gene_id, strain, mzt, count_type) %>% 
  summarize(mean_mzt_expr = mean(total_tpm)) %>% 
  group_by(expt, gene_id, strain, count_type) %>% 
  mutate(mean_abundance = mean(mean_mzt_expr))

avg_abundance <- avg_abundance %>% 
  ungroup() %>% 
  spread(mzt, mean_mzt_expr) %>% 
  mutate(log2abundance = log2(post_mzt + 1) - log2(pre_mzt + 1))

plt_dat <- mutate(avg_abundance,
                  mean_abundance = log2(mean_abundance + 1))
```

```{r}
# plot MAs per expt.

plt_dat_exons <- filter(plt_dat, count_type == "exon") 

plt_dat_exons <- split(plt_dat_exons, plt_dat_exons$expt)

plts <- imap(plt_dat_exons, 
  ~ggplot(.x, 
       aes(mean_abundance, log2abundance)) +
  geom_point(size = 0.1) + 
  scale_x_log10() + 
  facet_grid(~strain, 
             scale = "free") +
  labs(x = "mean abundance (log 10)",
       y = expression(frac("post-mzt", 
                           "pre-mzt")),
       title = .y) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5),
    strip.text = element_text(size = 8)
  )
)

plt <- plot_grid(plotlist = plts, nrow = 3, ncol = 2)

save_plot("ma_plots_exons.png", 
          plt, 
          base_width = 6,
          base_height = 4.5,
          nrow = 3,
          ncol = 2)


## intronic

plt_dat_introns <- filter(plt_dat, count_type == "intron") 

plt_dat_introns <- split(plt_dat_introns,
                         plt_dat_introns$expt)

plts <- imap(plt_dat_introns, 
  ~ggplot(.x, 
       aes(mean_abundance, log2abundance)) +
  geom_point(size = 0.1) + 
  facet_grid(~strain, 
             scale = "free") +
  labs(x = "",
       y = expression(frac("post-mzt", 
                           "pre-mzt")),
       title = .y) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5),
    strip.text = element_text(size = 8)
  )
)

plt <- plot_grid(plotlist = plts, nrow = 3, ncol = 2)

save_plot("ma_plots_introns.png", 
          plt, 
          base_width = 6,
          base_height = 4.5,
          nrow = 3,
          ncol = 2)
```

```{r classify_txs_mzt}
## select transcripts with tpm > 0.5 as maternal

mzt_class <- mutate(avg_abundance, 
                    mzt_class = ifelse(pre_mzt > 0.1 & log2abundance > 1,
                           "matzyg", 
                           ifelse(pre_mzt <  0.1 & log2abundance > 0,
                                  "zyg",
                                  ifelse(pre_mzt > 0.1,
                                         "mat", 
                                         NA))))

mzt_class <- dplyr::select(mzt_class, expt, gene_id, strain, count_type, mzt_class)    

mzt_class <- spread(mzt_class, count_type, mzt_class)


mzt_class <- filter(mzt_class, !(is.na(exon) & is.na(intron))) %>% 
  mutate(mzt_class = ifelse(is.na(intron),
                            exon,
                            ifelse(is.na(exon),
                                   intron,
                                   ifelse(str_detect(intron, "zyg") & exon %in% c("mat", "matzyg"),
                                          "matzyg",
                                          ifelse(str_detect(intron, "zyg") & exon %in% c("zyg"),
                                                 "zyg",
                                                 NA)))))

                       
gene2tx <- gtf[, c("gene_id", "gene_name")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)
gene2tx <- as_data_frame(gene2tx)

mzt_class <- left_join(mzt_class, gene2tx, by = "gene_id") %>% unique()
```

```{r}
if(!file.exists("eisen_data.tsv")){
  download.file("https://ndownloader.figshare.com/files/400050",
              "eisen_data.tsv")
}
edata <- read_tsv("eisen_data.tsv")

```

```{r}

cas_cols <- str_subset(colnames(edata), "cas")

edata_confident <- edata[rowSums(edata[, cas_cols], na.rm = T) > 0 ,]

res_eisen <- edata_confident %>% 
  dplyr::select(NAME, CLASS)

res_eisen <- inner_join(mzt_class, res_eisen, 
            by = c("gene_name" = "NAME")) %>% 
  dplyr::rename(kr_class = mzt_class,
                eisen_class = CLASS)

summary_dat <- res_eisen %>% 
 group_by(kr_class, eisen_class) %>% 
 summarize(n())

summary_dat 

classification_summary <- res_eisen %>% 
  mutate(eisen_class = ifelse(is.na(eisen_class),
                              "not classified",
                              eisen_class),
         kr_class = ifelse(is.na(kr_class),
                                 "not classified",
                                 kr_class),
         kr_class = factor(kr_class,
                           levels = c(
                             "mat",
                             "matzyg",
                             "zyg",
                             "not classified"
                           )))


plt <- ggplot(classification_summary, aes(kr_class)) +
  geom_bar(aes(fill = eisen_class)) +
  facet_wrap(~expt) +
  labs(y = "# of genes", x = "intron based classification") +
  scale_fill_brewer(palette = "Set1", 
                    name = "SNP based\nclassification\n(Lott et al 2011)") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

plt

save_plot("eisen_vs_intron_classification.pdf", plt,
          base_height = 7)
```

