---
title: "Examine intronic signals during MZT in Drosophila embryos"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Background

The Maternal to Zygotic Transistion (MZT) is a developmental stage in which the embryo initiates transcription of mRNAs from the zygotic genome. Maternally deposited mRNAs begin to be degraded and are replaced by zygotic messages. 

Zygotic mRNAs can be distinguished from maternal mRNAs, due to the presence or higher proportions of introns. Maternally derived mRNAs are likely to be spliced, and any remaining intron lariat degraded prior to fertilization. In contrast zygotic transcription will produce mRNAs in that need to be spliced, and thus there will be some proportion of unspliced mRNA, and intron lariat derivied from the zygotic transcripts. 

Using ribo-zero depleted total-RNA-Seq libraries, one would expect to be able to recover intronic mRNA fragements, and be able to assign for each transcript, the relative proportion of each transcript that is maternally or zygotically derived. 


## Approach

Each total RNA-Seq lib was trimmed to remove Tru-Seq adapters (Rissland libraries), or barcodes and custom 3' adapter (Bartel libraries). The trimmed reads were then processed with `salmon` to estimate `TPMs` for each transcript. The `salmon` transcript database was modified to include a pre-mRNA entry, which was designated as an exon that spanned the entire gene. This pre-mRNA annotation matches the `gene` entry from the annotaiton file. By including the pre-mRNA transcript in the `salmon` database, one can get an estimate of the fraction of pre-mRNA or mature mRNA present in each sample. 

First we'll examine the extent to which pre-mRNA transcripts increase during embryonic development. Then, if a clear increase is present, we can classify transcripts into a % maternal and % zygotic, and additional classify the temporal pattern of zygotic transcription i.e. are there distinct sets of early versus late transcripts, and do their transcriptional pattern suggest function (TF transcription prior to target activation/repression, miRNA transcription before target repression, RBP transcription prior to translational control, etc.)

 
## RNA-Seq Libraries
```{r load_libs}
source("../../R/globals.r")
```



```{r get_metadata}
mdata1 <- read_tsv(file.path(data_dir, "raw_data", "drosophila", "GSE83616_run_info_geo.txt"))
mdata2 <- read_tsv(file.path(data_dir, "raw_data", "drosophila", "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("eichhorn", "rissland")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```


## Alignment %

```{r alignment}

alignment <- map(mdata$Run, 
                  ~fromJSON(txt = file.path(data_dir, 
                          "salmon", 
                          .x, 
                          "aux_info", 
                          "meta_info.json")) %>% .$percent_mapped)

alignment <- data_frame(library = mdata$Run,
                        percent_mapped = unlist(alignment))


inner_join(alignment, 
          mdata, 
          by = c("library" = "Run"))
```

## Load in salmon data

```{r load_dat, message = F}
files <- dir(file.path(data_dir, "salmon"),
            recursive = T,
            pattern = "*quant.sf",
            full.names = T)

dat <- map(files, read_tsv)

# name with libraryID
names(dat) <- dirname(files) %>% 
  str_split(., "[/]", simplify = T) %>% 
  .[, 8]

dat <- bind_rows(dat, .id = "libraryID")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("libraryID" = "Run"))
```

### transcript to gene mapping
```{r drop_monoexonic}
gtf <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  bed_merge() %>% 
  dplyr::summarize(total_exons = n())

monoexonic_genes <- dplyr::filter(gtf_summary, total_exons == 1) %>% pull(gene_id)
```


```{r remove_monoexonic}

dat <- anti_join(dat, 
                    data_frame(gene_id = monoexonic_genes),
                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat %>% 
  mutate(type = ifelse(str_detect(Name, 
                                  "^pre_"), 
                       "pre", "mature")) -> dat

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_ annotations

dat <- mutate(dat, 
              gene_id = ifelse(type == "pre",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))

dat <- mutate(dat,
              monoexonic = gene_id %in% monoexonic_genes)
```

### plot pre_mRNA ratio

```{r, calc_premRNA}
# sum primary and mature TPMs for each gene 
summary_dat <- dplyr::select(dat, TPM, gene_id, 
              libraryID, type, developmental_stage,
              study, strain) %>% 
  group_by(gene_id, libraryID, 
           developmental_stage, type,
           strain) %>% 
  dplyr::summarize(total_tpm = sum(TPM, na.rm = F)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- summary_dat %>% 
  spread(type, total_tpm) %>% 
  mutate(primary_ratio  = pre / (mature + pre))


## calc mean primary ratio
tpm_means <- tpm_summary %>% 
  group_by(gene_id, 
           developmental_stage,
           strain) %>% 
  summarize(mean_primary = mean(primary_ratio,
                                na.rm = T)) 

tpm_means <- mutate(tpm_means, 
                       stage = factor(developmental_stage, 
                                         levels = c(
                                           "Stage 11 oocyte",
                                           "Stage 12 oocyte",
                                           "Stage 13 oocyte",
                                           "Stage 14 oocyte",
                                           "Activated egg",
                                           "0-1 hr embryo",
                                           "1-2 hr embryo",
                                           "2-3 hr embryo",
                                           "3-4 hr embryo",
                                           "4-5 hr embryo",
                                           "5-6 hr embryo")
                                         )
                       )

tpm_means <- mutate(tpm_means,
                       strain = ifelse(strain == "wt",
                                       paste(strain, 
                                             "(Eichhorn)"),
                                       paste(strain,
                                             "(Rissland)")))
                                        
tpm_means %>% 
  group_by(gene_id, developmental_stage) %>% 
  summarize(mean_primary = mean(mean_primary, na.rm = T)) %>% 
  mutate(strain = "all strains") %>% 
  bind_rows(tpm_means, .) %>% 
  arrange(gene_id) %>% 
  ungroup() -> all_combined

ggplot(all_combined, 
       aes(developmental_stage, mean_primary)) +
  geom_boxplot(aes(fill = developmental_stage), coef = 1e3) + 
  facet_grid(~strain, scale = "free") +
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_brewer(palette = "Set1",
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5)
  )
```

## Classify transcripts into maternal or zygotic

Next I'll try to classify transcripts into catagories based upon their abundance changes throughout embryogenesis. 

First need to identify transcripts that are differentially expressed between maternal and zygotic states. Let's try a two state comparison of 
pre MZT and post MZT, which based on the correlation matrix shown above, is strongly segregated between the two.

First generate salmon quant.sf files for pre-mRNAs or mature mRNAs independently for loading into `tximport`
```{r, split_salmon}

tx2gene_mapping <- dplyr::select(dat, 
                         transcript_id, 
                         gene_id) %>% 
  unique() %>% 
  as.data.frame()

files <- dir(file.path(data_dir, "salmon"),
            recursive = T,
            pattern = "*quant.sf",
            full.names = T)

sra_ids <- dirname(files) %>% 
  str_split(., "/", simplify = T) %>% 
  .[, 8]

keep <- sra_ids %in% mdata$Run
files <- files[keep]
sra_ids <- sra_ids[keep]

all_dat <- map(files, read_tsv)

all_pre <- map(all_dat, 
                 ~dplyr::filter(.x,
                                str_detect(Name, "^pre_")))

all_mat <- map(all_dat, 
               ~dplyr::filter(.x,
                              !str_detect(Name, "^pre_")))

dir.create("salmon_modified", recursive = T)

walk2(all_pre,
      sra_ids,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".premRNA"))))
walk2(all_mat,
      sra_ids,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".maturemRNA"))))
```


```{r}
files <- dir("salmon_modified", full.names = T)

mature <- str_subset(files, "maturemRNA$")
pre <- str_subset(files, "premRNA$")

matures <- tximport(mature, 
         type = "salmon",
         tx2gene = tx2gene_mapping)

pres <- tximport(pre, 
         type = "salmon",
         tx2gene = tx2gene_mapping)

# make metadata sample ordered by columns in count matrix
pdata <- inner_join(data_frame(Run = sra_ids),
           mdata, by = "Run")

pdata <- mutate(pdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"))
```


```{r deseq2}

dds <- DESeqDataSetFromTximport(matures, pdata, ~mzt)
dds_mature <- DESeq(dds)
res_mature <- results(dds_mature)
res_mature_tidy <- results(dds_mature, tidy = TRUE)

dds <- DESeqDataSetFromTximport(pres, pdata, ~mzt)
dds_pre <- DESeq(dds)
res_pre <- results(dds_pre)
res_pre_tidy <- results(dds_pre, tidy = TRUE)
```

```{r}
pre_plot_dat <- res_pre_tidy %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0, "Down", "Not Sig")))
n_sig_up <- nrow(dplyr::filter(pre_plot_dat, sig == "Up")) %>% 
  formatC(big.mark = ",")
n_sig_down <- nrow(dplyr::filter(pre_plot_dat, sig == "Down"))  %>% 
  formatC(big.mark = ",")
n_notsig <- nrow(dplyr::filter(pre_plot_dat, sig == "Not Sig"))  %>% 
  formatC(big.mark = ",")

p <- ggplot(pre_plot_dat, 
       aes(baseMean, log2FoldChange)) +
  geom_point(aes(colour = sig), size = 0.1) +
  scale_colour_viridis(discrete = TRUE,
                       name = "",
                       labels = c(paste0("Sig down (n = ", n_sig_down, ")"),
                                  paste0("Not Sig (n= ", n_notsig, ")"),
                                  paste0("Sig up (n = ", n_sig_up, ")"))) +
  scale_x_log10() + 
  labs(x = expression(paste("log"[10], " Average Expression")),
       y = expression(paste("log"[2], " ", frac("Zygotic", "Maternal"))),
       title = "pre-mRNA") +
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(legend.position = "top")
  
p

mature_plot_dat <- res_mature_tidy %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0, "Down", "Not Sig")))
n_sig_up <- nrow(dplyr::filter(mature_plot_dat, sig == "Up")) %>% 
  formatC(big.mark = ",")
n_sig_down <- nrow(dplyr::filter(mature_plot_dat, sig == "Down"))  %>% 
  formatC(big.mark = ",")
n_notsig <- nrow(dplyr::filter(mature_plot_dat, sig == "Not Sig"))  %>% 
  formatC(big.mark = ",")

p <- ggplot(mature_plot_dat, 
       aes(baseMean, log2FoldChange)) +
  geom_point(aes(colour = sig), size = 0.1) +
  scale_colour_viridis(discrete = TRUE,
                       name = "",
                       labels = c(paste0("Sig down (n = ", n_sig_down, ")"),
                                  paste0("Not Sig (n= ", n_notsig, ")"),
                                  paste0("Sig up (n = ", n_sig_up, ")"))) +
  scale_x_log10() + 
  labs(x = expression(paste("log"[10], " Average Expression")),
       y = expression(paste("log"[2], " ", frac("Zygotic", "Maternal"))),
       title = "mature-mRNA") +
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(legend.position = "top")
  
p
```


## Overlap between zygotic transcripts 

How similar are the significant transcripts recovered based on exonic or intronic signal?

```{r plot_zyg_mat_pre}
library(eulerr)

sig_mat_up <- res_mature_tidy %>% filter(padj < 0.05, log2FoldChange > 0)
sig_pre_up <- res_pre_tidy %>% filter(padj < 0.05, log2FoldChange > 0)
all_sig_up <- union(sig_mat_up$row, sig_pre_up$row)
lgl_up_df <- data.frame(
  `mature_mRNA` =  all_sig_up %in% sig_mat_up$row,
  `pre_mRNA` =  all_sig_up %in% sig_pre_up$row
)

sig_mat_down <- res_mature_tidy %>% filter(padj < 0.05, log2FoldChange < 0)
sig_pre_down <- res_pre_tidy %>% filter(padj < 0.05, log2FoldChange < 0)
all_sig_down <- union(sig_mat_down$row, sig_pre_down$row)
lgl_down_df <- data.frame(
  `mature_mRNA` =  all_sig_down %in% sig_mat_down$row,
  `pre_mRNA` =  all_sig_down %in% sig_pre_down$row
)

fit_up <- euler(lgl_up_df)
fit_down <- euler(lgl_down_df)

plot(fit_up, fill = viridis(2), quantities = TRUE, main = "Increased abundance in Zygotic Stages")
plot(fit_down, fill = viridis(2), quantities = TRUE, main = "Decreased abundance in Zygotic Stages")
```


What are the transcripts that have increased intronic epression, but no increases in mature?

```{r hmap_intronic_only, fig.height = 10, fig.width = 8}
interest_txs <- sig_pre_up[!sig_pre_up$row %in% sig_mat_up$row, ]

dds_pre_rlogs <- rlog(dds_pre)

# add in colnames
cols <- pdata %>% 
  mutate(cnames = str_c(developmental_stage, strain , sep = "::")) %>% 
  pull(cnames)

rlog_counts_pre <- assay(dds_pre_rlogs)
colnames(rlog_counts_pre) <- cols

dds_mat_rlogs <- rlog(dds_mature)

# add in colnames
cols <- pdata %>% 
  mutate(cnames = str_c(developmental_stage, strain , sep = "::")) %>% 
  pull(cnames)

rlog_counts_mat <- assay(dds_mat_rlogs)
colnames(rlog_counts_mat) <- cols

sig_pre_rlogs <- rlog_counts_pre[rownames(rlog_counts_pre) %in% interest_txs$row &
                                 rownames(rlog_counts_pre) %in% rownames(rlog_counts_mat), ]
sig_mat_rlogs <- rlog_counts_mat[rownames(rlog_counts_mat) %in% interest_txs$row &
                                 rownames(rlog_counts_mat) %in% rownames(rlog_counts_pre), ]


ht1 <- Heatmap(sig_pre_rlogs, 
         name = "pre-mRNA abundance\nrlog",
         show_row_names = F,
         show_row_dend = F)

ht2 <- Heatmap(sig_mat_rlogs, 
         name = "mature-mRNA abundance\nrlog",
         show_row_names = F,
         show_row_dend = F)

print(ht1 + ht2)
```


### Compare to SNP data called by Michael Eisen's group

```{r eisen_comp, eval = F}
library("org.Dm.eg.db")
edata <- read_tsv("https://ndownloader.figshare.com/files/400050")


res_tidy <- bind_rows(premRNA = res_pre_tidy, 
          maturemRNA = res_mature_tidy, .id = "rna_type") %>% 
  as_tibble()

res_pre_tidy$symbol <- mapIds(org.Dm.eg.db, 
                            keys=res_tidy$row, 
                            column="SYMBOL", 
                            keytype="ENSEMBL",
                            multiVals="first")

res_tidy$entrez <- mapIds(org.Dm.eg.db, 
                            keys=res_tidy$row, 
                            column="ENTREZID", 
                            keytype="ENSEMBL",
                            multiVals="first")

res_tidy$name =   mapIds(org.Dm.eg.db,
                           keys=res_tidy$row, 
                           column="GENENAME",
                           keytype="ENSEMBL",
                           multiVals="first")


res_pre_eisen <- edata %>% 
  dplyr::select(NAME, CLASS) %>% 
  left_join(res_pre_tidy, ., by = c("symbol" = "NAME")) %>% 
  dplyr::rename(eisen_class = CLASS)

res_pre_eisen %>% 
  semi_join(sig_pre_up) %>% 
  dplyr::count(eisen_class)
```

### Compare to Zelda peaks

Chip-Seq data taken from http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1002266 . 

```{r zelda, eval = F}

readxls::read_xlhttps://doi.org/10.1371/journal.pgen.1002266.s007

```


### Output tables

The interesting transcripts that have increased intronic expression but do not have increased mature levels will be written to tables to send to Olivia.

```{r outtbls}

gene2name <- dplyr::select(gtf, gene_id, gene_name) %>% unique()

res_mature_tidy_simple <- dplyr::select(res_mature_tidy, 
                                        row:log2FoldChange, padj)
interesting_txs <- left_join(interest_txs, 
                             gene2name, 
                             by = c("row" = "gene_id")) %>% 
  dplyr::select(gene_name, row:log2FoldChange, padj) %>% 
  arrange(padj) %>% 
  left_join(res_mature_tidy_simple, 
            by = "row",
            suffix = c("_pre-mRNA", "_mature-mRNA")) %>% 
  dplyr::rename(gene_id = row)

all_txs <- left_join(res_pre_tidy, 
                             gene2name, 
                             by = c("row" = "gene_id")) %>% 
  dplyr::select(gene_name, row:log2FoldChange, padj) %>% 
  arrange(padj) %>% 
  left_join(res_mature_tidy_simple, 
            by = "row",
            suffix = c("_pre-mRNA", "_mature-mRNA")) %>% 
  dplyr::rename(gene_id = row)

today <- Sys.Date()
write_tsv(interesting_txs, paste0("upregulated_intronic_", today, ".tsv"))
write_tsv(all_txs, paste0("all_genes_", today, ".tsv"))
```