---
title: "Nanopore analysis"
author: "Kent Riemondy RBI"
date: "7/29/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(kentr)
source(here::here("R/globals.r"))
eisa_dir <- file.path(project_dir, "dbases-rerun", "drosophila", "eisa")
fig_dir <- "figs_ont"
dir.create(fig_dir, recursive = TRUE)
tbl_dir <- "processed_files"
out_dir <- "output_files"
dir.create(out_dir)
```

```{r}
mdata <- read_tsv(file.path(data_dir,
                             "raw_data",
                             "drosophila",
                             "RISSLAND",
                             "GSE98106_run_info_geo.txt"))

#simplify mdata
mdata <- dplyr::select(mdata,
                        Run_s, 
                        developmental_stage_s,
                        rip_antibody_s,
                        source_name_s,
                        strain_s)

# select relevant libs
mdata <- dplyr::filter(mdata,
                        source_name_s == "Embryo",
                        strain_s %in% c("w1118", "eGFP-ME31B"),
                        rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)


#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

fa_type <- "eisa_masked"

pdata <- mutate(mdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"),
                strain_id = str_replace(strain, "-", "_"))

```


```{r}
plot_ont_tx <- function(ont_align, 
                             transcripts,
                             gtf_obj,
                             scale_y = FALSE,
                             gnome = "dm6",
                             annotation_label = NULL,
                             ...){
  library(GenomicFeatures)
  library(Gviz)
  library(GenomicAlignments)
  
  options(ucscChromosomeNames=FALSE)

  #subset gtf to get min max coords to plot
  gfeatures <- gtf_to_gviz(gtf_obj, transcripts)
  
  chrom <- as.vector(unique(gfeatures$chromosome))[1]
  start <- min(gfeatures$start)
  end <- max(gfeatures$end)
  strand <- as.vector(unique(gfeatures$strand))[1]
  
  # get coverage
  # cov_list <- split(ont_align, ont_align$sample) %>% 
  #   map(~group_by(.x, rname) %>% 
  #         summarize(n = n()) %>% 
  #         inner_join(gfeatures, ., by  = c("transcript" = "rname")) %>% 
  #         mutate(transcript = str_c(transcript, " (", n, ")")))
  cov_list <- split(ont_align, ont_align$sample) %>% 
    map(~group_by(.x, rname) %>% 
          summarize(n = n())) %>% 
    map(~{res <- .x$n; names(res) <- .x$rname; res}) 
  
  grange_feat <- gtf_obj[gtf_obj$transcript_id %in% features]
  
  dtracks <- pmap(list(cov_list,
                       names(cov_list),
                       viridis(length(cov_list))),
                  function(x, y, z){
                    feats <- grange_feat
                    feats$transcript_id <- str_c(grange_feat$transcript_id,
                                                 " (", 
                                                 x[grange_feat$transcript_id],
                                                 ")")

                    GeneRegionTrack(makeTxDbFromGRanges(feats), 
                              name =  as.character(y),
                              showID = TRUE,
                              col = 'black', fill = z,
                              transcriptAnnotation = "transcript"
                    )})
  
  # grtrack <- GeneRegionTrack(gfeatures,
  #                            genome = gnome, 
  #                            chromosome=chrom, 
  #                            name="",
  #                         #   stacking = "dense",
  #                            col.title = "black",
  #                         showId = TRUE,
  #                         transcriptAnnotation = "transcript")
  options(scipen=16)
  if(is.null(annotation_label)){
    annotation_label <- unique(gfeatures$symbol)[1]
  }
  plotTracks(dtracks,
             # chromosome = chrom,
             # from = start,
             # to = end,
             reverseStrand = strand == "-", 
             col.title = "black",
             col.axis = "black",
             background.title ="transparent",
          #   lwd = .5, 
             innerMargin = 10,
             margin = 50,
             main = annotation_label,
             ...)

}
```



```{r}
switch_isos <- read_excel(file.path("..",
                                    "2021-06-publication",
                                    "output_files",
                                    "isoform_summary.xlsx"),
           sheet = 2) 


ont_align_summary <- read_tsv("../2021-05-23/all_merged_read_alignments.tsv")
  

gtf_obj <- import(annots$drosophila$gtf)

to_plot <- c(
  "PNUTS",
  "RhoGEF3",
  "Miro"
)

walk(to_plot[1], 
    ~{
      features <- switch_isos %>% filter(gene_name == .x) %>% pull(transcript_id)
      ont_align <- filter(ont_align_summary, rname %in% features)
      
      pdf(file.path("../figures/pdfs/tx-nano-plots/", str_c(.x, "_coverage.pdf")),
          width = 6,
          height = 3)
      plot_ont_tx(ont_align, features, gtf_obj,
                 title.width = 1.5,
                 sizes = c(0.1, rep(0.25, 5)),
                 fontsize = 8,
                 cex.title = 1)
      dev.off()
    })
```

```{r}
base_dir <- "~/Projects/mzt-introns/data-rerun/bigwigs/drosophila/star/RISSLAND"
sras <- filter(mdata, strain == "w1118") %>% pull(Run)
ids <- filter(mdata, strain == "w1118") %>% pull(developmental_stage)

bws <- c("_fwd.bw", "_rev.bw")
names(bws) <- c("+", "-")

bw_fns <- map(bws, function(x) {
  out <- file.path(base_dir, paste0(sras, x))
  names(out) <- ids
  out
}) 


```

## Representative transcripts

Show coverage profiles from transcripts that swtich
```{r, eval = FALSE}
additional_annots <- read_tsv(file.path(out_dir, "all_isoform_annotations.tsv"))
base_url <- "http://amc-sandbox.ucdenver.edu/User33/rissland/hub/dm6/ont/"
bbs <- c("0-1_hr_1and2.bb","4-5_hr_1and2.bb")
bbs <- paste0(base_url, bbs)
names(bbs) <- c("0-1 hr", "4-5 hr")


gtf_obj <- import(annots$drosophila$gtf)

tx_summary_plots <- function(bws, 
                             gene,
                             isoform_annotation_df,
                             ...){
  
  to_plot <- filter(isoform_annotation_df, gene_name == gene)
  strand <- unique(to_plot$strand)[1]

  features <- unique(to_plot$featureID)
  plot_ont_tx(bw_fns[[strand]], 
                 features,
                 gtf_obj,
                 scale_y = FALSE,
                 cex.main = 0.75,
                 rotation.title=0,
                 ...
                 )
}

genes_to_plot <- c(
  "PNUTS")
  # "cnn",
  # "Zn72D",
  # "Rb97D",
  # "pyd",
  # "CG8331",
  # "Pk92B",
  # "yata",
  # "hts",
  # "bl",
  # "eIF-2gamma",
  # "Npc1a",
  # "Su(var)2-10",
  # "toc",
  # "Hph",
  # "blot",
  # "Mhcl",
  # "SC35",
  # "CG4658",
  # "arm",
  # "Nmnat",
  # "Adar",
  # "Plip",
  # "lola")

tx_figs <- file.path(fig_dir, "ont_tx_figs")
dir.create(tx_figs, showWarnings = FALSE)

walk(genes_to_plot,
    ~{
      pdf(file.path(tx_figs,
                    str_c(.x, "_tx_coverage.pdf")),
          width = 6,
          height = 3)
      if (!.x %in% additional_annots$gene_name){
        message(.x, " not switching iso")
        return(NULL)
      }
      tx_summary_plots(bw_fns, gene = .x, 
                 isoform_annotation_df = additional_annots,
                 title.width = 1.5,
                 sizes = c(0.1, rep(0.25, 5)),
                 fontsize = 8,
                 cex.title = 1)
      dev.off()
    })

```