---
title: "3'UTR motif search"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Zebrafish MZT analysis cont.

Previously clusters of gene expression patterns were defined by using pre-mRNA signals to calssify transcripts. 

## RNA-Seq Libraries
```{r load_libs}
source("../../R/globals.r")
```

```{r clustsr}
clusters_pre <- read_tsv("../2018-03-15/premrna_cluster_ids.tsv")
clusters_mat <- read_tsv("../2018-03-15/maturerna_cluster_ids.tsv")

library(kentr)
gtffile <- "~/Projects/shared_dbases/annotation/zebrafish/GRCz10/Danio_rerio.GRCz10.91.gtf"

get_3utrs <- function(gtf_file, gene_ids, gene_col = "gene_id", transcript_col = "transcript_id"){
  gtf <- import(gtf_file)
  gtf <- as.data.frame(gtf)
  gtf <- filter(gtf, gene_id %in% gene_ids)
  gtf <- dplyr::select(gtf, 
                       chrom = seqnames, 
                       start, end, type, strand, exon_id, one_of(c(gene_col, transcript_col)))
  gtf <- dplyr::filter(gtf, 
                       type %in% c("exon", "CDS"))
  cds_gtf <- filter(gtf, type == "CDS") %>% 
         group_by(transcript_id) %>% 
         summarize(cds_start = unique(ifelse(strand == "+",
                                  min(start), 
                                  max(end))),
                   cds_end = unique(ifelse(strand == "+", 
                                max(end),
                                min(start)))) %>% 
    left_join(gtf, ., by = "transcript_id")
  
  utr3_gtf <- cds_gtf %>% 
    filter(type == "exon") %>% 
    group_by(transcript_id) %>% 
    mutate(utr3_exon = ifelse(strand == "+",
                              ifelse(start >= cds_end,
                                     "yes",
                                     ifelse(cds_end >= start & cds_end <= end,
                                            "overlapping",
                                            "no")),
                              ifelse(end <= cds_end,
                                     "yes",
                                     ifelse(cds_end >= start & cds_end <= end,
                                            "overlapping",
                                            "no")))) %>% 
    ungroup() %>% 
    filter(utr3_exon != "no")
  
  utr3_res <- utr3_gtf %>% 
    group_by(transcript_id) %>% 
    mutate(start = ifelse(strand == "+" & utr3_exon == "overlapping",
                          cds_end,
                          start),
           end = ifelse(strand == "-" & utr3_exon == "overlapping",
                        cds_end,
                        end))  %>% 
    filter(start != end) %>% 
    select(chrom:transcript_id)
  
  res <- group_by(utr3_res, gene_id, strand) %>% 
    bed_merge() %>% 
    ungroup()
  res
}

genome_file <- "~/Projects/shared_dbases/genomes/zebrafish/GRCz10/Danio_rerio.GRCz10.dna.toplevel.fa"

genes <- clusters_pre %>% filter(cluster == 7) %>% pull(gene_id)

utrs <- get_3utrs(gtffile, genes)
utrs <- group_by(utrs, gene_id) %>% 
  mutate(n = row_number(),
         id = str_c(gene_id, "_", n)) %>% 
  select(chrom, start, end, id, gene_id, strand) %>% 
  ungroup()

write_bed <- function(df, path = ""){
  res <- mutate_if(df, is.numeric, as.character)
  write_tsv(res, path, col_names = F)
}
               
write_bed(utrs, "cluster_7_utrs.bed")

utr_seqs <- kentr::get_sequences(utrs, genome_file)

gtf <- import(gtffile)
gtf <- as.data.frame(gtf)

all_genes <- filter(gtf, type == "gene") %>% select(gene_id) %>% unique()  %>% pull(gene_id)
all_utrs <- get_3utrs(gtffile, all_genes)
all_utrs <- group_by(all_utrs, gene_id) %>% 
  mutate(n = row_number(),
         id = str_c(gene_id, "_", n)) %>% 
  select(chrom, start, end, id, gene_id, strand) %>% 
  ungroup()

write_bed(all_utrs, "all_utrs.bed")

mat_genes <- clusters_mat %>% filter(cluster %in% c(1, 2, 5)) %>% pull(gene_id)
mat_utrs <- get_3utrs(gtffile, mat_genes)
mat_utrs <- group_by(mat_utrs, gene_id) %>% 
  mutate(n = row_number(),
         id = str_c(gene_id, "_", n)) %>% 
  select(chrom, start, end, id, gene_id, strand) %>% 
  ungroup()

write_bed(mat_utrs, "maternal_utrs.bed")
```

```{r kmers}
utr_seqs <- kentr::get_sequences(utrs, genome_file)
all_utr_seqs <- kentr::get_sequences(all_utrs, genome_file)

kmers <- mutate(utr_seqs, 
                kmers = get_kmers(seq, n = 6)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

kmers_all <- mutate(all_utr_seqs, 
                kmers = get_kmers(seq, n = 6)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

bg <- kmers_all %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmer_count <- kmers %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmers_total <- left_join(kmer_count, 
                         bg,
                         by = "kmer",
                         suffix = c("_test", "_bg"))
kmers_total <- mutate(kmers_total, 
                      n_seq_test = length(genes),
                      n_seq_bg = length(all_genes))
kmers_total %>% 
  na.omit(.) %>%
  rowwise() %>% 
  do(tidy(binom.test(.$counts_test, .$n_seq_test, 
                     p = .$counts_bg / .$n_seq_bg),
          alternative="greater")) -> out

bind_cols(kmers_total %>% na.omit(), out) %>% 
  mutate(padj = p.adjust(p.value)) %>% 
  arrange(padj) %>% View()
```


## TF enrichment

The JASPAR database of PWMs was recently updated and they generated TF binding site predictions for both dropshoplia and zebrafish (yah!). The data is available in BED format, so can be directly used for looking at TF enrichment.  

First we'll start with generating a set of promoter regions to query. 

```{r get_jaspar}
gtffile <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"
chromfile <- "~/Projects/shared_dbases/genomes/drosophila/Drosophila_melanogaster.BDGP6.dna.toplevel.fa.fai"
genomefile <- "~/Projects/shared_dbases/genomes/drosophila/Drosophila_melanogaster.BDGP6.dna.toplevel.fa"
gtf <- import(gtffile)
gtf <- as.data.frame(gtf)

tss <- filter(gtf, type == "gene") %>%
  mutate(end = ifelse(strand == "+",
                      start + 1,
                      end),
         start = ifelse(strand == "-",
                        end - 1,
                        end)) %>% 
  select(chrom = seqnames, start, end, gene_name, gene_id, strand)

tss_slop <- bed_slop(tss, genome = read_genome(chromfile), 
                     left = 1000, right = 200, strand = T)
write_bed(tss_slop, "tss.bed")
tss_seqs <- get_sequences(tss_slop, genomefile)
write_fasta(tss_seqs, "tss_dm6.fasta")
```

Also try homer with custom fastas
```{r}
gtffile <- "~/Projects/shared_dbases/annotation/zebrafish/GRCz10/Danio_rerio.GRCz10.91.gtf"
chromfile <- "~/Projects/shared_dbases/genomes/zebrafish/GRCz10/Danio_rerio.GRCz10.dna.toplevel.fa.fai"
genomefile <- "~/Projects/shared_dbases/genomes/zebrafish/GRCz10/Danio_rerio.GRCz10.dna.toplevel.fa"

gtf <- import(gtffile)
gtf <- as.data.frame(gtf)

tss <- filter(gtf, type == "gene") %>%
  mutate(end = ifelse(strand == "+",
                      start + 1,
                      end),
         start = ifelse(strand == "-",
                        end - 1,
                        end)) %>% 
  select(chrom = seqnames, start, end, gene_name, gene_id, strand)

tss_slop <- bed_slop(tss, genome = read_genome(chromfile), left = 1000, right = 200, strand = T)

write_bed(tss_slop, "tss_zf.bed")
tss_seqs <- get_sequences(tss_slop, genomefile)
write_fasta(tss_seqs, "tss_zf10.fasta")

clust7_genes <- clusters_pre %>% filter(cluster == 7) %>% pull(gene_id)
clust7_tss <- filter(tss_seqs, gene_id %in% clust7_genes)
write_fasta(clust7_tss, "tss_cluster7_zf10.fasta")

```