---
title: "Intron mask"
author: "Kent Riemondy RBI"
date: "12/26/2019"
output: html_document
---


## Generate expressed intron mask

To define introns, all intronic mRNA regions with coverage in the pre-ZGA samples will be masked and excluded from the intronic set used for counting. 

```{r coverage_tracks}
mdata <- read_tsv(file.path(data_dir, "raw_data", "drosophila", 
                             "RISSLAND", "GSE98106_run_info_geo.txt"))

mdata <- dplyr::select(mdata,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata <- dplyr::filter(mdata,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s) %>% 
  filter(developmental_stage_s == "0-1 hr embryo")


bws <- dir(file.path(data_dir, "bigwigs", "drosophila", "RISSLAND"), 
           full.names = TRUE, 
           pattern = ".bw")

bws <- str_subset(bws, str_c(mdata$Run_s, 
                             collapse = "|"))

fwd_bws <- str_subset(bws, "fwd.bw")
rev_bws <- str_subset(bws, "rev.bw")

fwd_df <- map_dfr(fwd_bws,
                  ~read_bigwig(.x, set_strand = "+")) 

rev_df <- map_dfr(rev_bws,
                  ~read_bigwig(.x, set_strand = "-")) 

# sum coverage
fwd_df <- bed_partition(fwd_df, 
                        score = sum(score))

rev_df <- bed_partition(rev_df, 
                        score = sum(score))

ivl_df <- bind_rows(list(
  "+" = fwd_df,
  "-" = rev_df),
  .id = "strand")

ivl_df <- dplyr::select(ivl_df, chrom:score, strand)
```

Next exclude exonic regions and keep regions with at least 1 reads. These regions will be excluded from intron intervals for computing intron coverage. Also add repeatMasker regions. (downloaded from ucsc table browser 2018-07-09)

```{r drop_exons}

gtf_df <- tidy_gtf(annots$drosophila$gtf) %>% 
  bed_sort()

exons <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                transcript_id,
                exon_number,
                strand) %>% 
  bed_sort() %>% 
  group_by(strand)

ivl_df <- group_by(ivl_df, strand)
intron_coverage <- bed_subtract(ivl_df, exons)

masked_ivls <- filter(intron_coverage, score > 0) %>% 
  group_by(strand) %>% 
  bed_merge() %>% 
  ungroup()

```

```{r}
masked_out <- mutate(masked_ivls,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

dir.create("annotations")
write_tsv(masked_out, "annotations/fly_intron_mask.bed", col_names = F)
```

Make another mask the exclude exonic sequence, with some padding. 

```{r, eval = FALSE}
exons <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                transcript_id,
                exon_number,
                strand) %>% 
  bed_sort() %>%
  group_by(strand) %>% 
  bed_merge() %>% 
  group_by(strand)

intron_coverage <- bed_subtract(ivl_df, exons) %>% 
   filter(score > 0) %>% 
  dplyr::select(-score)

exon_mask <- filter(exons, end - start > 200) %>% 
  bed_slop(both = -75, genome = chrom_sizes) %>% 
  dplyr::select(chrom:end, strand)


intron_coverage <- bind_rows(intron_coverage, 
                             exon_mask)

masked_ivls <-group_by(intron_coverage, strand) %>% 
  bed_merge() %>% 
  ungroup()


masked_out <- mutate(masked_ivls,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(masked_out, "fly_exon_intron_plus_exons_juncts_mask.bed", col_names = F)
```


## Convert to transcript coordinates


```{r}
library(GenomicFeatures)

bed_df <- read_bed("annotations/fly_intron_mask.bed", n = 6)
bed_df <- filter(bed_df, chrom %in% unique(gtf_df$chrom)) 

gtf_fn <- "~/Projects/mzt-introns/dbases/drosophila/primary_transcripts.gtf"

bed <- makeGRangesFromDataFrame(bed_df)
gtf <- import(gtf_fn)
ggtf <- makeTxDbFromGRanges(gtf)
gtx <- exonsBy(ggtf, by = "tx", use.names = T)
db_out <- mapToTranscripts(x = bed, transcripts = gtx, ignore.strand = FALSE)
db_out <- as_data_frame(db_out)

db_out <- mutate(db_out,
                     start = start - 1,
                     end = end,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom = seqnames,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(db_out, "annotations/fly_intron_mask_txcoords.bed", col_names = F)
```


Per transcript masking



```{r}
library(GenomicFeatures)

bed_df <- read_bed("annotations/fly_intron_mask.bed", n = 6)
bed_df <- filter(bed_df, chrom %in% unique(gtf_df$chrom)) 

gtf_fn <- "~/Projects/mzt-introns/dbases/drosophila/primary_transcripts_per_tx.gtf"

bed <- makeGRangesFromDataFrame(bed_df)
gtf <- import(gtf_fn)
ggtf <- makeTxDbFromGRanges(gtf)
gtx <- exonsBy(ggtf, by = "tx", use.names = T)
db_out <- mapToTranscripts(x = bed, transcripts = gtx, ignore.strand = FALSE)
db_out <- as_data_frame(db_out)

db_out <- mutate(db_out,
                     start = start - 1,
                     end = end,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom = seqnames,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(db_out, "annotations/fly_intron_mask_txcoords_per_tx.bed", col_names = F)
```
