---
title: "Examine intronic signals during MZT in chicken embryos"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Background

The Maternal to Zygotic Transistion (MZT) is a developmental stage in which the embryo initiates transcription of mRNAs from the zygotic genome. Maternally deposited mRNAs begin to be degraded and are replaced by zygotic messages. 

Zygotic mRNAs can be distinguished from maternal mRNAs, due to the presence or higher proportions of introns. Maternally derived mRNAs are likely to be spliced, and any remaining intron lariat degraded prior to fertilization. In contrast zygotic transcription will produce mRNAs in that need to be spliced, and thus there will be some proportion of unspliced mRNA, and intron lariat derivied from the zygotic transcripts. 

Using ribo-zero depleted total-RNA-Seq libraries, one would expect to be able to recover intronic mRNA fragements, and be able to assign for each transcript, the relative proportion of each transcript that is maternally or zygotically derived. 


## Approach

Each total RNA-Seq lib was trimmed to remove Tru-Seq adapters (Rissland libraries), or barcodes and custom 3' adapter (Bartel libraries). The trimmed reads were then processed with `salmon` to estimate `TPMs` for each transcript. The `salmon` transcript database was modified to include a pre-mRNA entry, which was designated as an exon that spanned the entire gene. This pre-mRNA annotation matches the `gene` entry from the annotaiton file. By including the pre-mRNA transcript in the `salmon` database, one can get an estimate of the fraction of pre-mRNA or mature mRNA present in each sample. 

First we'll examine the extent to which pre-mRNA transcripts increase during embryonic development. Then, if a clear increase is present, we can classify transcripts into a % maternal and % zygotic, and additional classify the temporal pattern of zygotic transcription i.e. are there distinct sets of early versus late transcripts, and do their transcriptional pattern suggest function (TF transcription prior to target activation/repression, miRNA transcription before target repression, RBP transcription prior to translational control, etc.)

 
## RNA-Seq Libraries
```{r load_libs}
source("../../R/globals.r")
```



```{r get_metadata}

mdata <- read_tsv(file.path(data_dir, "raw_data",
                             "chicken", 
                             "HWANG", "PRJNA342320_download_log.txt"))

mdata$study <- 'HWANG'

#cleanup sample names
mdata <- mdata %>% 
  separate(sample_title, c("stage", "replicate"), sep = "_") %>% 
  select(-fastq_aspera, -fastq_ftp)

mdata
```



## Load in featurecounts data

```{r load_dat, message = F}
count_data_dir <- file.path(data_dir, 
                            "featurecounts",
                            "chicken")
count_files <- dir(count_data_dir,
                   pattern = "_counts.tsv$",
                   recursive = T,
                   full.names = T)

sample_names <- str_split(count_files, "\\/") %>% 
  map_chr(., ~.x[10]) %>% 
  str_remove("_counts.tsv") 

sample_names_simple <- str_replace(sample_names, "_exon|_intron", "")
 

dat <- map(count_files, 
           ~read_tsv(.x,
                     skip = 2,  
                     col_names = F) %>% 
             select(gene_id = X1, 
                    length = X6, 
                    counts = X7))
             

# name with libraryID
names(dat) <- sample_names

dat <- bind_rows(dat, .id = "sample_name")

dat <- mutate(
  dat,
  run_accession = str_replace(sample_name,
                    "_exon|_intron",
                    ""),
  count_type = ifelse(str_detect(sample_name, "exon"),
                      "exon",
                      "intron")
)

#join with main data
dat <- inner_join(dat, 
                  mdata,
                  by = c("run_accession"))
```

### transcript to gene mapping
```{r drop_monoexonic}
gtf <- "~/Projects/shared_dbases/annotation/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.94.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf)
gtf$seqnames <- as.character(gtf$seqnames)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  bed_merge() %>% 
  dplyr::summarize(total_exons = n())

monoexonic_genes <- dplyr::filter(gtf_summary, total_exons == 1) %>%
  pull(gene_id)
```


```{r remove_monoexonic}

dat <- anti_join(dat, 
                 data_frame(gene_id = monoexonic_genes),
                 by = "gene_id")

dat <- group_by(dat, run_accession) %>% 
  mutate(lib_size = sum(counts),
         eff_length = as.numeric(length) - 150 + 1) %>% 
  filter(eff_length > 0 ) %>% 
  mutate(rate = log(counts) - log(eff_length),
         total_rates = log(sum(exp(rate))),
         tpm = exp(rate - total_rates + log(1e6)))
```

### plot pre_mRNA ratio

```{r, calc_premRNA}
# sum primary and mature TPMs for each gene across replicates
summary_dat <- dplyr::select(dat, tpm, gene_id, 
              run_accession, count_type, stage,
              study, replicate) %>% 
  group_by(gene_id, run_accession, 
           stage, count_type) %>% 
  dplyr::summarize(total_tpm = sum(tpm, na.rm = F)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- ungroup(dat) %>% 
  dplyr::select(tpm, gene_id, 
              count_type, stage,
              study, replicate) %>% 
  spread(count_type, tpm, fill = 0L) %>% 
  mutate(primary_ratio  = intron / (intron + exon))


## calc mean primary ratio
tpm_means <- tpm_summary %>% 
  group_by(gene_id, 
           stage) %>% 
  summarize(mean_primary = mean(primary_ratio,
                                na.rm = T)) 

tpm_means <- mutate(tpm_means, 
                       stage = factor(stage, 
                                         levels = c("Oocyte",
                                                    "Zygote",
                                                    "EGKI",
                                                    "EGKIII",
                                                    "EGKVI",
                                                    "EGKVIII",
                                                    "EGKX"
                                                    )
                                         )
                       ) %>% 
  ungroup()

plt <- ggplot(tpm_means, 
       aes(stage, mean_primary)) +
  geom_boxplot(aes(fill = stage), coef = 1e3) + 
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_brewer(palette = "Set1",
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5)
  )

save_plot("chicken-pre-mRNA-ratios.pdf", plt, 
          base_width = 6,
          base_height = 4.5)

```



## Load in Salmon data
```{r read_in_salmon}

fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon", 
                        "chicken",
                        mdata$study,
                        .x,
                        mdata$run_accession,
                        "quant.sf"))

dat <- map(files, ~map(.x, read_tsv))

names(dat) <- fa_types

dat <- map(dat, function(x) {
  names(x) <- mdata$run_accession
  x
} )

dat <- map(dat, ~bind_rows(.x, .id = "run_accession"))

dat <- bind_rows(dat, .id = "expt")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("run_accession"))


# convert gene symbols to gene id
#tx2gid <- gtf %>% 
 # tbl_df() %>% 
#  filter(type == "exon") %>%
#  dplyr::select(gene_id, transcript_id) %>% 
#  unique()

#dat <- left_join(dat, tx2gid,
#                 by = c("Name" = "transcript_id")) %>% 
#  mutate(gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
#                          Name,
#                          gene_id))

#dat <- anti_join(dat, 
#                 data_frame(gene_id = monoexonic_genes),
#                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(count_type = ifelse(str_detect(Name, 
                                        "^pre_"), 
                             "intron", 
                             "exon"))

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, 
                 by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_annotations
salmon_dat <- mutate(dat, 
              gene_id = ifelse(count_type == "intron",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))
```

## alignment info

```{r}
library(jsonlite)
fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked"
)

files <- map(fa_types, 
             ~file.path(data_dir,
                        "salmon", 
                        "chicken",
                        mdata$study,
                        .x,
                        mdata$run_accession,
                        "aux_info",
                        "meta_info.json"))

align_dat <- map(files, ~map(.x, jsonlite::fromJSON)) %>% 
  map(., ~map(.x, ~.x$percent_mapped))

names(align_dat) <- fa_types


map(
  align_dat,
  ~ data_frame(
    percent_alignment = unlist(.x),
    sra_number = mdata$run_accession,
    stage = mdata$stage,
    replicate = mdata$replicate
  )
) %>% bind_rows(., .id = "fasta_type")


```




## DE testing

First read in salmon data into tximport
```{r}
fa_types <- c(
  "primary_transcripts",
  "primary_transcripts_masked"
)
# drop oocyte for now

pdata <- filter(mdata, !str_detect(stage, "Oocyte")) %>% 
  mutate(mzt = ifelse(stage %in% c("Oocyte",
                                          "Zygote",
                                          "EGKI",
                                          "EGKIII"),
                             "maternal",
                             "zygotic"))

tx_files <- map(fa_types, 
                 ~file.path(data_dir,
                   "salmon", 
                   "chicken",
                   pdata$study,
                   .x,
                   pdata$run_accession,
                   "quant.sf"))

tids <- read_tsv(tx_files[[1]][1]) %>% 
  dplyr::select(Name)

tx2gene <- gtf %>% 
  filter(type == "transcript") %>% 
  select(transcript_id, gene_id) %>% 
  tbl_df() %>% 
  unique()

tx2gene_mapping <- full_join(tids, tx2gene,
                             by = c("Name" = "transcript_id"))

# add in "gene id" for pre mRNA transcript
tx2gene_mapping <- mutate(tx2gene_mapping,
                          gene_id = ifelse(is.na(gene_id) & str_detect(Name, "^pre_"),
                                           Name,
                                           gene_id)) %>% 
  dplyr::rename(transcript_id = Name)
          
                                 
rs_tx_objs <- map(tx_files, 
               ~tximport(.x, 
                         type = "salmon",
                         tx2gene = tx2gene_mapping))

```

Run deseq with post-mzt v. pre-mzt
```{r}
rs_ddobjs <- map(rs_tx_objs, 
              function(x) {
                dds <- DESeqDataSetFromTximport(x, 
                                                pdata, 
                                                ~mzt + replicate)
                dds <- DESeq(dds)
                dds
              })

names(rs_ddobjs) <- fa_types
```

Extract results and make MA plots
```{r}
ris_res <- map_dfr(rs_ddobjs, ~results(.x, 
                                       name = "mzt_zygotic_vs_maternal", 
                                       tidy = T), 
                   .id = "fasta_type") %>% 
  tbl_df() %>% 
  mutate(count_type = ifelse(startsWith(row, "pre_"), 
                        "intron",
                        "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))

plot_ma <- function(df, plt_title = NULL) {
  n_sig_up <- nrow(dplyr::filter(df, sig == "Up")) %>%
    formatC(big.mark = ",")
  n_sig_down <-
    nrow(dplyr::filter(df, sig == "Down"))  %>%
    formatC(big.mark = ",")
  n_notsig <-
    nrow(dplyr::filter(df, sig == "Not Sig"))  %>%
    formatC(big.mark = ",")
  df <- mutate(df, 
               sig = factor(sig, 
                               levels = c("Down",
                                          "Not Sig",
                                          "Up")))
  pre_plot <- ggplot(df,
                     aes(baseMean,
                         log2FoldChange)) +
    geom_point(aes(colour = sig), size = 0.1) +
    scale_colour_viridis(
      discrete = TRUE,
      name = "",
      drop = FALSE,
      labels = c(
        paste0("Significantly decreased (n = ", n_sig_down, ")"),
        paste0("Not Significant (n= ", n_notsig, ")"),
        paste0("Significantly increased (n = ", n_sig_up, ")")
      )
    ) +
    facet_wrap(~count_type) +
    scale_x_log10() +
    labs(x = expression(paste("log"[10], " Average Expression")),
         y = expression(paste("log"[2], " ", frac(
           "Zygotic", "Maternal"
         ))),
         title = plt_title) +
    guides(colour = guide_legend(nrow = 3,
                                 override.aes = list(size = 3))) +
    theme(legend.position = "top")
  
  pre_plot
}


plts <- imap(split(ris_res, 
           list(ris_res$fasta_type,
                         ris_res$count_type)), 
    ~plot_ma(.x, .y))

plt <- plot_grid(plotlist = plts)
save_plot("ma_plot_salmon.pdf", plt, nrow = 2, ncol = 2)

```

## Plot out expression

```{r}
ris_res_masked <- filter(ris_res, 
                      fasta_type == "primary_transcripts_masked") 
  
sig_pregenes <- ris_res_masked %>% 
  filter(str_detect(row, "^pre"), 
         padj < 0.05, 
         log2FoldChange > 0) %>% 
  pull(row)

rs_ddobj_masked <- rs_ddobjs$primary_transcripts_masked

rs_ddobj_rlog<- rlog(rs_ddobj_masked)

# add in colnames
cols <- pdata %>% 
  mutate(cnames = str_c(stage, replicate, sep = "_")) %>% 
  pull(cnames)

rlog_counts <- assay(rs_ddobj_rlog)
colnames(rlog_counts) <- cols

premrnas <- str_subset(rownames(rlog_counts), "^pre")
mrnas <- setdiff(rownames(rlog_counts), premrnas)


sig_genes_ids <- str_remove(sig_pregenes, "^pre_")

mrna_rlogs <- rlog_counts[intersect(sig_genes_ids,
                                        rownames(rlog_counts)), ] 

premrna_rlogs <- rlog_counts[intersect(sig_pregenes,
                                        rownames(rlog_counts)), ] 

rownames(premrna_rlogs) <- str_remove(rownames(premrna_rlogs),
                                      "^pre_")

shared_genes <- intersect(rownames(premrna_rlogs), rownames(mrna_rlogs))

premrna_rlogs <- premrna_rlogs[shared_genes, ]
mrna_rlogs <- mrna_rlogs[shared_genes, ]

norm_counts <- counts(rs_ddobj_masked, normalized = T)
colnames(norm_counts) <- cols
zscore_norm_counts <- t(scale(t(log1p(norm_counts))))

  
ht1 <- Heatmap(zscore_norm_counts[sig_pregenes, ], 
         name = "pre-mRNA abundance\nzscore",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = FALSE,
         column_order = cols)
pdf("premrnas_heatmap_zscores.pdf")
  draw(ht1, column_title = "Chicken")
dev.off()

log_norm_counts <- log1p(norm_counts)
ht1 <- Heatmap(log_norm_counts[sig_pregenes, ], 
         name = "pre-mRNA abundance\nlog expression",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = FALSE,
         column_order = cols)
pdf("premrnas_heatmap_logexpr.pdf")
  draw(ht1, column_title = "Chicken")
dev.off()
```

```{r order_by_time_expr}

sig_pre_gene_log_counts <- log_norm_counts[sig_pregenes, ]

time_to_expr <- sig_pre_gene_log_counts %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>% 
  gather(stage, expr, -gene) %>% 
  separate(stage, c("stage", "replicate"), "_") %>% 
  as_tibble() %>% 
  group_by(gene, stage) %>% 
  summarize(m_expr = mean(expr),
            expressed = m_expr > 0) %>% 
  mutate(stage = factor(stage,
                        levels = c("Oocyte",
                                   "Zygote",
                                   "EGKI",
                                   "EGKIII",
                                   "EGKVI",
                                   "EGKVIII",
                                   "EGKX"
                        ))) %>% 
  arrange(stage, .by_group = TRUE) %>% 
  filter(expressed) %>% 
  group_by(gene) %>% 
  summarize(stage_expression = dplyr::first(stage))
            
time_to_expr <- left_join(data_frame(gene = rownames(sig_pre_gene_log_counts)),
                          time_to_expr, by = "gene")

ht1 <- Heatmap(sig_pre_gene_log_counts, 
         name = "pre-mRNA\nlog expression",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = FALSE,
         column_order = cols, 
         split = time_to_expr$stage_expression)

pdf("premrnas_heatmap_logexpr_time_staged.pdf", height = 16, width = 6)
  draw(ht1, column_title = "Chicken")
dev.off()
```

## Examine motifs

### 3' UTRs
```{r clustsr}
library(kentr)
gtffile <- "~/Projects/shared_dbases/annotation/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.94.gtf"
chromfile <- "~/Projects/shared_dbases/genomes/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.dna.toplevel.fa.fai"
genomefile <- "~/Projects/shared_dbases/genomes/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.dna.toplevel.fa"

gnome <- read_tsv(chromfile, col_types = "ci---", 
                  col_names= c("chrom", "size"))

get_3utrs <- function(gtf_file, gene_ids, 
                      gene_col = "gene_id", 
                      transcript_col = "transcript_id"){
  gtf <- import(gtf_file)
  gtf <- as.data.frame(gtf)
  gtf <- filter(gtf, gene_id %in% gene_ids)
  gtf <- dplyr::select(gtf, 
                       chrom = seqnames, 
                       start, end, type, strand, exon_id, one_of(c(gene_col, transcript_col)))
  gtf <- dplyr::filter(gtf, 
                       type %in% c("exon", "CDS"))
  cds_gtf <- filter(gtf, type == "CDS") %>% 
         group_by(transcript_id) %>% 
         summarize(cds_start = unique(ifelse(strand == "+",
                                  min(start), 
                                  max(end))),
                   cds_end = unique(ifelse(strand == "+", 
                                max(end),
                                min(start)))) %>% 
    left_join(gtf, ., by = "transcript_id")
  
  utr3_gtf <- cds_gtf %>% 
    filter(type == "exon") %>% 
    group_by(transcript_id) %>% 
    mutate(utr3_exon = ifelse(strand == "+",
                              ifelse(start >= cds_end,
                                     "yes",
                                     ifelse(cds_end >= start & cds_end <= end,
                                            "overlapping",
                                            "no")),
                              ifelse(end <= cds_end,
                                     "yes",
                                     ifelse(cds_end >= start & cds_end <= end,
                                            "overlapping",
                                            "no")))) %>% 
    ungroup() %>% 
    filter(utr3_exon != "no")
  
  utr3_res <- utr3_gtf %>% 
    group_by(transcript_id) %>% 
    mutate(start = ifelse(strand == "+" & utr3_exon == "overlapping",
                          cds_end,
                          start),
           end = ifelse(strand == "-" & utr3_exon == "overlapping",
                        cds_end,
                        end))  %>% 
    filter(start != end) %>% 
    select(chrom:transcript_id)
  
  utr3_res <- mutate(utr3_res, start = start - 1)
  
  res <- group_by(utr3_res, gene_id, strand) %>% 
    bed_merge() %>% 
    ungroup()
  res
}


gtf <- import(gtffile)
gtf <- as.data.frame(gtf)

all_genes <- ris_res_masked %>% 
  pull(row) %>% 
  unique()

all_utrs <- get_3utrs(gtffile, all_genes)
all_utrs <- group_by(all_utrs, gene_id) %>% 
  mutate(n = row_number(),
         id = str_c(gene_id, "_", n)) %>% 
  select(chrom, start, end, id, gene_id, strand) %>% 
  ungroup() 

# pad any utrs less than 10 in length
all_annotated_utrs <- all_utrs %>% filter(end - start >= 10)
all_short_utrs <- all_utrs %>% filter(end - start < 10) %>% 
  bed_slop(., right = 750, strand = TRUE, genome = gnome)

all_utrs <- bind_rows(all_annotated_utrs, all_short_utrs) %>% 
  bed_sort()

down_genes <- ris_res_masked %>% 
  filter(count_type == "exon", log2FoldChange < 0, padj < 0.05) %>% 
  pull(row)

write_bed <- function(df, path = ""){
  res <- mutate_if(df, is.numeric, as.character)
  write_tsv(res, path, col_names = F)
}

utrs <- filter(all_utrs, gene_id %in% down_genes)
rna_outdir <- "utr_fastas"
dir.create(rna_outdir)
write_bed(utrs, file.path(rna_outdir, "downregulated_utrs.bed"))


write_bed(all_utrs, file.path(rna_outdir, "all_utrs.bed"))

genomefile <- "~/Projects/shared_dbases/genomes/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.dna.toplevel.fa"

utrs_seqs <- get_sequences(utrs, genomefile)
write_fasta(utrs_seqs, file.path(rna_outdir,
                                 "downregulated_utrs.fasta"))

all_utrs_seqs <- get_sequences(all_utrs, genomefile)
write_fasta(all_utrs_seqs, file.path(rna_outdir, "all_utrs.fasta"))
```


```{r down_genes_pattern}
Heatmap(zscore_norm_counts[down_genes, ], 
         name = "pre-mRNA\nlog expression",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = FALSE,
         column_order = cols)
```


```{r kmers}

kmers <- mutate(utrs_seqs, 
                kmers = get_kmers(seq, n = 6)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

kmers_all <- mutate(all_utrs_seqs, 
                kmers = get_kmers(seq, n = 6)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

bg <- kmers_all %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmer_count <- kmers %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmers_total <- left_join(kmer_count, 
                         bg,
                         by = "kmer",
                         suffix = c("_test", "_bg"))
kmers_total <- mutate(kmers_total, 
                      n_seq_test = nrow(utrs_seqs),
                      n_seq_bg = nrow(all_utrs_seqs))
out <- kmers_total %>% 
  na.omit(.) %>%
  rowwise() %>% 
  do(broom::tidy(binom.test(.$counts_test, .$n_seq_test, 
                     p = .$counts_bg / .$n_seq_bg),
          alternative ="greater")) 

res <- bind_cols(kmers_total, out) %>% 
  mutate(padj = p.adjust(p.value),
         prop.test = counts_test / n_seq_test,
         prop.bg = counts_bg / n_seq_bg,
         enrichment = log2(prop.test / prop.bg),
         log10pval = -log10(padj)) %>% 
  mutate(qtile = quantile(log10pval, .995),
         enriched = ifelse(log10pval > qtile, T, F)) %>% 
  arrange(padj)

labeled_seqs <- filter(res, enriched)

library(ggrepel)
mer6_plt <- ggplot(res, aes(enrichment, log10pval)) +
  geom_point() +
  geom_text_repel(data = labeled_seqs, 
                  aes(label = kmer), color = "red", 
                  force = 10) +
  labs(title = "6mers enriched in 3' UTRs downregulated mRNAs", x = "enrichment (log2)")
kmer_res <- list(`6mer` = res)
```

```{r 8mers}

kmers <- mutate(utrs_seqs, 
                kmers = get_kmers(seq, n = 8)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

kmers_all <- mutate(all_utrs_seqs, 
                kmers = get_kmers(seq, n = 8)) %>% 
                select(gene_id, start, end, kmers) %>% 
  unnest() %>% 
  group_by(gene_id, kmer) %>% 
  summarize(counts = sum(counts)) %>% 
  ungroup()

bg <- kmers_all %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmer_count <- kmers %>% 
  mutate(success = ifelse(counts >= 1, 1L, 0L)) %>% 
  group_by(kmer) %>% 
  summarize(counts = sum(success))

kmers_total <- left_join(kmer_count, 
                         bg,
                         by = "kmer",
                         suffix = c("_test", "_bg"))
kmers_total <- mutate(kmers_total, 
                      n_seq_test = nrow(utrs_seqs),
                      n_seq_bg = nrow(all_utrs_seqs))
out <- kmers_total %>% 
  na.omit(.) %>%
  rowwise() %>% 
  do(broom::tidy(binom.test(.$counts_test, .$n_seq_test, 
                     p = .$counts_bg / .$n_seq_bg),
          alternative ="greater")) 

res <- bind_cols(kmers_total, out) %>% 
  mutate(padj = p.adjust(p.value),
         prop.test = counts_test / n_seq_test,
         prop.bg = counts_bg / n_seq_bg,
         enrichment = log2(prop.test / prop.bg),
         log10pval = -log10(padj)) %>% 
  arrange(padj) %>% 
  mutate(rank = row_number(),
         enriched = ifelse(rank <= 20, T, F)) %>% 
  arrange(padj)

labeled_seqs <- filter(res, enriched)

mer8_plt <- ggplot(res, aes(enrichment, log10pval)) +
  geom_point() +
  geom_text_repel(data = labeled_seqs, 
                  aes(label = kmer), color = "red", 
                  force = 10) +
  labs(title = "8mers enriched in 3' UTRs downregulated mRNAs", x = "enrichment (log2)")
kmer_res$`8mer` <- res
mer8_plt


plt <- plot_grid(mer6_plt, mer8_plt)
save_plot("kmer_enrichment.pdf", plt, nrow = 1, ncol = 2, base_height = 6)

```


```{r}
mirs <- read_file("ftp://mirbase.org/pub/mirbase/CURRENT/mature.fa.gz")
mirs %>% 
  str_split(., ">") %>% 
  .[[1]] %>% 
  str_trim(.) %>% 
  data.frame(mirs = .) %>% 
  separate(mirs, c("name", "seq"), sep = "\n") %>% 
  na.omit() -> mir_df

chicken_mirs <- filter(mir_df, str_detect(name, "^gga"))

chicken_mirs <- chicken_mirs %>% 
  mutate(mer6  = str_sub(revComp(str_replace_all(chicken_mirs$seq, "U", "T")), 2, 8),
         mer8 = str_sub(revComp(str_replace_all(chicken_mirs$seq, "U", "T")), 2, 10))

left_join(kmer_res)
```
### Promoter TF binding enrichment 

```{r}
gtffile <- "~/Projects/shared_dbases/annotation/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.94.gtf"
chromfile <- "~/Projects/shared_dbases/genomes/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.dna.toplevel.fa.fai"
genomefile <- "~/Projects/shared_dbases/genomes/chicken/gallus_gallus_5/Gallus_gallus.Gallus_gallus-5.0.dna.toplevel.fa"
gtf <- import(gtffile)
gtf <- as.data.frame(gtf)

tss <- filter(gtf, type == "gene") %>%
  mutate(start = start - 1) %>% #convert gtf to bed coords
  mutate(end = ifelse(strand == "+",
                      start + 1,
                      end),
         start = ifelse(strand == "-",
                        end - 1,
                        start)) %>% 
  select(chrom = seqnames, start, end, gene_name, gene_id, strand)


tss_slop <- bed_slop(tss, 
                     genome = gnome, 
                     left = 5000, right = 500, strand = T) %>% 
  unique() %>% 
  bed_sort()

write_bed(tss_slop, "chicken_tss.bed")

tss_slop <- bed_slop(tss, 
                     genome = gnome, 
                     left = 2000, right = 500, strand = T) %>% 
  unique() %>% 
  bed_sort()

write_bed(tss_slop, "chicken_tss_2k.bed")


tss_slop <- get_sequences(tss_slop, genomefile)
write_fasta(tss_slop, "all_tss.fasta")


up_genes <- ris_res_masked %>% 
  filter(count_type == "intron", log2FoldChange > 0, padj < 0.05) %>% 
  pull(row) %>% 
  str_remove(., "^pre_")

tss_slop_up <- filter(tss_slop, gene_id %in% up_genes)

write_fasta(tss_slop_up, "upregulated_tss.fasta")
```

#### 5k window 

```{r jaspar_enrichment, fig.height= 4, fig.width= 3}

jaspar_tfs <- read_tsv(file.path(db_dir, "chicken",
                            "tfsummary",
                            "tf_5k_500bp_summary.tsv.gz"),
                       col_names = c("tmp",
                                     "promoter"),
                       col_types = c('cc')) %>% 
  separate("tmp", into = c("n_binding_sites", "tf"), sep = " ") %>% 
  mutate(n_binding_sites = as.integer(n_binding_sites))

time_to_expr <- time_to_expr %>% 
  mutate(stage_expression = ifelse(stage_expression == "Zygote",
                                   "pre_mzt",
                                   "mzt"))

clusters <- time_to_expr %>% 
  mutate(gene_id = str_remove(gene, "pre_")) %>% 
  select(-gene)

clusters <- mutate(clusters,
                   cluster = as.character(stage_expression)) %>% 
  select(-stage_expression) %>% 
  group_by(cluster) %>% 
  mutate(total_genes_in_cluster = n())

tf_res <- left_join(clusters, jaspar_tfs, 
                       by = c("gene_id" = "promoter")) %>% 
  group_by(tf, cluster) %>% 
  mutate(bound_in_cluster = n()) %>% 
  select(tf, cluster, bound_in_cluster, total_genes_in_cluster) %>% 
  ungroup() %>% 
  unique()

bkgd <- jaspar_tfs %>% 
    filter(promoter %in% ris_res$row) %>% 
  group_by(tf) %>% 
  summarize(n_bound_in_bg = n()) %>% 
  mutate(total_genes_in_bg = length(unique(str_remove(ris_res$row,
                                                      "^pre_"))),
         not_bound_in_bg = total_genes_in_bg - n_bound_in_bg) %>% 
  ungroup()


tf_summary <- left_join(tf_res, 
          bkgd,
          by = c("tf")) %>% 
  ungroup()

summary_stats <- tf_summary %>% 
  filter(!is.na(tf)) %>% 
  group_by(tf, cluster) %>% 
  do(broom::tidy(phyper(.$bound_in_cluster - 1, .$n_bound_in_bg, 
                        .$not_bound_in_bg, .$total_genes_in_cluster, lower.tail = FALSE))) %>% 
  ungroup() %>% 
  mutate(padj = p.adjust(x)) %>% 
  inner_join(tf_summary, ., by = c("tf", "cluster")) %>% 
  mutate(enrichment = log2((bound_in_cluster / total_genes_in_cluster) /
                      (n_bound_in_bg / total_genes_in_bg)),
         log10_padj = -log10(padj)) %>% 
  select(everything(), pval = x)

summary_stats <- group_by(summary_stats, 
                          cluster) %>% 
  arrange(desc(log10_padj), .by_group = T) %>% 
  mutate(rank = row_number())

plt <- ggplot(summary_stats, aes(rank, log10_padj)) +
  geom_point() +
  facet_wrap(~cluster, scales = "free")

save_plot("tf_enrichment.pdf", plt)
View(summary_stats)
```

Using just JASPAR predictions has too many predictions to examine enrichment meaningfully. For many TFs over 80% of genes have predicted binding sites with - 5k and + 500bp of the TSS. 

Next try two approachs:
  1) use conservation to restain the predicitons to conserved elements
  2) Try making the window around the promoter smaller (2kbp?)
  
```{r jaspar_w_gerp_enrichment, fig.height= 4, fig.width= 3}
  
jaspar_tfs <- read_tsv(file.path(db_dir, "chicken",
                            "tfsummary",
                            "tf_5k_500bp_gerp_summary.tsv.gz"),
                       col_names = c("tmp",
                                     "promoter"),
                       col_types = c('cc')) %>% 
  separate("tmp", into = c("n_binding_sites", "tf"), sep = " ") %>% 
  mutate(n_binding_sites = as.integer(n_binding_sites))


clusters <- time_to_expr %>% 
  mutate(gene_id = str_remove(gene, "pre_")) %>% 
  select(-gene)

clusters <- mutate(clusters,
                   cluster = as.character(stage_expression)) %>% 
  select(-stage_expression) %>% 
  group_by(cluster) %>% 
  mutate(total_genes_in_cluster = n())

tf_res <- left_join(clusters, jaspar_tfs, 
                       by = c("gene_id" = "promoter")) %>% 
  group_by(tf, cluster) %>% 
  mutate(bound_in_cluster = n()) %>% 
  select(tf, cluster, bound_in_cluster, total_genes_in_cluster) %>% 
  ungroup() %>% 
  unique()

bkgd <- jaspar_tfs %>% 
    filter(promoter %in% ris_res$row) %>% 
  group_by(tf) %>% 
  summarize(n_bound_in_bg = n()) %>% 
  mutate(total_genes_in_bg = length(unique(str_remove(ris_res$row,
                                                      "^pre_"))),
         not_bound_in_bg = total_genes_in_bg - n_bound_in_bg) %>% 
  ungroup


tf_summary <- left_join(tf_res, 
          bkgd,
          by = c("tf")) %>% 
  ungroup()

summary_stats <- tf_summary %>% 
  filter(!is.na(tf)) %>% 
  group_by(tf, cluster) %>% 
  do(broom::tidy(phyper(.$bound_in_cluster - 1, .$n_bound_in_bg, 
                        .$not_bound_in_bg, .$total_genes_in_cluster, lower.tail = FALSE))) %>% 
  ungroup() %>% 
  mutate(padj = p.adjust(x)) %>% 
  inner_join(tf_summary, ., by = c("tf", "cluster")) %>% 
  mutate(enrichment = log2((bound_in_cluster / total_genes_in_cluster) /
                      (n_bound_in_bg / total_genes_in_bg)),
         log10_padj = -log10(padj)) %>% 
  select(everything(), pval = x)


ggplot(summary_stats, aes(enrichment, log10_padj)) +
  geom_point() +
  facet_grid(~cluster)
```

#### 2k window

```{r jaspar_2k, fig.height= 4, fig.width= 3}

jaspar_tfs <- read_tsv(file.path(db_dir, "chicken",
                            "tfsummary",
                            "tf_2k_500bp_summary.tsv.gz"),
                       col_names = c("tmp",
                                     "promoter"),
                       col_types = c('cc')) %>% 
  separate("tmp", into = c("n_binding_sites", "tf"), sep = " ") %>% 
  mutate(n_binding_sites = as.integer(n_binding_sites))


clusters <- time_to_expr %>% 
  mutate(gene_id = str_remove(gene, "pre_")) %>% 
  select(-gene)

clusters <- mutate(clusters,
                   cluster = as.character(stage_expression)) %>% 
  select(-stage_expression) %>% 
  group_by(cluster) %>% 
  mutate(total_genes_in_cluster = n())

tf_res <- left_join(clusters, jaspar_tfs, 
                       by = c("gene_id" = "promoter")) %>% 
  group_by(tf, cluster) %>% 
  mutate(bound_in_cluster = n()) %>% 
  select(tf, cluster, bound_in_cluster, total_genes_in_cluster) %>% 
  ungroup() %>% 
  unique()

bkgd <- jaspar_tfs %>% 
    filter(promoter %in% ris_res$row) %>% 
  group_by(tf) %>% 
  summarize(n_bound_in_bg = n()) %>% 
  mutate(total_genes_in_bg = length(unique(str_remove(ris_res$row,
                                                      "^pre_"))),
         not_bound_in_bg = total_genes_in_bg - n_bound_in_bg) %>% 
  ungroup


tf_summary <- left_join(tf_res, 
          bkgd,
          by = c("tf")) %>% 
  ungroup()

summary_stats <- tf_summary %>% 
  filter(!is.na(tf)) %>% 
  group_by(tf, cluster) %>% 
  do(broom::tidy(phyper(.$bound_in_cluster - 1, .$n_bound_in_bg, 
                        .$not_bound_in_bg, .$total_genes_in_cluster, lower.tail = FALSE))) %>% 
  ungroup() %>% 
  mutate(padj = p.adjust(x)) %>% 
  inner_join(tf_summary, ., by = c("tf", "cluster")) %>% 
  mutate(enrichment = log2((bound_in_cluster / total_genes_in_cluster) /
                      (n_bound_in_bg / total_genes_in_bg)),
         log10_padj = -log10(padj)) %>% 
  select(everything(), pval = x)


ggplot(summary_stats, aes(enrichment, log10_padj)) +
  geom_point() +
  facet_grid(~cluster)
```

```{r jaspar_w_gerp_enrichment_2k, fig.height= 4, fig.width= 3}

jaspar_tfs <- read_tsv(file.path(db_dir, "chicken",
                            "tfsummary",
                            "tf_2k_500bp_gerp_summary.tsv.gz"),
                       col_names = c("tmp",
                                     "promoter"),
                       col_types = c('cc')) %>% 
  separate("tmp", into = c("n_binding_sites", "tf"), sep = " ") %>% 
  mutate(n_binding_sites = as.integer(n_binding_sites))


clusters <- time_to_expr %>% 
  mutate(gene_id = str_remove(gene, "pre_")) %>% 
  select(-gene)

clusters <- mutate(clusters,
                   cluster = as.character(stage_expression)) %>% 
  select(-stage_expression) %>% 
  group_by(cluster) %>% 
  mutate(total_genes_in_cluster = n())

tf_res <- left_join(clusters, jaspar_tfs, 
                       by = c("gene_id" = "promoter")) %>% 
  group_by(tf, cluster) %>% 
  mutate(bound_in_cluster = n()) %>% 
  select(tf, cluster, bound_in_cluster, total_genes_in_cluster) %>% 
  ungroup() %>% 
  unique()

bkgd <- jaspar_tfs %>% 
    filter(promoter %in% ris_res$row) %>% 
  group_by(tf) %>% 
  summarize(n_bound_in_bg = n()) %>% 
  mutate(total_genes_in_bg = length(unique(str_remove(ris_res$row,
                                                      "^pre_"))),
         not_bound_in_bg = total_genes_in_bg - n_bound_in_bg) %>% 
  ungroup


tf_summary <- left_join(tf_res, 
          bkgd,
          by = c("tf")) %>% 
  ungroup()

summary_stats <- tf_summary %>% 
  filter(!is.na(tf)) %>% 
  group_by(tf, cluster) %>% 
  do(broom::tidy(phyper(.$bound_in_cluster - 1, .$n_bound_in_bg, 
                        .$not_bound_in_bg, .$total_genes_in_cluster, lower.tail = FALSE))) %>% 
  ungroup() %>% 
  mutate(padj = p.adjust(x)) %>% 
  inner_join(tf_summary, ., by = c("tf", "cluster")) %>% 
  mutate(enrichment = log2((bound_in_cluster / total_genes_in_cluster) /
                      (n_bound_in_bg / total_genes_in_bg)),
         log10_padj = -log10(padj)) %>% 
  select(everything(), pval = x)


ggplot(summary_stats, aes(enrichment, log10_padj)) +
  geom_point() +
  facet_grid(~cluster)
```

### Get promoter fastas for homer

```{r}

clusters <- time_to_expr %>% 
  mutate(gene_id = str_remove(gene, "pre_")) %>% 
  select(-gene)


tss <- filter(gtf, type == "gene") %>%
  mutate(start = start - 1) %>% #convert gtf to bed coords
  mutate(end = ifelse(strand == "+",
                      start + 1,
                      end),
         start = ifelse(strand == "-",
                        end - 1,
                        start)) %>% 
  select(chrom = seqnames, start, end, gene_name, gene_id, strand)

gnome <- read_tsv(chromfile, col_types = "ci---", 
                  col_names= c("chrom", "size"))

tss_slop <- bed_slop(tss, 
                     genome = gnome, 
                     left = 5000, right = 500, strand = T, 
                     trim = TRUE) %>% 
  unique() %>% 
  bed_sort()

tss_per_cluster <- left_join(clusters, tss_slop, by = "gene_id")

per_cluster_tss_seqs <- get_sequences(tss_per_cluster, genomefile)

outdir <- "promoter_fastas"
dir.create(outdir)

all_tss_seqs <- get_sequences(tss_slop, genomefile)
write_fasta(all_tss_seqs, file.path(outdir, "all_tss.fasta"))

cluster_seqs <- split(per_cluster_tss_seqs,
                      per_cluster_tss_seqs$stage_expression)

iwalk(cluster_seqs, 
     ~write_fasta(.x, file.path(outdir, 
                                paste0(.y,
                                       "_cluster.fasta"))))
```