---
title: "Zelda Binding"
author: "Kent Riemondy RBI"
date: "4/6/2018"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
```

```{r libs}
library(valr)
library(tidyverse)
library(rtracklayer)
library(doParallel)
source("../../R/globals.r")
```

## Examine zelda binding


```{r eisen_dat}
if(!file.exists("eisen_zelda_chip.tsv")){
  download.file("https://doi.org/10.1371/journal.pgen.1002266.s007", "eisen_zelda_chip.tsv")
}

zelda_chip <- read_tsv("eisen_zelda_chip.tsv", 
              col_names = T, 
              skip = 1)

zelda_peaks <- dplyr::select(zelda_chip,
                             chrom = Chr, 
                             start = From,
                             end = To,
                             gene = TSS_gene,
                             dist = TSS_dist,
                             score = Height)

zelda_10k <- zelda_peaks %>% filter(dist <= 10000)
zelda_10k_summed <- group_by(zelda_10k, gene) %>% summarize(binding_sites = n(),
                                                            score = mean(score))
```

```{r diff_genes}
de_res <- read_tsv("../2018-04-18_drosophila/all_genes_2018-05-01.tsv")

de_res <- left_join(de_res,
                    zelda_10k_summed,
                    by = c("gene_name" = "gene")) %>% 
  mutate(binding_sites = ifelse(is.na(binding_sites),
                                0,
                                binding_sites),
         score = ifelse(is.na(score),
                        0,
                        score))
   #      score = log2(score + 1))

clusters <- read_tsv("../2018-04-18_drosophila/fly_pre_km_clusters.tsv")

de_res <- left_join(de_res, clusters, by = c("gene_id", "gene_name"))

counts_cols <- read_tsv("../2018-04-18_drosophila/fly_pre_norm_counts.tsv", col_names = F, n_max = 1)
counts <- read_tsv("../2018-04-18_drosophila/fly_pre_norm_counts.tsv", col_names = F, skip = 1)
colnames(counts) <- c("gene_id", counts_cols)
counts <- as.data.frame(counts)
rownames(counts) <- counts[, 1]
counts[, 1] <- NULL

zscores <- t(scale(t(log2(counts + 1))))

cluster_res <- de_res %>% filter(!is.na(cluster))
annotation_data <- cluster_res[, c("gene_id", "score", "binding_sites", "cluster")] %>% as.data.frame()
rownames(annotation_data) <- annotation_data[, 1]
annotation_data[, 1] <- NULL
```

```{r hmaps, fig.width = 8, fig.heigth = 12}
library(ComplexHeatmap)
library(circlize)


time_order <- c(
  "Stage 11 oocyte::wt",
  "Stage 12 oocyte::wt",
  "Stage 13 oocyte::wt",
  "Stage 14 oocyte::wt",
  "Activated egg::wt",
  "0-1 hr embryo::wt",
  "0-1 hr embryo::w1118",
  "0-1 hr embryo::eGFP-ME31B",
  "1-2 hr embryo::w1118",
  "1-2 hr embryo::eGFP-ME31B",  
  "2-3 hr embryo::wt",
  "2-3 hr embryo::w1118",
  "2-3 hr embryo::eGFP-ME31B",  
  "3-4 hr embryo::wt",
  "3-4 hr embryo::w1118",
  "3-4 hr embryo::eGFP-ME31B",
  "4-5 hr embryo::w1118",
  "4-5 hr embryo::eGFP-ME31B",
  "5-6 hr embryo::wt"
)

ht <- Heatmap(zscores[rownames(annotation_data), ],
              name = "pre-mRNA \nZscore",
              split = annotation_data$cluster,
              show_row_names = F,
              show_row_dend = F,
              column_order = time_order,
              cluster_columns = F)

ha <- HeatmapAnnotation(annotation_data[, c("score", "binding_sites")],
                        col = list(
                          score = colorRamp2(c(0, max(annotation_data$score, na.rm = T)), 
                                             c("white", "darkred")),
                          binding_sites = colorRamp2(c(0, max(annotation_data$binding_sites, 
                                                              na.rm = T)), c("white", "darkblue"))),
                        which = "row", width = unit(1, "cm"),
                        annotation_legend_param = list(score = list(title = "Zelda\nbinding\nscore"),
                                                       binding_sites = list(title = "N Zelda\nbinding\nsites")))
ht + ha

pdf("zelda_binding_premRNA_clusters_zscores.pdf",width = 5)
print(ht + ha)
dev.off()


ht <- Heatmap(counts[rownames(annotation_data), ],
              name = "pre-mRNA \n counts",
              split = annotation_data$cluster,
              show_row_names = F,
              show_row_dend = F,
              column_order = time_order,
              cluster_columns = F)

ha <- HeatmapAnnotation(annotation_data[, c("score", "binding_sites")],
                        col = list(
                          score = colorRamp2(c(0, max(annotation_data$score, na.rm = T)), 
                                             c("white", "darkred")),
                          binding_sites = colorRamp2(c(0, max(annotation_data$binding_sites, 
                                                              na.rm = T)), c("white", "darkblue"))),
                        which = "row", width = unit(1, "cm"),
                        annotation_legend_param = list(score = list(title = "Zelda\nbinding\nscore"),
                                                       binding_sites = list(title = "N Zelda\nbinding\nsites")))
ht + ha

pdf("zelda_binding_premRNA_clusters_logcounts.pdf", width = 5)
print(ht + ha)
dev.off()
```

```{r zld_boxplots}
annotation_data <- mutate(annotation_data, cluster = as.character(cluster))
score <- ggplot(annotation_data, 
       aes(cluster, score)) +
  geom_boxplot(aes(fill = cluster), coef = Inf) +
  scale_fill_viridis_d() +
  labs(y = "Zelda Chip-Seq Signal") +
  theme(legend.pos = "none")

bsites <- ggplot(annotation_data, 
       aes(cluster, binding_sites)) +
  geom_boxplot(aes(fill = cluster), coef = Inf) +
  scale_fill_viridis_d() +
  labs(y = "# of Zelda Chip-Seq Peaks") +
  theme(legend.pos = "none")

plt <- plot_grid(score, bsites, nrow = 1)
save_plot("zelda_boxplots.pdf", plt, nrow = 1, ncol = 2)
```


### mature mRNA

```{r mature_hmaps, fig.width = 8, fig.heigth = 12}
de_res <- read_tsv("../2018-04-18_drosophila/all_genes_2018-05-01.tsv")

de_res <- left_join(de_res,
                    zelda_10k_summed,
                    by = c("gene_name" = "gene")) %>% 
  mutate(binding_sites = ifelse(is.na(binding_sites),
                                0,
                                binding_sites),
         score = ifelse(is.na(score),
                        0,
                        score))
   #      score = log2(score + 1))

clusters <- read_tsv("../2018-04-18_drosophila/fly_mat_km_clusters.tsv")

de_res <- left_join(de_res, clusters, by = c("gene_id", "gene_name"))

counts_cols <- read_tsv("../2018-04-18_drosophila/fly_mature_norm_counts.tsv", col_names = F, n_max = 1)
counts <- read_tsv("../2018-04-18_drosophila/fly_mature_norm_counts.tsv", col_names = F, skip = 1)
colnames(counts) <- c("gene_id", counts_cols)
counts <- as.data.frame(counts)
rownames(counts) <- counts[, 1]
counts[, 1] <- NULL

zscores <- t(scale(t(log2(counts + 1))))

cluster_res <- de_res %>% filter(!is.na(cluster))
annotation_data <- cluster_res[, c("gene_id", "score", "binding_sites", "cluster")] %>% as.data.frame()
rownames(annotation_data) <- annotation_data[, 1]
annotation_data[, 1] <- NULL

ht <- Heatmap(zscores[rownames(annotation_data), ],
              name = "mature-mRNA \nZscore",
              split = annotation_data$cluster,
              show_row_names = F,
              show_row_dend = F,
              column_order = time_order,
              cluster_columns = F)

ha <- HeatmapAnnotation(annotation_data[, c("score", "binding_sites")],
                        col = list(
                          score = colorRamp2(c(0, max(annotation_data$score, na.rm = T)), 
                                             c("white", "darkred")),
                          binding_sites = colorRamp2(c(0, max(annotation_data$binding_sites, 
                                                              na.rm = T)), c("white", "darkblue"))),
                        which = "row", width = unit(1, "cm"),
                        annotation_legend_param = list(score = list(title = "Zelda\nbinding\nscore"),
                                                       binding_sites = list(title = "N Zelda\nbinding\nsites")))
ht + ha

pdf("zelda_binding_matmRNA_clusters_zscores.pdf", width = 5)
print(ht + ha)
dev.off()



```

```{r cluster_4_genes, eval = F}
de_res %>% 
  filter(cluster == 4) %>% 
  arrange(`padj_pre-mRNA`)

genes_to_query <- de_res %>% 
  filter(cluster == 4) %>% 
  arrange(`padj_pre-mRNA`) %>% 
  pull(gene_id)

library(kentr)
gtffile <- "~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf"

get_3utrs <- function(gtf_file, gene_ids, gene_col = "gene_id", transcript_col = "transcript_id"){
  gtf <- import(gtf_file)
  gtf <- as.data.frame(gtf)
  gtf <- filter(gtf, gene_id %in% gene_ids)
  gtf <- dplyr::select(gtf, 
                       chrom = seqnames, 
                       start, end, type, strand, exon_id, one_of(c(gene_col, transcript_col)))
  gtf <- dplyr::filter(gtf, 
                       type %in% c("exon", "CDS"))
  cds_gtf <- filter(gtf, type == "CDS") %>% 
         group_by(transcript_id) %>% 
         summarize(cds_start = unique(ifelse(strand == "+",
                                  min(start), 
                                  max(end))),
                   cds_end = unique(ifelse(strand == "+", 
                                max(end),
                                min(start)))) %>% 
    left_join(gtf, ., by = "transcript_id")
  
  cds_gtf <- cds_gtf %>% 
    filter(type == "exon") %>% 
    group_by(transcript_id) %>% 
    mutate(., start = ifelse(strand == "+", cds_end, start), 
              end = ifelse(strand == "+", end, cds_start)) %>%
    ungroup() %>% 
    select(chrom:transcript_id)  %>%
    filter(start <= end)
  
  res <- group_by(cds_gtf, gene_id, strand) %>% 
    bed_merge() %>% 
    ungroup()
  res
}

genome_file <- "~/Projects/shared_dbases/genomes/drosophila/Drosophila_melanogaster.BDGP6.dna.toplevel.fa"

utrs <- get_3utrs(gtffile, de_res$gene_id)
utr_seqs <- kentr::get_sequences(utrs, genome_file)

#kmers <- mutate(utr_seqs, 
#                kmers = get_kmers(seq, n = 6)) %>% 
#                select(gene_id, start, end, kmers) %>% 
#  unnest()
#
#cluster_ids <- select(cluster_res, cluster, gene_id)
#kmers <- left_join(kmers, cluster_ids, by = "gene_id")
#
### calc utr length and sum multiple exons per utr
#kmers <- mutate(kmers,
#                exon_len = end - start) %>% 
#  group_by(gene_id, kmer, cluster) %>% 
#  summarize(counts = sum(counts),
#            utr_len = sum(exon_len))
#
#kmers <- ungroup(kmers)
#
#missing_kmers <- expand.grid(rep(list(c("A", "T", "C", "G")), 6)) %>%
#  apply(. , 1, function(x) str_c(x, collapse = "")) %>% 
#  data_frame(kmer = .)
#
#missing_kmers <- crossing(kmer = missing_kmers$kmer, 
#                          gene_id = de_res$gene_id) %>% 
#  left_join(de_res[, c("gene_id", "cluster")], 
#            by = "gene_id")
#
#kmers <- left_join(missing_kmers, 
#                   kmers, by = c("kmer", "gene_id", "cluster"))
#rm(missing_kmers)
#
#kmers <- mutate(kmers, 
#                kmer_density = ifelse(is.na(counts),
#                                      0,
#                                      counts / utr_len))
#
#mer6_summary <- group_by(kmers, cluster, kmer) %>% 
#  summarize(kmer_density = mean(counts, na.rm = T)) %>% 
#  ungroup()
#
#mer6_summary <- mutate(mer6_summary,
#                       cluster = ifelse(is.na(cluster),
#                                        "background",
#                                        as.character(cluster)))
#mer6_summary <- spread(mer6_summary, cluster, kmer_density)

```

### DroidDB database

Use predictions from the droidb database to curate a set of TFs regulating genes in each expression cluster. Use hypergeometric tests to test for each cluster if there is an enrichment for TF target genes over a background set of genes (all expressed genes or all genes?).

```{r droidb, fig.height= 4, fig.width= 3}
if (!file.exists("droidb_tf_gene.txt")){
  download.file("http://droidb.org/data/DroID_v2015_12/tf_gene.txt", "droidb_tf_gene.txt")
}

droid <- read_tsv("droidb_tf_gene.txt")
up_pre <- read_tsv("../2018-04-18_drosophila/up_pre_not_up_mature.tsv") %>% 
  mutate(cluster = "increased_premRNA_no_mature") %>% 
  select(gene_id = row, cluster, gene_name)

clusters <- mutate(clusters, cluster = as.character(cluster)) %>% 
  bind_rows(., up_pre) %>% 
  group_by(., cluster) %>% 
  mutate(total_genes_in_cluster = n())

droid_res <- left_join(clusters, droid, by = c("gene_id" = "FLY_TARGET_GENE"))

droid_summary <- group_by(droid_res, TF_SYMBOL, cluster, total_genes_in_cluster) %>% 
  summarize(bound_genes_in_cluster = n())

bkgd <- dplyr::select(droid, FLY_TARGET_GENE, TF_SYMBOL) %>% 
  unique() %>% 
  group_by(TF_SYMBOL) %>% 
  summarize(bound_in_bg = n())

expr_genes <- read_tsv("../2018-04-18_drosophila/all_genes_2018-05-01.tsv")

total_genes <- length(unique(expr_genes$gene_name))
bkgd <- mutate(bkgd, 
               total_genes_in_bg = total_genes, 
               not_bound_in_bg = total_genes_in_bg - bound_in_bg)

droid_summary <- left_join(droid_summary, 
          bkgd,
          by = c("TF_SYMBOL")) %>% 
  ungroup()

summary_stats <- droid_summary %>% 
  filter(!is.na(TF_SYMBOL)) %>% 
  group_by(TF_SYMBOL, cluster) %>% 
  do(broom::tidy(phyper(.$bound_genes_in_cluster - 1, .$bound_in_bg, 
                        .$not_bound_in_bg, .$total_genes_in_cluster, lower.tail = FALSE))) %>% 
  ungroup() %>% 
  mutate(padj = p.adjust(x)) %>% 
  inner_join(droid_summary, ., by = c("TF_SYMBOL", "cluster")) %>% 
  mutate(enrichment = log2((bound_genes_in_cluster / total_genes_in_cluster) /
                      (bound_in_bg / total_genes_in_bg))) %>% 
  select(everything(), pval = x)


tf_summary_matrix <- summary_stats %>% 
  select(TF_SYMBOL, cluster, padj) %>% 
  spread(TF_SYMBOL, padj, fill = 1.0) %>% 
  as.data.frame() %>% 
  column_to_rownames("cluster") %>% 
  as.matrix() 

tf_summary_matrix <- -log10(tf_summary_matrix)
tf_summary_matrix <- t(tf_summary_matrix)
tf_summary_matrix <- tf_summary_matrix[rowSums(tf_summary_matrix) != 0, ]
tf_summary_matrix <- tf_summary_matrix[order(tf_summary_matrix[, "4"], decreasing = T),]

ht <- Heatmap(tf_summary_matrix,
        cluster_rows = F,
        name = "-log10\nFDR")

pdf("tf_enrichment.pdf", height = 12)
print(ht)
dev.off()
```


```{r pre-mzt_genes}
## based on microarray (Unmasking activation of the zygotic genome using chromosomal deletions in the Drosophila embryo
#S De Renzis O Elemento S Tavazoie EF Wieschaus (2007))

renzis <- read_delim(file.path(docs_dir, "renzis_early_zygotic.txt"), 
           delim = " ", skip = 1, col_names = F)

mat_clusters <- read_tsv("../2018-04-18_drosophila/fly_mat_km_clusters.tsv")
pre_clusters <- read_tsv("../2018-04-18_drosophila/fly_pre_km_clusters.tsv")
classifications <- read_tsv("../2018-04-18_drosophila/eisen_vs_intron_classification.tsv")

all_clusters <- full_join(mat_clusters, pre_clusters, by = "gene_id", suffix = c("_mature", "_pre")) %>% 
  left_join(., classifications, by = "gene_id")

all_clusters <- left_join(all_clusters, renzis, by = c("gene_id" = "X2"))

all_clusters
```

```{r polii}
if (!file.exists("supplemental_file_1.xlsx")){
  download.file("http://datadryad.org/bitstream/handle/10255/dryad.52325/supplemental_file_1.xlsx", "supplemental_file_1.xlsx")
}
pol2_chipseq_description <- readxl::read_excel("supplemental_file_1.xlsx")

pol2_chipseq <- readxl::read_excel("supplemental_file_1.xlsx", sheet = 2)


pol2_clusters <- inner_join(all_clusters, pol2_chipseq, by = c("gene_id" = "fb_gene_id")) %>% 
  filter(!is.na(preMBT_group) | !is.na(MBT_group))

pol2_clusters %>% filter(cluster_pre == 4, is.na(X1)) -> clust4_not_known_binding_summary

table(clust4_not_known_binding_summary$MBT_group)
table(clust4_not_known_binding_summary$cluster_mature)
```


```{r overlap}

renzis_premBT <- renzis$X2
pol2_chipseq_premBT <- filter(pol2_chipseq, !is.na(preMBT_group) ) %>% pull(fb_gene_id)
clust4 <- all_clusters %>% filter(cluster_pre == 4) %>% pull(gene_id)

preMBT_genes <- list("microarray\nearly zygotic" = renzis_premBT,
                     "pol II Chip\nearly zygotic" = pol2_chipseq_premBT,
                     "pre-mRNA\nearly zygotic" = clust4)

library(eulerr)
set.seed(1)
overlap_euler <- euler(preMBT_genes)
plot(overlap_euler, quantities = T, auto.key = T, fill = brewer.pal(3, "Set1"),
      counts = list(cex = 1.5, col = "black")) 

pdf("microarry_chipseq_overlap.pdf")
plot(overlap_euler, quantities = T, auto.key = T, fill = brewer.pal(3, "Set1"),
      counts = list(cex = 2, col = "black"))
dev.off()
```