---
title: "Build human intron exon references"
author: "Kent Riemondy RBI"
date: "8/2/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r libs}
library(valr)
library(tidyverse)
library(rtracklayer)
library(doParallel)
source("../../R/globals.r")
```


```{r annot}
gtf <- "~/Projects/shared_dbases/annotation/gencode.v26.primary_assembly.annotation.gtf"
gtf <- import(gtf)
gtf <- as.data.frame(gtf) %>% tbl_df()
```

## Compute coverage from pre-ZGA samples ()

To define introns, all intronic mRNA regions with coverage in the pre-ZGA samples (1-cell to be restrictive) will be masked and excluded from the intronic set used for counting. There are 5 samples from the 1-cell stage that will be summed and used to define the masking annotations for introns.

```{r coverage_tracks}
mdata <- read_tsv(file.path(data_dir, "raw_data", "human", 
                             "XUE", "PRJNA189204.txt"))

# select relevant libs
filtered_mdata <- dplyr::filter(mdata, 
                                str_detect(sample_title, 
                                           "human oocyte|human pronuclei"))


bws <- dir(file.path(data_dir, "bigwigs", "human", "XUE"), 
           full.names = TRUE, 
           pattern = ".bw")

bws <- str_subset(bws, str_c(filtered_mdata$run_accession, 
                             collapse = "|"))

bws <- str_subset(bws, "rev.bw|fwd.bw")
# FYI these libraries are unstranded 
bw_df <- map_dfr(bws,
                  ~read_bigwig(.x, set_strand = "+")) 

# sum coverage
bw_df <- bed_partition(bw_df, 
                        score = sum(score))

ivl_df <- bind_rows(list(
  "-" = fwd_df,
  "+" = rev_df),
  .id = "strand")

ivl_df <- dplyr::select(ivl_df, chrom:score, strand)
```

Next exclude exonic regions and keep regions with at least 2 reads. These regions will be excluded from intron intervals for computing intron coverage. Also add repeatMasker regions. (downloaded from ucsc table browser 2018-07-09)

```{r drop_exons}

gtf_df <- mutate(gtf,
                 start = start - 1,
                 chrom = as.character(seqnames),
                 strand = as.character(strand)) %>% 
  bed_sort()

exons <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                transcript_id,
                exon_number,
                strand) %>% 
  bed_sort() %>% 
  group_by(strand)

ivl_df <- group_by(ivl_df, strand)
intron_coverage <- bed_subtract(ivl_df, exons)

masked_ivls <- filter(intron_coverage, score > 1) %>% 
  group_by(strand) %>% 
  bed_merge() %>% 
  ungroup()

#rpts <- read_bed(file.path(db_dir, "drosophila", 
#                           "drosophila_rmsk.bed.gz"), n_fields = 6)

#masked_ivls <- bind_rows(rpts, masked_ivls)
#masked_ivls <- dplyr::select(masked_ivls,
#                             chrom:end, strand)

introns <- filter(gtf_df, type == "transcript") %>% 
  select(chrom, start, end, gene_name, strand) %>% 
  unique() %>% 
  group_by(strand) %>% 
  bed_subtract(exons) %>% 
  bed_subtract(masked_ivls) %>% 
  unique()

introns_out <- introns %>% 
  mutate(score = 0L) %>% 
  select(chrom, start, end, gene_name, score, strand) %>% 
  mutate_if(is.numeric, format, scientific = F) %>% 
  mutate_all(str_trim)

write_tsv(introns_out, "fly_introns.bed", col_names = F)
introns_out %>% 
  dplyr::select(GeneID = gene_name, Chr = chrom, Start = start, End = end, Strand = strand) %>% 
  write_tsv(., "fly_introns.saf")

exons_out <- filter(gtf_df, type == "exon") %>%  
  dplyr::select(chrom,
                start, 
                end, 
                gene_name,
                exon_number,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(exons_out, "fly_exons.bed", col_names = F)

exons_out %>% 
  dplyr::select(GeneID = gene_name, Chr = chrom, Start = start, End = end, Strand = strand) %>% 
  write_tsv(., "fly_exons.saf")

```