---
title: "Sea Urchin time course"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Sea Urchin MZT analysis

 
## RNA-Seq Libraries
```{r load_libs}
source("../../R/globals.r")
```



```{r get_metadata}
mdata1 <- read_csv(file.path(data_dir, "raw_data", "seaurchin", 
                             "TU", "sraresult.csv"))

mdata2 <- read_tsv(file.path(data_dir, "raw_data", "seaurchin", 
                             "TU", "PRJNA81157.txt"))

mdata <- left_join(mdata2, mdata1, 
                   by = c("experiment_accession" = "Experiment Accession"))
                   
fish_pond <- file.path(data_dir, 
                           "salmon",
                           "seaurchin",
                           "TU")
fishes <- dir(fish_pond,
                 pattern = "quant.sf$",
                 recursive = T,
                 full.names = T)

fish_names <- str_split(fishes, "\\/") %>% map_chr(., ~.x[10])

mdata <- mdata %>% filter(run_accession %in% fish_names)
mdata
```


## Alignment %

```{r alignment}

alignment <- map(fish_names, 
                  ~fromJSON(txt = file.path(
                    fish_pond, 
                          .x, 
                          "aux_info", 
                          "meta_info.json")) %>% .$percent_mapped)

alignment <- data_frame(fish_names,
                        percent_mapped = unlist(alignment))


inner_join(alignment, 
          mdata, 
          by = c("fish_names" = "run_accession"))
```

## Load in salmon data

```{r load_dat, message = F}

dat <- map(fishes, read_tsv)

# name with libraryID
names(dat) <- fish_names

dat <- bind_rows(dat, .id = "run_accession")

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("run_accession"))
```

### transcript to gene mapping
```{r drop_monoexonic}
gtf <- file.path(db_dir, "seaurchin", "GCF_000002235.4_Spur_4.2_genomic.gtf")
gtf <- import(gtf)
gtf <- as.data.frame(gtf)

gtf_summary <- dplyr::filter(gtf, type == "exon") %>% 
  dplyr::select(seqnames, start, end, strand, gene_id) %>% 
  dplyr::rename(chrom = seqnames) %>% 
  unique() %>% 
  group_by(gene_id) %>% 
  bed_merge() %>% 
  dplyr::summarize(total_exons = n())

monoexonic_genes <- dplyr::filter(gtf_summary, total_exons == 1) %>% pull(gene_id)
```


```{r remove_monoexonic}

dat <- anti_join(dat, 
                    data_frame(gene_id = monoexonic_genes),
                 by = c("Name" = "gene_id"))

# classify transcripts as primary or mature
dat %>% 
  mutate(type = ifelse(str_detect(Name, 
                                  "^pre_"), 
                       "pre", "mature")) -> dat

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length:EffectiveLength, 
                        NumReads))

gene2tx <- gtf[, c("gene_id", "transcript_id")]
gene2tx <- unique(gene2tx)
gene2tx <- na.omit(gene2tx)

# add in geneid attribute
dat <- left_join(dat, gene2tx, by = c("Name" = "transcript_id")) %>% 
  dplyr::rename(transcript_id = Name)

# add geneid attribute for pre_ annotations

dat <- mutate(dat, 
              gene_id = ifelse(type == "pre",
                               str_replace(transcript_id, "^pre_", ""),
                               gene_id))

dat <- mutate(dat,
              monoexonic = gene_id %in% monoexonic_genes)

dat <- separate(dat,
                `Experiment Title`, 
                into = c("organism", "misc", "stageName", "libtype"), 
                sep = ";")
```

### plot pre_mRNA ratio

```{r, calc_premRNA}
# sum primary and mature TPMs for each gene 
summary_dat <- dplyr::select(dat, TPM, gene_id, 
              run_accession, type, stageName) %>% 
  group_by(gene_id, run_accession, type) %>% 
  dplyr::summarize(total_tpm = sum(TPM, na.rm = F),
                   stageName = unique(stageName)) %>% 
  ungroup() 

## calc primary ratio
tpm_summary <- summary_dat %>% 
  spread(type, total_tpm) %>% 
  mutate(primary_ratio  = pre / (mature + pre))

## calc mean primary ratio
tpm_means <- tpm_summary %>% 
  group_by(gene_id, 
           stageName) %>% 
  summarize(mean_primary = mean(primary_ratio,
                                na.rm = T)) 

tpm_means <- dplyr::ungroup(tpm_means)

plt <- ggplot(tpm_means, 
       aes(stageName, mean_primary)) +
  geom_boxplot(aes(fill = stageName), coef = 1e3) + 
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_brewer(palette = "Set1",
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5)
  )

plt
```




## Classify transcripts into maternal or zygotic

Next I'll try to classify transcripts into catagories based upon their abundance changes throughout embryogenesis. 

First generate salmon quant.sf files for pre-mRNAs or mature mRNAs independently for loading into `tximport`
```{r, split_salmon}

tx2gene_mapping <- dplyr::select(dat, 
                         transcript_id, 
                         gene_id) %>% 
  unique() %>% 
  as.data.frame()

all_dat <- map(fishes, read_tsv)

all_pre <- map(all_dat, 
                 ~dplyr::filter(.x,
                                str_detect(Name, "^pre_")))

all_mat <- map(all_dat, 
               ~dplyr::filter(.x,
                              !str_detect(Name, "^pre_")))

dir.create("salmon_modified", recursive = T)

walk2(all_pre,
      fish_names,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".premRNA"))))
walk2(all_mat,
      fish_names,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".maturemRNA"))))
```


```{r}
files <- dir("salmon_modified", full.names = T)

mature <- str_subset(files, "maturemRNA$")
pre <- str_subset(files, "premRNA$")

matures <- tximport(mature, 
         type = "salmon",
         tx2gene = tx2gene_mapping)

pres <- tximport(pre, 
         type = "salmon",
         tx2gene = tx2gene_mapping)

# make metadata sample ordered by columns in count matrix
pdata <- inner_join(data_frame(run_accession = fish_names),
           mdata, by = "run_accession")

pdata <- separate(pdata,
                `Experiment Title`, 
                into = c("organism", "misc", "stageName", "libtype"), 
                sep = ";")

pdata <- mutate(pdata,
                mzt = ifelse(stageName %in% c(" embryo 10hpf", " embryo 0hpf"),
                             "maternal",
                             "zygotic"))
```


```{r deseq2}

dds <- DESeqDataSetFromTximport(matures, pdata, ~mzt)
dds_mature <- DESeq(dds)
res_mature <- results(dds_mature)
res_mature_tidy <- results(dds_mature, tidy = TRUE)

dds <- DESeqDataSetFromTximport(pres, pdata, ~mzt)
dds_pre <- DESeq(dds)
res_pre <- results(dds_pre)
res_pre_tidy <- results(dds_pre, tidy = TRUE)
```

```{r plot_MAs}
pre_plot_dat <- res_pre_tidy %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0, "Down", "Not Sig")))
n_sig_up <- nrow(dplyr::filter(pre_plot_dat, sig == "Up")) %>% 
  formatC(big.mark = ",")
n_sig_down <- nrow(dplyr::filter(pre_plot_dat, sig == "Down"))  %>% 
  formatC(big.mark = ",")
n_notsig <- nrow(dplyr::filter(pre_plot_dat, sig == "Not Sig"))  %>% 
  formatC(big.mark = ",")

p <- ggplot(pre_plot_dat, 
       aes(baseMean, log2FoldChange)) +
  geom_point(aes(colour = sig), size = 0.1) +
  scale_colour_viridis(discrete = TRUE,
                       name = "",
                       labels = c(paste0("Sig down (n = ", n_sig_down, ")"),
                                  paste0("Not Sig (n= ", n_notsig, ")"),
                                  paste0("Sig up (n = ", n_sig_up, ")"))) +
  scale_x_log10() + 
  labs(x = expression(paste("log"[10], " Average Expression")),
       y = expression(paste("log"[2], " ", frac("Zygotic", "Maternal"))),
       title = "pre-mRNA") +
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(legend.position = "top")
  
p

mature_plot_dat <- res_mature_tidy %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0, "Down", "Not Sig")))
n_sig_up <- nrow(dplyr::filter(mature_plot_dat, sig == "Up")) %>% 
  formatC(big.mark = ",")
n_sig_down <- nrow(dplyr::filter(mature_plot_dat, sig == "Down"))  %>% 
  formatC(big.mark = ",")
n_notsig <- nrow(dplyr::filter(mature_plot_dat, sig == "Not Sig"))  %>% 
  formatC(big.mark = ",")

p <- ggplot(mature_plot_dat, 
       aes(baseMean, log2FoldChange)) +
  geom_point(aes(colour = sig), size = 0.1) +
  scale_colour_viridis(discrete = TRUE,
                       name = "",
                       labels = c(paste0("Sig down (n = ", n_sig_down, ")"),
                                  paste0("Not Sig (n= ", n_notsig, ")"),
                                  paste0("Sig up (n = ", n_sig_up, ")"))) +
  scale_x_log10() + 
  labs(x = expression(paste("log"[10], " Average Expression")),
       y = expression(paste("log"[2], " ", frac("Zygotic", "Maternal"))),
       title = "mature-mRNA") +
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(legend.position = "top")
  
p
```


```{r plot_zyg_mat_pre}
library(eulerr)

sig_mat_up <- res_mature_tidy %>% filter(padj < 0.05, log2FoldChange > 0)
sig_pre_up <- res_pre_tidy %>% filter(padj < 0.05, log2FoldChange > 0)
all_sig_up <- union(sig_mat_up$row, sig_pre_up$row)
lgl_up_df <- data.frame(
  `mature_mRNA` =  all_sig_up %in% sig_mat_up$row,
  `pre_mRNA` =  all_sig_up %in% sig_pre_up$row
)

sig_mat_down <- res_mature_tidy %>% filter(padj < 0.05, log2FoldChange < 0)
sig_pre_down <- res_pre_tidy %>% filter(padj < 0.05, log2FoldChange < 0)
all_sig_down <- union(sig_mat_down$row, sig_pre_down$row)
lgl_down_df <- data.frame(
  `mature_mRNA` =  all_sig_down %in% sig_mat_down$row,
  `pre_mRNA` =  all_sig_down %in% sig_pre_down$row
)

fit_up <- euler(lgl_up_df)
fit_down <- euler(lgl_down_df)

plot(fit_up, fill = viridis(2), quantities = TRUE, main = "Increased abundance in Zygotic Stages")
plot(fit_down, fill = viridis(2), quantities = TRUE, main = "Decreased abundance in Zygotic Stages")
```


```{r write_out_sig}
write_tsv(res_pre_tidy, "deseq_results_pre.tsv")
write_tsv(res_mature_tidy, "deseq_results_mature.tsv")

```


```{r kmeans}
library(flexclust) #kmeans++

mature_counts_lrt <- log2(counts(dds_mature, normalized =T) + 1)
pre_counts_lrt <- log2(counts(dds_pre, normalized = T) + 1)

colnames(mature_counts_lrt) <- colData(dds_mature)$stageName
colnames(pre_counts_lrt) <- colData(dds_pre)$stageName

mature_zscores <- t(scale(t(mature_counts_lrt)))
pre_zscores <- t(scale(t(pre_counts_lrt)))

sig_pre <- res_pre_tidy[res_pre_tidy$padj < 0.01, ]
sig_mat <- res_mature_tidy[res_mature_tidy$padj < 0.01, ]

sig_pre_zscores <- pre_zscores[rownames(pre_zscores) %in% sig_pre$row, ]
sig_mature_zscores <- mature_zscores[rownames(mature_zscores) %in% sig_mat$row, ] 

set.seed(20180408)
pre_km <- flexclust::kcca(sig_pre_zscores, k = 10, 
                        control = list(initcent = "kmeanspp"))
set.seed(20180408)
mat_km <- flexclust::kcca(sig_mature_zscores, k = 10, 
                        control = list(initcent = "kmeanspp"))


```


```{r hmaps}
col_order <- sort(pdata$stageName)

ht1 <- Heatmap(sig_pre_zscores,
         name = "pre-mRNA abundance\nzscore",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = F,
         split = pre_km@cluster,
         km_title = "%i", 
         column_order = col_order)

ht1

pdf("premRNA_kmeans.pdf", width = 8)
print(ht1)
dev.off()

common_genes <- intersect(rownames(sig_mature_zscores), rownames(sig_pre_zscores))
ht1_1 <- Heatmap(sig_pre_zscores[common_genes, ],
         name = "pre-mRNA abundance\nzscore",
         show_row_names = F,
         show_row_dend = F,
         split = pre_km@cluster[common_genes],
         km_title = "%i", 
         cluster_columns = F,
         column_order = col_order)

ht1_2 <- Heatmap(sig_mature_zscores[common_genes, ],
         name = "mature-mRNA abundance\nzscore",
         show_row_names = F,
         show_row_dend = F,
         cluster_columns = F,
         column_order = column_order(ht1))

pdf("pre_mRNA_kmeans_with_mat.pdf", width = 8)
print(ht1_1 + ht1_2)
dev.off()


ht2 <- Heatmap(sig_mature_zscores, 
         name = "mature-mRNA abundance\nzscore",
         show_row_names = F,
         show_row_dend = F,
         split = mat_km@cluster,
         km_title = "%i",
         cluster_columns = F,
         column_order = col_order)
ht2

pdf("mat_mRNA_kmeans.pdf", width = 8)
print(ht2)
dev.off()

```


```{r gene_names}
tx2gene_name <- gtf %>% 
  filter(type == "gene") %>% 
  select(gene_name, gene_id) %>% 
  unique()

pre_clusters <- data_frame(gene_id = names(pre_km@cluster),
                           cluster = pre_km@cluster)
mat_clusters <- data_frame(gene_id = names(mat_km@cluster),
                           cluster = mat_km@cluster)

pre_clusters <- left_join(pre_clusters, tx2gene_name, by = "gene_id")
mat_clusters <- left_join(mat_clusters, tx2gene_name, by = "gene_id")
```