---
title: "Convert genome coordinates to transcript coords"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

##

```{r load_libs}
source("../../R/globals.r")
library(GenomicFeatures)
```


```{r read_bed}
gtf_to_bed <- function(gtf){
  gtf_df <- dplyr::rename(gtf,
                          chrom = seqnames) %>% 
    mutate(start = start - 1)
  gtf_df
}
  
gtf_fn <- "~/Projects/mzt-introns/dbases/drosophila/primary_transcripts.gtf"
gtf <- import(gtf_fn)
gtf <- as.data.frame(gtf) %>% tbl_df()

gtf_df <- gtf_to_bed(gtf) %>% 
  group_by(strand)

bed_df <- read_bed("../2018-05-10/fly_intron_mask.bed", n = 6)


bed <- makeGRangesFromDataFrame(bed_df)
gtf <- import(gtf_fn)
ggtf <- makeTxDbFromGRanges(gtf)
gtx <- exonsBy(ggtf, by = "tx", use.names = T)
db_out <- GenomicFeatures::mapToTranscripts(x = bed, transcripts = gtx)
db_out <- as_data_frame(db_out)

db_out <- mutate(db_out,
                     start = start - 1,
                     end = end,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom = seqnames,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(db_out, "fly_intron_mask_txcoords.bed", col_names = F)


bed_df <- read_bed("../2018-05-10/fly_exon_intron_plus_exons_juncts_mask.bed", n = 6)


bed <- makeGRangesFromDataFrame(bed_df)
gtf <- import(gtf_fn)
ggtf <- makeTxDbFromGRanges(gtf)
gtx <- exonsBy(ggtf, by = "tx", use.names = T)
db_out <- GenomicFeatures::mapToTranscripts(x = bed, transcripts = gtx)
db_out <- as_data_frame(db_out)

db_out <- mutate(db_out,
                     start = start - 1,
                     end = end,
                     score = 0,
                     name = ".") %>% 
  ungroup() %>% 
  dplyr::select(chrom = seqnames,
                start, 
                end, 
                name,
                score,
                strand) %>% 
  bed_sort() %>% 
  mutate_if(is.numeric, format, scientific = F)  %>% 
  mutate_all(str_trim)

write_tsv(db_out, "fly_intron_mask_plus_exon_txcoords.bed", col_names = F)
```
