---
title: "MZT differential expression vignette"
date: '`r Sys.Date()`'
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```


```{r}
library(readr)
library(here)
library(dplyr)
library(stringr)
library(ggplot2)
library(tidyr)
library(tibble)
library(rtracklayer)
library(tximport)
library(DESeq2)
```


This vignette will demonstrate how to use the salmon output files to classify trasncripts into maternal, maternal and zygotic, or zygotic. 

For this vignette the testdata set comprised of 5 timepoints from drosophila development will be examined. These test files were downsampled to reduce the file size and only mapped to the `2L` chromosome. 

```{r}
salmon_dir <- "../data-test/salmon_bt2_masked/eisa_masked/"

sr_ids <- c("SRR5469986.sub",
            "SRR5469988.sub",
            "SRR5469990.sub",
            "SRR5469992.sub",
            "SRR5469994.sub",
            "SRR5469998.sub",
            "SRR5470000.sub",
            "SRR5470002.sub",
            "SRR5470004.sub",
            "SRR5470006.sub")

stage <- c("0-1 hr embryo",
            "1-2 hr embryo",
            "2-3 hr embryo",
            "3-4 hr embryo",
            "4-5 hr embryo",
            "0-1 hr embryo",
            "1-2 hr embryo",
            "2-3 hr embryo",
            "3-4 hr embryo",
            "4-5 hr embryo")

strain <- rep(c("w1118", "eGFP-ME31B"), each = 5)

mdata <- data.frame(sr_ids, stage, strain)
```


```{r }
gtf_fn <- "../dbases-test/drosophila/ext/Drosophila_melanogaster.BDGP6.84.2L.gtf"
gtf <- import(gtf_fn) %>% 
  as.data.frame() %>% 
  dplyr::rename(chrom = seqnames) %>% 
  mutate(start = start - 1L)

gene2symbol <- gtf %>%
  filter(type == "gene") %>% 
  select(gene_id, gene_name) %>% 
  distinct()

gene2tx <- gtf %>%
   filter(type == "transcript") %>%
   select(gene_id, transcript_id) %>%
   na.omit() %>%
   distinct()

gene2intron <- read_tsv("../dbases-test/eisa/tx2gene.tsv") 
```


## Examine intron signal across timepoints 

As a quick QC check, the proportion of reads assigned to introns will be examined across the developmental timepoints. The expectation is that intron signal will be minimal at the 0-1 hour time point, due to the masking of any intron signal at this stage, and that intron signal will increase coincident with the induction of zygotic transcription. 

```{r}
salmon_files <- file.path(salmon_dir,
                   mdata$sr_ids,
                   "quant.sf")

txi <- tximport(salmon_files, 
                type = "salmon",
                tx2gene = gene2intron)

tpms <- txi$abundance %>% 
  as.data.frame() 
colnames(tpms) <- mdata$sr_ids
tpms <- rownames_to_column(tpms, "feature_id") %>% 
  mutate(count_type = ifelse(str_detect(feature_id, "-I[0-9]*$"),
                             "intron",
                             "exon"),
         gene_id = str_remove(feature_id, "-I[0-9]*$")) %>% 
  pivot_longer(cols = -c(count_type, gene_id, feature_id),
               names_to = "run",
               values_to = "TPM")

tpms <- left_join(tpms, mdata, by = c('run' = 'sr_ids'))
```

```{r}
tpm_summary <- tpms %>%
  select(-feature_id) %>%
  pivot_wider(names_from = count_type,
              values_from = TPM) %>%
  na.omit() %>%
  mutate(intron_ratio  = intron / (intron + exon))

## calc mean primary ratio across developmental stages
tpm_means <- tpm_summary %>% 
  group_by(gene_id, 
           stage) %>% 
  summarize(mean_primary = mean(intron_ratio, na.rm = T)) 

dev_stages_ordered <- c(
  "0-1 hr embryo",
  "1-2 hr embryo",
  "2-3 hr embryo",
  "3-4 hr embryo",
  "4-5 hr embryo")

tpm_means <- mutate(tpm_means, 
                       stage = factor(stage, 
                                      levels = dev_stages_ordered)
                       )

p <- ggplot(tpm_means, 
       aes(stage, mean_primary)) +
  geom_boxplot(outlier.shape = NA, coef = 1e10) + 
  labs(x = "",
       y = expression(frac("Intron TPM", 
                           "Intron + Exon TPM"))) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5)
  )

p
```


## Differential expression testing

Next DESeq2 will be used to compare the expression of spliced transcripts and introns in the post-mzt samples to the pre-mzt. For this analysis intron and splice transcript signals will be aggregated to gene-level estimates. Additionally the samples will be binned into two stages for comparison: maternal representing the pre-major ZGA wave stages, and zygotic, represent samples at or after the major wave. Datasets with additional time points and replicates would benefit from using a timecourse or ANOVA-like approach that would treat each timepoint independently rather than a two-state comparison. 

```{r}
mdata$mzt <- ifelse(mdata$stage %in% c("0-1 hr embryo", "1-2 hr embryo"),
                    "maternal",
                    "zygotic")
ddobj <- DESeqDataSetFromTximport(txi, 
                                  mdata, 
                                  ~mzt + strain)
ddobj <- DESeq(ddobj)

res <- results(ddobj, 
                   name = "mzt_zygotic_vs_maternal", 
                   tidy = TRUE) %>% 
  as_tibble() %>% 
  mutate(count_type = ifelse(str_detect(row, "-I[0-9]*$"), 
                             "intron",
                             "exon"),
         sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0,
                             "Down", "Not Sig"))) %>% 
  filter(!is.na(sig))

out_tbl <- res %>% 
   mutate(gene_id = str_remove(row, "-I[0-9]*$")) %>% 
   left_join(gene2symbol, by = "gene_id")

out_tbl
```

```{r}
ggplot(out_tbl, aes(baseMean, log2FoldChange)) +
  geom_point(aes(color = count_type)) +
  scale_x_log10(labels = scales::comma) +
  scale_color_manual(values = c("grey", "blue"))
```


## Select catagories

Using the outputs from DESeq2 we can now classify genes to maternal, maternal + zygotic, or zygotic.  

```{r}
fold_change_cutoff <- 1
padj_cutoff <- 0.05
tpm_cut_off <- 1

mean_tpms <- tpms %>% 
  mutate(gene_id = feature_id) %>% 
  group_by(gene_id, stage) %>% 
  summarize(mean_tpm = mean(TPM)) %>% 
  group_by(gene_id) %>% 
  mutate(max_tpm = max(mean_tpm)) %>% 
  pivot_wider(names_from = stage, values_from = mean_tpm) %>% 
  inner_join(res, by = c("gene_id" = "row"))

exonic_df <- mean_tpms %>% 
  filter(!str_detect(gene_id, "-I$")) 

intronic_df <- mean_tpms %>% 
  filter(str_detect(gene_id, "-I$")) 

maternal_genes <- exonic_df %>% 
    filter(`0-1 hr embryo` >= tpm_cut_off) %>% 
    pull(gene_id) %>% 
    unique()
  
zygotic_genes_exon <- exonic_df %>% 
    filter(padj < padj_cutoff, 
           log2FoldChange > fold_change_cutoff,
           max_tpm > tpm_cut_off) %>% 
    arrange(padj) %>% 
    pull(gene_id) %>% 
    unique()
  
zygotic_genes_intron <- intronic_df %>% 
    filter(padj < padj_cutoff, 
           log2FoldChange > fold_change_cutoff,
           max_tpm > tpm_cut_off) %>% 
    arrange(padj) %>%
    pull(gene_id) %>% 
    unique()
  
all_zygotic_combined <- unique(union(zygotic_genes_exon,
                                       str_remove(zygotic_genes_intron, 
                                                  "-I[0-9]*$") ))
  
pure_zygotic_combined <- setdiff(all_zygotic_combined, maternal_genes)
  
mat_zyg_combined<- intersect(all_zygotic_combined, maternal_genes)
  
pure_maternal <- setdiff(setdiff(maternal_genes, 
                                 mat_zyg_combined),
                           all_zygotic_combined)
  
  
gene_lists <- lst(
  all_zygotic_combined,
  pure_zygotic_combined,
  mat_zyg_combined,
  pure_maternal
)
  
gene_df <- lapply(seq_along(gene_lists), function(x) {
  data.frame(gene_id = gene_lists[[x]],
             mzt_class = names(gene_lists)[x])}) %>% 
  do.call(rbind, .)  %>% 
    left_join(gene2symbol, by = "gene_id")
  
table(gene_df$mzt_class)
```



## Session info

```{r}
sessionInfo()
```

