---
title: "Examine intronic signals during MZT in Drosophila embryos"
author: "Kent Riemondy RBI"
date: "10/17/2017"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Background

The Maternal to Zygotic Transistion (MZT) is a developmental stage in which the embryo initiates transcription of mRNAs from the zygotic genome. Maternally deposited mRNAs begin to be degraded and are replaced by zygotic messages. 

Zygotic mRNAs can be distinguished from maternal mRNAs, due to the presence or higher proportions of introns. Maternally derived mRNAs are likely to be spliced, and any remaining intron lariat degraded prior to fertilization. In contrast zygotic transcription will produce mRNAs in that need to be spliced, and thus there will be some proportion of unspliced mRNA, and intron lariat derivied from the zygotic transcripts. 

Using ribo-zero depleted total-RNA-Seq libraries, one would expect to be able to recover intronic mRNA fragements, and be able to assign for each transcript, the relative proportion of each transcript that is maternally or zygotically derived. 


## Approach

Each total RNA-Seq lib was trimmed to remove Tru-Seq adapters (Rissland libraries), or barcodes and custom 3' adapter (Bartel libraries). The trimmed reads were then processed with `salmon` to estimate `TPMs` for each transcript. The `salmon` transcript database was modified to include a pre-mRNA entry, which was designated as an exon that spanned the entire gene. This pre-mRNA annotation matches the `gene` entry from the annotaiton file. By including the pre-mRNA transcript in the `salmon` database, one can get an estimate of the fraction of pre-mRNA or mature mRNA present in each sample. 

First we'll examine the extent to which pre-mRNA transcripts increase during embryonic development. Then, if a clear increase is present, we can classify transcripts into a % maternal and % zygotic, and additional classify the temporal pattern of zygotic transcription i.e. are there distinct sets of early versus late transcripts, and do their transcriptional pattern suggest function (TF transcription prior to target activation/repression, miRNA transcription before target repression, RBP transcription prior to translational control, etc.)

 
## RNA-Seq Libraries
```{r load_libs}
library(tidyverse)
library(valr)
library(ComplexHeatmap)
library(cowplot)
library(viridis)
```



```{r get_metadata}
mdata1 <- read_tsv("../data/raw_data/GSE83616_run_info_geo.txt")
mdata2 <- read_tsv("../data/raw_data/GSE98106_run_info_geo.txt")


#simplify mdata
mdata1 <- dplyr::select(mdata1,
                       Run_s, 
                       developmental_stage_s,
                       genotype_s,
                       source_name_s,
                       Assay_Type_s)

#simplify mdata
mdata2 <- dplyr::select(mdata2,
                       Run_s, 
                       developmental_stage_s,
                       rip_antibody_s,
                       source_name_s,
                       strain_s)

# select relevant libs
mdata2 <- dplyr::filter(mdata2,
                     source_name_s == "Embryo",
                     strain_s %in% c("w1118", "eGFP-ME31B"),
                     rip_antibody_s == "none") %>% 
  dplyr::select(-rip_antibody_s)

mdata1 <- dplyr::filter(mdata1,
                     genotype_s == "wt",
                     Assay_Type_s == "RNA-Seq") %>% 
  dplyr::select(-Assay_Type_s) %>% 
  dplyr::rename(strain_s = genotype_s)


mdata <- list(mdata1, mdata2)
names(mdata) <- c("eichhorn", "rissland")

mdata <- bind_rows(mdata, .id = "study")
#cleanup names
mdata <- mdata %>% 
  mutate(source_name_s = ifelse(source_name_s == "embryos", 
                                "Embryo",
                                source_name_s),
         developmental_stage_s = str_replace(developmental_stage_s,
                                             " h ",
                                             " hr "))

# cleanup colnames
colnames(mdata) <- str_replace(colnames(mdata),
                               "_s$",
                               "")

mdata
```


## Alignment %

```{r alignment}
alignment <- read.table("../data/salmon/align_info.txt",
                        stringsAsFactors = F)
alignment <- alignment[, -c(2:3)]

colnames(alignment) <- c("library", "% of reads aligned")

alignment <- alignment %>% 
  mutate(library = dirname(library) %>% str_split("/", simplify = T) %>% .[, 1],
         `% of reads aligned` = str_replace(`% of reads aligned`, ",$", "") %>% as.numeric()) 

inner_join(alignment, 
          mdata, 
          by = c("library" = "Run"))
```

## Load in salmon data

```{r load_dat, message = F}
files <- dir("../data/salmon",
            recursive = T,
            pattern = "*quant.sf",
            full.names = T)

dat <- map(files, read_tsv)

# name with libraryID
names(dat) <- dirname(files) %>% 
  str_split(., "/", simplify = T) %>% 
  .[, 4]

dat <- bind_rows(dat, .id = "libraryID")
```


```{r join}

#join with main data
dat <- inner_join(dat, 
                 mdata,
                by = c("libraryID" = "Run"))
```

### transcript to gene mapping
```{r drop_monoexonic}
gtf <- read_tsv("../dbases/primary_transcript_db.gtf",
                comment = "#", col_names = F)

names(gtf) <- c("chrom",
                "source",
                "feature",
                "start",
                "end",
                "none",
                "strand",
                "frame",
                "attributes")

trx_gtf <- gtf %>% filter(feature == "transcript")

trx_attrs <- trx_gtf$attributes
gene_ids <- trx_attrs %>% 
  str_split(";") %>% 
  map(., ~str_subset(.x, "gene_id")) %>% 
  unlist() 

transcript_ids <- trx_attrs %>% 
  str_split(";") %>% 
  map(., ~str_subset(.x, "transcript_id")) %>% 
  unlist() 

trx_gtf$gene_id <- gene_ids
trx_gtf$transcript_id <- transcript_ids


trx_map <- trx_gtf %>% 
  dplyr::select(transcript_id, gene_id) %>% 
  mutate(gene_id = str_match(gene_id, 
                              'gene_id \\\"([^\\\\]+)\\\"') %>% 
  .[, 2],
         transcript_id = str_match(transcript_id, 
                                    'transcript_id \\\"([^\\\\]+)\\\"') %>% 
  .[, 2]) %>% 
  unique()
  
```

### monoexonic genes

```{r drop_monoexonic}
gtf <- read_tsv("~/Projects/shared_dbases/annotation/drosophila/Drosophila_melanogaster.BDGP6.84.gtf",
                comment = "#", col_names = F)
names(gtf) <- c("chrom",
                "source",
                "feature",
                "start",
                "end",
                "none",
                "strand",
                "frame",
                "attributes")

gtf_attrs <- gtf$attributes
gene_ids <- gtf_attrs %>% 
  str_split(";") %>% 
  map(., ~str_subset(.x, "gene_id")) %>% 
  unlist() 
gtf$gene_ids <- gene_ids

gtf_summary <- dplyr::filter(gtf, feature == "exon") %>% 
  dplyr::select(chrom, start, end, strand, gene_ids) %>% 
  unique() %>% 
  group_by(gene_ids) %>% 
  bed_merge() %>% 
  summarize(total_exons = n())

monoexonic_genes <- dplyr::filter(gtf_summary, 
                                  total_exons == 1) %>% 
  dplyr::pull(gene_ids) %>% 
  str_match(., 'gene_id \\\"([^\\\\]+)\\\"') %>% 
  .[, 2]
  
```

### add in gene ids

```{r}
dat <- left_join(dat,
                 trx_map,
                 by = c("Name" = "transcript_id"))
```


```{r remove_monoexonic}

dat <- anti_join(dat, 
                    data_frame(gene_id = monoexonic_genes),
                 by = "gene_id")

# classify transcripts as primary or mature
dat <- dat %>% 
  mutate(type = ifelse(str_detect(Name, 
                                  "^pre_"), 
                       "pre", "mature")) 

# drop unneeded data
dat <- dplyr::select(dat,
                     -c(Length, EffectiveLength, NumReads))
```

Only multiexonic messages will be used for calculating the proportion of pre-mRNA messages

### plot pre_mRNA ratio

```{r, calc_premRNA}
# sum primary and mature TPMs for each gene 
dplyr::select(dat, TPM, gene_id, 
              libraryID, type, developmental_stage,
              study, strain) %>% 
  group_by(gene_id, libraryID, 
           developmental_stage, type,
           strain) %>% 
  summarize(total_tpm = sum(TPM, na.rm = F)) %>% 
  ungroup() -> summary_dat

summary_dat %>% 
  group_by(gene_id, libraryID, 
           developmental_stage,
           strain) %>% 
  dplyr::arrange(type, .by_group = T) %>% 
  mutate(primary_ratio = nth(total_tpm, 2) / 
           (nth(total_tpm, 2) + nth(total_tpm, 1)))  -> tpm_summary


tpm_summary %>% 
  group_by(gene_id) %>% 
  mutate(max_tpm = mean(total_tpm)) %>% 
  dplyr::filter(max_tpm > 1) -> tpm_summary

tpm_summary %>% 
  dplyr::select(-type, -total_tpm) %>% 
  unique() -> tpm_filtered

tpm_filtered %>% 
  group_by(gene_id, 
           developmental_stage,
           strain) %>% 
  summarize(mean_primary = mean(primary_ratio,
                                na.rm = T)) -> tpm_filtered

tpm_filtered <- mutate(tpm_filtered, 
                       stage = factor(developmental_stage, 
                                         levels = c(
                                           "Stage 11 oocyte",
                                           "Stage 12 oocyte",
                                           "Stage 13 oocyte",
                                           "Stage 14 oocyte",
                                           "Activated egg",
                                           "0-1 hr embryo",
                                           "1-2 hr embryo",
                                           "2-3 hr embryo",
                                           "3-4 hr embryo",
                                           "4-5 hr embryo",
                                           "5-6 hr embryo")
                                         )
                       )

tpm_filtered <- mutate(tpm_filtered,
                       strain = ifelse(strain == "wt",
                                       paste(strain, 
                                             "(Eichhorn)"),
                                       paste(strain,
                                             "(Rissland)")))
                                        
p <- ggplot(tpm_filtered, 
       aes(stage, mean_primary)) +
  stat_boxplot(geom ='errorbar', 
               coef = 1e3) +
  geom_boxplot(aes(fill = stage, 
                   ymin=..lower.., 
                   ymax=..upper..),
               coef = 1e3) + 
  facet_grid(~strain, scale = "free") +
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_viridis(discrete = T,
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5),
    strip.text = element_text(size = 8)
  )

p
save_plot("premRNA_ratio.pdf", p, base_width = 6)

tpm_filtered %>% 
  group_by(gene_id, developmental_stage) %>% 
  summarize(mean_primary = mean(mean_primary, na.rm = T)) %>% 
  mutate(strain = "all strains") %>% 
  bind_rows(tpm_filtered, .) %>% 
  arrange(gene_id) -> all_combined

ggplot(all_combined, 
       aes(developmental_stage, mean_primary)) +
  geom_boxplot(aes(fill = developmental_stage), coef = 1e3) + 
  facet_grid(~strain, scale = "free") +
  labs(x = "",
       y = expression(frac("pre-mRNA TPM", 
                           "mature + pre-mRNA TPM"))) +
  scale_fill_brewer(palette = "Set1",
                    name = "Transcript Type",
                    breaks = c("pre", "mature"),
                    labels = c("pre-mRNA", "mature-mRNA")) +
  theme(
    axis.text.x = element_text(angle = 90,
                               vjust = 0.5,
                               hjust = 0.5)
  )
#ggsave("total_transcript_distribution.pdf")
```


```{r tpm_v_premRNA_ratio}
tpm_summary %>% 
  dplyr::select(gene_id, total_tpm, primary_ratio, libraryID) %>% 
  group_by(gene_id, libraryID) %>%
  mutate(gene_total_tpm = sum(total_tpm)) %>% 
  group_by(gene_id) %>% 
  summarize(mean_tpm = mean(gene_total_tpm),
            mean_pre_ratio = mean(primary_ratio, na.rm = T)) -> tpm_v_ratio

ggplot(tpm_v_ratio, aes(mean_tpm, mean_pre_ratio)) + 
  geom_point() +
  scale_x_log10()

```

## Heatmaps
```{r find_increasing}
tpm_filtered %>% 
  ungroup() %>% 
  split(., .$strain) -> per_strain_tpms

set_rowids <- function(df, colname = "gene_id"){
  rownames(df) <- df[[colname]]
  df[, 1] <- NULL
  df
}

map(per_strain_tpms, ~.x %>% 
          dplyr::select(-stage, -strain) %>% 
  spread(developmental_stage, mean_primary) %>% 
    as.data.frame() %>%
    set_rowids(.)) -> mats

#compute z-scores

map(mats, 
    ~t(scale(t(.x)))) -> zmats

map2(zmats,
     names(zmats),
     ~Heatmap(na.omit(.x), 
              name = "pre-mRNA proportion\n (Z-Score)",
                    show_row_names = F, 
                    show_row_dend = F,
              column_title = .y,
                )) -> plts

plts

all_mat <- do.call(cbind, mats)
all_zmat <- t(scale(t(all_mat)))

Heatmap(na.omit(all_zmat), name = "pre-mRNA proportion\n (Z-Score)",
        column_title = "All datasets",
                    show_row_names = F, 
                    show_row_dend = F) -> all_plt

all_plt
```


```{r}
tpm_summary %>% 
  dplyr::select(gene_id, developmental_stage, strain, total_tpm, type) %>% 
  split(., .$type) -> tpm_split 

map(tpm_split, ~.x %>% 
  mutate(col_id = paste0(strain, "::", developmental_stage)) %>% 
  dplyr::select(-developmental_stage, -type, -strain) %>% 
  spread(col_id, total_tpm) %>% 
    as.data.frame() %>%
    set_rowids(.)) -> tpm_mats

map(tpm_mats, ~cor(.x, method = "spearman")) -> tpm_cors

map2(tpm_cors,
    c("Mature Transcripts", 
      "pre-mRNA Transcripts"),
    ~Heatmap(.x, 
             column_title = .y,
             name = "Spearman Correlation\nof TPM values"))
```


## Classify transcripts into maternal or zygotic

Next I'll try to classify transcripts into catagories based upon their abundance changes throughout embryogenesis. 

First need to identify transcripts that are differentially expressed between maternal and zygotic states. Let's try a two state comparison of 
pre MZT and post MZT, which based on the correlation matrix shown above, is strongly segregated between the two.

```{r}
library(tximport)
library(edgeR)

colnames(dat)[colnames(dat) == "Name"] <- "transcript_id" 

tx2gene_mapping <- dplyr::select(dat, 
                         transcript_id, 
                         gene_id) %>% 
  unique() %>% 
  as.data.frame()

files <- dir("../data/salmon",
            recursive = T,
            pattern = "*quant.sf",
            full.names = T)

sra_ids <- dirname(files) %>% 
  str_split(., "/", simplify = T) %>% 
  .[, 4]


keep <- sra_ids %in% mdata$Run
files <- files[keep]
sra_ids <- sra_ids[keep]

all_dat <- map(files, read_tsv)
all_dat <- map(all_dat, ~dplyr::rename(.x, transcript_id = Name))

all_pre <- map(all_dat, 
                 ~dplyr::filter(.x,
                                str_detect(transcript_id, "^pre_")))

all_mat <- map(all_dat, 
               ~dplyr::filter(.x,
                              !str_detect(transcript_id, "^pre_")))

dir.create("salmon_modified", recursive = T)
walk2(all_pre,
      sra_ids,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".premRNA"))))
walk2(all_mat,
      sra_ids,
      ~write_tsv(.x, file.path("salmon_modified", 
                               paste0(.y, ".maturemRNA"))))
```


```{r}
files <- dir("salmon_modified", full.names = T)

mature <- str_subset(files, "maturemRNA$")
pre <- str_subset(files, "premRNA$")

matures <- tximport(mature, 
                    countsFromAbundance = "no",
         type = "none",
         tx2gene = tx2gene_mapping,
         geneIdCol = "gene_id",
         txIdCol = "transcript_id",
         abundanceCol = "TPM",
         countsCol = "NumReads",
         lengthCol = "EffectiveLength")

pres <- tximport(pre, 
                    countsFromAbundance = "no",
         type = "none",
         tx2gene = tx2gene_mapping,
         geneIdCol = "gene_id",
         txIdCol = "transcript_id",
         abundanceCol = "TPM",
         countsCol = "NumReads",
         lengthCol = "EffectiveLength")

# make metadata sample ordered by columns in count matrix

pdata <- inner_join(data_frame(Run = sra_ids),
           mdata, by = "Run")

pdata <- mutate(pdata, 
                mzt = ifelse(str_detect(developmental_stage,
                                        "^[2345]-[3456] hr embryo"),
                             "zygotic",
                             "maternal"))
```


```{r deseq2}
library(DESeq2)
library(viridis)
dds <- DESeqDataSetFromTximport(matures, pdata, ~mzt)
dds_mature <- DESeq(dds)
res_mature <- results(dds_mature)


dds <- DESeqDataSetFromTximport(pres, pdata, ~mzt)
dds_pre <- DESeq(dds)
res_pre <- results(dds_pre)
res_pre_tidy <- results(dds_pre, tidy = TRUE)
```


```{r}
pre_plot_dat <- res_pre_tidy %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj < 0.05 & log2FoldChange > 0, "Up", 
                      ifelse(padj < 0.05 & log2FoldChange < 0, "Down", "Not Sig")))
n_sig_up <- nrow(dplyr::filter(pre_plot_dat, sig == "Up")) %>% 
  formatC(big.mark = ",")
n_sig_down <- nrow(dplyr::filter(pre_plot_dat, sig == "Down"))  %>% 
  formatC(big.mark = ",")
n_notsig <- nrow(dplyr::filter(pre_plot_dat, sig == "Not Sig"))  %>% 
  formatC(big.mark = ",")

p <- ggplot(pre_plot_dat, 
       aes(baseMean, log2FoldChange)) +
  geom_point(aes(colour = sig), size = 0.1) +
  scale_colour_viridis(discrete = TRUE,
                       name = "",
                       labels = c(paste0("Sig down (n = ", n_sig_down, ")"),
                                  paste0("Not Sig (n= ", n_notsig, ")"),
                                  paste0("Sig up (n = ", n_sig_up, ")"))) +
  scale_x_log10() + 
  labs(x = expression(paste("log"[10], " Average Expression")),
       y = expression(paste("log"[2], " ", frac("Zygotic", "Maternal")))) +
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(legend.position = "top")
  
p
save_plot("MA_zm_deseq_premRNA.pdf", p, base_width = 6)
```

```{r heatmap_zygotic, fig.width=4}
res_pre_sig <- res_pre_tidy %>% 
  as_tibble() %>% 
  dplyr::filter(!is.na(padj), padj < 0.05, log2FoldChange > 0)

dds_pre <- rlog(dds_pre)

# add in colnames
cols <- pdata %>% 
  mutate(cnames = str_c(developmental_stage, strain , sep = "::")) %>% 
  pull(cnames)

rlog_counts <- assay(dds_pre)
colnames(rlog_counts) <- cols

scale_pre_mat <-t(scale(t(rlog_counts)))

sig_pre_mat <- scale_pre_mat[rownames(scale_pre_mat) %in% res_pre_sig$row, ]

Heatmap(sig_pre_mat, 
        name = "pre-mRNA\nabundance\nZ-score",
        show_row_names = F,
        show_row_dend = F)

pdf("premRNA_abundance.pdf")
Heatmap(sig_pre_mat, 
        name = "pre-mRNA\nabundance\nZ-score",
        show_row_names = F,
        show_row_dend = F)
dev.off()

dds_pre <- rlog(dds_pre)


rlog_counts <- assay(dds_mature)
colnames(rlog_counts) <- cols

scale_mature_mat <-t(scale(t(rlog_counts)))

scale_mature_mat <- scale_mature_mat[rownames(scale_mature_mat) %in% res_pre_sig$row, ]

scale_mature_mat <- na.omit(scale_mature_mat)
Heatmap(scale_mature_mat, 
        name = "mature-mRNA abundance",
        show_row_names = F,
        show_row_dend = F)
```


### Eisen classification

```{r eisen_dat}
library("org.Dm.eg.db")
edata <- read_tsv("https://ndownloader.figshare.com/files/400050")

res_pre_tidy
res_pre_tidy$symbol <- mapIds(org.Dm.eg.db, 
                            keys=res_pre_tidy$row, 
                            column="SYMBOL", 
                            keytype="ENSEMBL",
                            multiVals="first")

res_pre_tidy$entrez <- mapIds(org.Dm.eg.db, 
                            keys=res_pre_tidy$row, 
                            column="ENTREZID", 
                            keytype="ENSEMBL",
                            multiVals="first")

res_pre_tidy$name =   mapIds(org.Dm.eg.db,
                           keys=res_pre_tidy$row, 
                           column="GENENAME",
                           keytype="ENSEMBL",
                           multiVals="first")
    

res_pre_sig <- res_pre_tidy %>% 
  as_tibble() %>% 
  dplyr::filter(!is.na(padj), padj < 0.05, log2FoldChange > 0)


res_pre_sig <- edata %>% 
  dplyr::select(NAME, CLASS) %>% 
  left_join(res_pre_sig, ., by = c("symbol" = "NAME"))

res_pre_sig %>% dplyr::count(CLASS)
```

### Calc mean gene level TPMs to classify striclty maternal or zygotic. 

```{r mean_gene_tpms}
colnames(matures$abundance) <- sra_ids 
colnames(pres$abundance) <- sra_ids

# mean mature-mRNA TPMs
gene_tpm_mature <- matures$abundance %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  as_tibble() %>% gather(Run, expr, -gene_id) %>% 
  left_join(pdata, by = "Run") %>% 
  group_by(mzt, gene_id) %>% 
  summarize(mean_tpm = mean(expr)) %>% 
  spread(mzt, mean_tpm, sep = "_mature_")

# mean pre-mRNA TPMs
gene_tpm_pre <- pres$abundance %>% 
  as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  as_tibble() %>% gather(Run, expr, -gene_id) %>% 
  left_join(pdata, by = "Run") %>% 
  group_by(mzt, gene_id) %>% 
  summarize(mean_tpm = mean(expr)) %>% 
  spread(mzt, mean_tpm, sep = "_pre_")

mzt_tpms <- left_join(gene_tpm_mature,
                      gene_tpm_pre,
                      by = c("gene_id"))

#calc log2 ratios of zygotic / maternal

mzt_tpms <- mutate(mzt_tpms, 
                   zm_mature_ratio = log2(mzt_mature_zygotic / mzt_mature_maternal),
                   zm_pre_ratio = log2(mzt_pre_zygotic / mzt_pre_maternal))
```

```{r zm_ratio_hists}
plot_dat <- mzt_tpms %>% 
  dplyr::select(gene_id, zm_mature_ratio, zm_pre_ratio) %>% 
  dplyr::rename(mature = zm_mature_ratio,
                `pre-mRNA` = zm_pre_ratio) %>% 
  gather(stage, zm_ratio, -gene_id) 

p <- ggplot(plot_dat, aes(zm_ratio)) +
  geom_density(aes(fill = stage), alpha = 0.75) +
  scale_fill_viridis(discrete = TRUE,
                     name = "") +
  geom_vline(xintercept = 0, linetype = "dashed") +
  labs(x = expression(paste("Log"[2], " ", 
                            frac("Zygotic", "Maternal"), 
                            " TPM"))) +
  theme_cowplot() +
  theme(legend.position = "top")

p
save_plot("zm_ratio.pdf", p, base_height = 4)
```

## Strictly Zygotic transcripts

```{r tpm_cutoff}
min_tpm <- 0.1
```
  
  Transcripts that are only expressed in zygotic samples, are by definition newlly transcribed. These will be defined as those transcripts with < `r min_tpm` TPM in maternal stages, and higher than `r min_tpm` in zygotic stages. 
  
```{r exprsed}
mzt_tpms <- mzt_tpms %>% 
  dplyr::filter(mzt_mature_maternal > min_tpm | 
                  mzt_mature_zygotic > min_tpm)
```

  First genes that are not expressed sufficiently high in either stage are dropped, resulting in `r nrow(mzt_tpms)` expressed genes.
  
```{r zygotic_txs}

zygotic <- mzt_tpms %>% 
    dplyr::filter(mzt_mature_maternal <= min_tpm, 
                mzt_mature_zygotic > min_tpm)
                
plot_dat <- zygotic %>% 
  dplyr::select(gene_id, zm_mature_ratio, zm_pre_ratio) %>% 
  dplyr::rename(mature = zm_mature_ratio,
                `pre-mRNA` = zm_pre_ratio) %>% 
  gather(stage, zm_ratio, -gene_id) 

p <- ggplot(plot_dat, aes(zm_ratio)) +
  geom_density(aes(fill = stage), alpha = 0.75) +
  scale_fill_viridis(discrete = TRUE,
                     name = "") +
  labs(x = expression(paste("Log"[2], " ", 
                            frac("Zygotic", "Maternal"), 
                            " TPM"))) +
  theme_cowplot() +
  theme(legend.position = "top")

p


zygotic$gene_id

zygotic_mat <- all_zmat[rownames(all_zmat) %in% mzt_tpms$gene_id, ]

Heatmap(na.omit(zygotic_mat), name = "pre-mRNA proportion\n (Z-Score)",
        column_title = "All datasets",
                    show_row_names = F, 
                    show_row_dend = F) 

```

### Maternal + Zygotic transcripts

  Transcripts expressed in the oocyte are maternal (> `r min_tpm` TPM), and if these transcripts increase in abundance in the zygote (either through `exonic` or `intronic` single increases), then they are maternal + zygotic. 
  
```{r mat_and_zy, eval = F}



```
To do:
  - classify early versus late zygotic and look for Zelda (zfl)